<!DOCTYPE html>
<!-- START: inst/pkgdown/templates/layout.html --><!-- Generated by pkgdown: do not edit by hand --><html lang="en" data-bs-theme="auto">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<title>Data Wrangling and Processing for Genomics: All in One View</title>
<meta name="viewport" content="width=device-width, initial-scale=1">
<script src="../assets/themetoggle.js"></script><link rel="stylesheet" type="text/css" href="../assets/styles.css">
<script src="../assets/scripts.js" type="text/javascript"></script><!-- mathjax --><script type="text/x-mathjax-config">
    MathJax.Hub.Config({
      config: ["MMLorHTML.js"],
      jax: ["input/TeX","input/MathML","output/HTML-CSS","output/NativeMML", "output/PreviewHTML"],
      extensions: ["tex2jax.js","mml2jax.js","MathMenu.js","MathZoom.js", "fast-preview.js", "AssistiveMML.js", "a11y/accessibility-menu.js"],
      TeX: {
        extensions: ["AMSmath.js","AMSsymbols.js","noErrors.js","noUndefined.js"]
      },
      tex2jax: {
        inlineMath: [['\\(', '\\)']],
        displayMath: [ ['$$','$$'], ['\\[', '\\]'] ],
        processEscapes: true
      }
    });
    </script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><!-- Responsive Favicon for The Carpentries --><link rel="apple-touch-icon" sizes="180x180" href="../favicons/dc/apple-touch-icon.png">
<link rel="icon" type="image/png" sizes="32x32" href="../favicons/dc/favicon-32x32.png">
<link rel="icon" type="image/png" sizes="16x16" href="../favicons/dc/favicon-16x16.png">
<link rel="manifest" href="../favicons/dc/site.webmanifest">
<link rel="mask-icon" href="../favicons/dc/safari-pinned-tab.svg" color="#5bbad5">
<meta name="msapplication-TileColor" content="#da532c">
<meta name="theme-color" media="(prefers-color-scheme: light)" content="white">
<meta name="theme-color" media="(prefers-color-scheme: dark)" content="black">
</head>
<body>
    <header id="top" class="navbar navbar-expand-md top-nav data"><svg xmlns="http://www.w3.org/2000/svg" class="d-none"><symbol id="check2" viewbox="0 0 16 16"><path d="M13.854 3.646a.5.5 0 0 1 0 .708l-7 7a.5.5 0 0 1-.708 0l-3.5-3.5a.5.5 0 1 1 .708-.708L6.5 10.293l6.646-6.647a.5.5 0 0 1 .708 0z"></path></symbol><symbol id="circle-half" viewbox="0 0 16 16"><path d="M8 15A7 7 0 1 0 8 1v14zm0 1A8 8 0 1 1 8 0a8 8 0 0 1 0 16z"></path></symbol><symbol id="moon-stars-fill" viewbox="0 0 16 16"><path d="M6 .278a.768.768 0 0 1 .08.858 7.208 7.208 0 0 0-.878 3.46c0 4.021 3.278 7.277 7.318 7.277.527 0 1.04-.055 1.533-.16a.787.787 0 0 1 .81.316.733.733 0 0 1-.031.893A8.349 8.349 0 0 1 8.344 16C3.734 16 0 12.286 0 7.71 0 4.266 2.114 1.312 5.124.06A.752.752 0 0 1 6 .278z"></path><path d="M10.794 3.148a.217.217 0 0 1 .412 0l.387 1.162c.173.518.579.924 1.097 1.097l1.162.387a.217.217 0 0 1 0 .412l-1.162.387a1.734 1.734 0 0 0-1.097 1.097l-.387 1.162a.217.217 0 0 1-.412 0l-.387-1.162A1.734 1.734 0 0 0 9.31 6.593l-1.162-.387a.217.217 0 0 1 0-.412l1.162-.387a1.734 1.734 0 0 0 1.097-1.097l.387-1.162zM13.863.099a.145.145 0 0 1 .274 0l.258.774c.115.346.386.617.732.732l.774.258a.145.145 0 0 1 0 .274l-.774.258a1.156 1.156 0 0 0-.732.732l-.258.774a.145.145 0 0 1-.274 0l-.258-.774a1.156 1.156 0 0 0-.732-.732l-.774-.258a.145.145 0 0 1 0-.274l.774-.258c.346-.115.617-.386.732-.732L13.863.1z"></path></symbol><symbol id="sun-fill" viewbox="0 0 16 16"><path d="M8 12a4 4 0 1 0 0-8 4 4 0 0 0 0 8zM8 0a.5.5 0 0 1 .5.5v2a.5.5 0 0 1-1 0v-2A.5.5 0 0 1 8 0zm0 13a.5.5 0 0 1 .5.5v2a.5.5 0 0 1-1 0v-2A.5.5 0 0 1 8 13zm8-5a.5.5 0 0 1-.5.5h-2a.5.5 0 0 1 0-1h2a.5.5 0 0 1 .5.5zM3 8a.5.5 0 0 1-.5.5h-2a.5.5 0 0 1 0-1h2A.5.5 0 0 1 3 8zm10.657-5.657a.5.5 0 0 1 0 .707l-1.414 1.415a.5.5 0 1 1-.707-.708l1.414-1.414a.5.5 0 0 1 .707 0zm-9.193 9.193a.5.5 0 0 1 0 .707L3.05 13.657a.5.5 0 0 1-.707-.707l1.414-1.414a.5.5 0 0 1 .707 0zm9.193 2.121a.5.5 0 0 1-.707 0l-1.414-1.414a.5.5 0 0 1 .707-.707l1.414 1.414a.5.5 0 0 1 0 .707zM4.464 4.465a.5.5 0 0 1-.707 0L2.343 3.05a.5.5 0 1 1 .707-.707l1.414 1.414a.5.5 0 0 1 0 .708z"></path></symbol></svg><a class="visually-hidden-focusable skip-link" href="#main-content">Skip to main content</a>
  <div class="container-fluid top-nav-container">
    <div class="col-md-8">
      <div class="large-logo">
        <img id="dc-logo" alt="Data Carpentry" src="../assets/images/data-logo.svg">
</div>
    </div>
    <div class="selector-container">
      <div id="theme-selector">
        <li class="nav-item dropdown" id="theme-button-list">
          <button class="btn btn-link nav-link px-0 px-lg-2 dropdown-toggle d-flex align-items-center" id="bd-theme" type="button" aria-expanded="false" data-bs-toggle="dropdown" data-bs-display="static" aria-label="Toggle theme (auto)">
            <svg class="bi my-1 theme-icon-active"><use href="#circle-half"></use></svg><i data-feather="chevron-down"></i>
          </button>
          <ul class="dropdown-menu dropdown-menu-end" aria-labelledby="bd-theme-text">
<li>
              <button type="button" class="btn dropdown-item d-flex align-items-center" data-bs-theme-value="light" aria-pressed="false">
                <svg class="bi me-2 theme-icon"><use href="#sun-fill"></use></svg>
                Light
                <svg class="bi ms-auto d-none"><use href="#check2"></use></svg></button>
            </li>
            <li>
              <button type="button" class="btn dropdown-item d-flex align-items-center" data-bs-theme-value="dark" aria-pressed="false">
                <svg class="bi me-2 theme-icon"><use href="#moon-stars-fill"></use></svg>
                Dark
                <svg class="bi ms-auto d-none"><use href="#check2"></use></svg></button>
            </li>
            <li>
              <button type="button" class="btn dropdown-item d-flex align-items-center active" data-bs-theme-value="auto" aria-pressed="true">
                <svg class="bi me-2 theme-icon"><use href="#circle-half"></use></svg>
                Auto
                <svg class="bi ms-auto d-none"><use href="#check2"></use></svg></button>
            </li>
          </ul>
</li>
      </div>

      <div class="dropdown" id="instructor-dropdown">
        <button class="btn btn-secondary dropdown-toggle bordered-button" type="button" id="dropdownMenu1" data-bs-toggle="dropdown" aria-expanded="false">
          <i aria-hidden="true" class="icon" data-feather="eye"></i> Instructor View <i data-feather="chevron-down"></i>
        </button>
        <ul class="dropdown-menu" aria-labelledby="dropdownMenu1">
<li><button class="dropdown-item" type="button" onclick="window.location.href='../aio.html';">Learner View</button></li>
        </ul>
</div>
    </div>
  </div>
  <hr></header><nav class="navbar navbar-expand-xl bottom-nav data" aria-label="Main Navigation"><div class="container-fluid nav-container">
    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbarSupportedContent" aria-controls="navbarSupportedContent" aria-expanded="false" aria-label="Toggle Navigation">
      <span class="navbar-toggler-icon"></span>
      <span class="menu-title">Menu</span>
    </button>
    <div class="nav-logo">
      <img class="small-logo" alt="Data Carpentry" src="../assets/images/data-logo-sm.svg">
</div>
    <div class="lesson-title-md">
      Data Wrangling and Processing for Genomics
    </div>
    <div class="search-icon-sm">
      <!-- TODO: do not show until we have search
        <i role="img" aria-label="Search the All In One page" data-feather="search"></i>
      -->
    </div>
    <div class="desktop-nav">
      <ul class="navbar-nav me-auto mb-2 mb-lg-0">
<li class="nav-item">
          <span class="lesson-title">
            Data Wrangling and Processing for Genomics
          </span>
        </li>
        <li class="nav-item">
          <a class="nav-link" href="../instructor/key-points.html">Key Points</a>
        </li>
        <li class="nav-item">
          <a class="nav-link" href="../instructor/instructor-notes.html">Instructor Notes</a>
        </li>
        <li class="nav-item">
          <a class="nav-link" href="../instructor/images.html">Extract All Images</a>
        </li>
        <li class="nav-item dropdown">
          <button class="nav-link dropdown-toggle" id="navbarDropdown" data-bs-toggle="dropdown" aria-expanded="false">
            More <i data-feather="chevron-down"></i>
          </button>
          <ul class="dropdown-menu" aria-labelledby="navbarDropdown">
<hr>
<li><a class="dropdown-item" href="discuss.html">Discussion</a></li>
<li><a class="dropdown-item" href="reference.html">Glossary</a></li>
          </ul>
</li>
      </ul>
</div>
    <!--
    <form class="d-flex col-md-2 search-form">
      <fieldset disabled>
      <input class="form-control me-2 searchbox" type="search" placeholder="" aria-label="">
        <button class="btn btn-outline-success tablet-search-button"  type="submit">
          <i class="search-icon" data-feather="search" role="img" aria-label="Search the All In One page"></i>
        </button>
      </fieldset>
    </form>
    -->
    <a id="search-button" class="btn btn-primary" href="../instructor/aio.html" role="button" aria-label="Search the All In One page">Search the All In One page</a>
  </div>
<!--/div.container-fluid -->
</nav><div class="col-md-12 mobile-title">
  Data Wrangling and Processing for Genomics
</div>

<aside class="col-md-12 lesson-progress"><div style="width: %" class="percentage">
    %
  </div>
  <div class="progress data">
    <div class="progress-bar data" role="progressbar" style="width: %" aria-valuenow="" aria-label="Lesson Progress" aria-valuemin="0" aria-valuemax="100">
    </div>
  </div>
</aside><div class="container">
      <div class="row">
        <!-- START: inst/pkgdown/templates/navbar.html -->
<div id="sidebar-col" class="col-lg-4">
  <div id="sidebar" class="sidebar">
      <nav aria-labelledby="flush-headingEleven"><button role="button" aria-label="close menu" alt="close menu" aria-expanded="true" aria-controls="sidebar" class="collapse-toggle" data-collapse="Collapse " data-episodes="Episodes ">
          <i class="search-icon" data-feather="x" role="img"></i>
        </button>
        <div class="sidebar-inner">
          <div class="row mobile-row" id="theme-row-mobile">
            <div class="col" id="theme-selector">
              <li class="nav-item dropdown" id="theme-button-list">
                <button class="btn btn-link nav-link px-0 px-lg-2 dropdown-toggle d-flex align-items-center" id="bd-theme" type="button" aria-expanded="false" data-bs-toggle="dropdown" data-bs-display="static" aria-label="Toggle theme (auto)">
                  <svg class="bi my-1 theme-icon-active"><use href="#circle-half"></use></svg><span class="d-lg-none ms-1" id="bd-theme-text">Toggle Theme</span>
                </button>
                <ul class="dropdown-menu dropdown-menu-right" aria-labelledby="bd-theme-text">
<li>
                    <button type="button" class="btn dropdown-item d-flex align-items-center" data-bs-theme-value="light" aria-pressed="false">
                      <svg class="bi me-2 theme-icon"><use href="#sun-fill"></use></svg>
                      Light
                      <svg class="bi ms-auto d-none"><use href="#check2"></use></svg></button>
                  </li>
                  <li>
                    <button type="button" class="btn dropdown-item d-flex align-items-center" data-bs-theme-value="dark" aria-pressed="false">
                      <svg class="bi me-2 theme-icon"><use href="#moon-stars-fill"></use></svg>
                      Dark
                      <svg class="bi ms-auto d-none"><use href="#check2"></use></svg></button>
                  </li>
                  <li>
                    <button type="button" class="btn dropdown-item d-flex align-items-center active" data-bs-theme-value="auto" aria-pressed="true">
                      <svg class="bi me-2 theme-icon"><use href="#circle-half"></use></svg>
                      Auto
                      <svg class="bi ms-auto d-none"><use href="#check2"></use></svg></button>
                  </li>
                </ul>
</li>
            </div>
          </div>
          <div class="row mobile-row">
            <div class="col">
              <div class="sidenav-view-selector">
                <div class="accordion accordion-flush" id="accordionFlush9">
                  <div class="accordion-item">
                    <h2 class="accordion-header" id="flush-headingNine">
                      <button class="accordion-button collapsed" id="instructor" type="button" data-bs-toggle="collapse" data-bs-target="#flush-collapseNine" aria-expanded="false" aria-controls="flush-collapseNine">
                        <i id="eye" aria-hidden="true" class="icon" data-feather="eye"></i> Instructor View
                      </button>
                    </h2>
                    <div id="flush-collapseNine" class="accordion-collapse collapse" aria-labelledby="flush-headingNine" data-bs-parent="#accordionFlush2">
                      <div class="accordion-body">
                        <a href="../aio.html">Learner View</a>
                      </div>
                    </div>
                  </div>
<!--/div.accordion-item-->
                </div>
<!--/div.accordion-flush-->
              </div>
<!--div.sidenav-view-selector -->
            </div>
<!--/div.col -->

            <hr>
</div>
<!--/div.mobile-row -->

          <div class="accordion accordion-flush" id="accordionFlush11">
            <div class="accordion-item">

              <button id="chapters" class="accordion-button show" type="button" data-bs-toggle="collapse" data-bs-target="#flush-collapseEleven" aria-expanded="false" aria-controls="flush-collapseEleven">
                <h2 class="accordion-header chapters" id="flush-headingEleven">
                  EPISODES
                </h2>
              </button>
              <div id="flush-collapseEleven" class="accordion-collapse show collapse" aria-labelledby="flush-headingEleven" data-bs-parent="#accordionFlush11">

                <div class="accordion-body">
                  <div class="accordion accordion-flush" id="accordionFlush1">
  <div class="accordion-item">
    <div class="accordion-header" id="flush-heading1">
        <a href="index.html">Summary and Schedule</a>
    </div>
<!--/div.accordion-header-->

  </div>
<!--/div.accordion-item-->
</div>
<!--/div.accordion-flush-->

<div class="accordion accordion-flush" id="accordionFlush2">
  <div class="accordion-item">
    <div class="accordion-header" id="flush-heading2">
        <a href="01-background.html">1. Background and Metadata</a>
    </div>
<!--/div.accordion-header-->

  </div>
<!--/div.accordion-item-->
</div>
<!--/div.accordion-flush-->

<div class="accordion accordion-flush" id="accordionFlush3">
  <div class="accordion-item">
    <div class="accordion-header" id="flush-heading3">
        <a href="02-quality-control.html">2. Assessing Read Quality</a>
    </div>
<!--/div.accordion-header-->

  </div>
<!--/div.accordion-item-->
</div>
<!--/div.accordion-flush-->

<div class="accordion accordion-flush" id="accordionFlush4">
  <div class="accordion-item">
    <div class="accordion-header" id="flush-heading4">
        <a href="03-trimming.html">3. Trimming and Filtering</a>
    </div>
<!--/div.accordion-header-->

  </div>
<!--/div.accordion-item-->
</div>
<!--/div.accordion-flush-->

<div class="accordion accordion-flush" id="accordionFlush5">
  <div class="accordion-item">
    <div class="accordion-header" id="flush-heading5">
        <a href="04-variant_calling.html">4. Variant Calling Workflow</a>
    </div>
<!--/div.accordion-header-->

  </div>
<!--/div.accordion-item-->
</div>
<!--/div.accordion-flush-->

<div class="accordion accordion-flush" id="accordionFlush6">
  <div class="accordion-item">
    <div class="accordion-header" id="flush-heading6">
        <a href="05-automation.html">5. Automating a Variant Calling Workflow</a>
    </div>
<!--/div.accordion-header-->

  </div>
<!--/div.accordion-item-->
</div>
<!--/div.accordion-flush-->

                </div>
              </div>
            </div>

            <hr class="half-width">
<div class="accordion accordion-flush lesson-resources" id="accordionFlush12">
              <div class="accordion-item">
                <h2 class="accordion-header" id="flush-headingTwelve">
                  <button class="accordion-button collapsed" id="lesson-resources" type="button" data-bs-toggle="collapse" data-bs-target="#flush-collapseTwelve" aria-expanded="false" aria-controls="flush-collapseTwelve">
                    RESOURCES
                  </button>
                </h2>
                <div id="flush-collapseTwelve" class="accordion-collapse collapse" aria-labelledby="flush-headingTwelve" data-bs-parent="#accordionFlush12">
                  <div class="accordion-body">
                    <ul>
<li>
                        <a href="../instructor/key-points.html">Key Points</a>
                      </li>
                      <li>
                        <a href="../instructor/instructor-notes.html">Instructor Notes</a>
                      </li>
                      <li>
                        <a href="../instructor/images.html">Extract All Images</a>
                      </li>
                      <hr>
<li><a class="dropdown-item" href="discuss.html">Discussion</a></li>
<li><a class="dropdown-item" href="reference.html">Glossary</a></li>
                    </ul>
</div>
                </div>
              </div>
            </div>
            <hr class="half-width lesson-resources">
<a href="../instructor/aio.html">See all in one page</a>


            <hr class="d-none d-sm-block d-md-none">
<div class="d-grid gap-1">

            </div>
          </div>
<!-- /div.accordion -->
        </div>
<!-- /div.sidebar-inner -->
      </nav>
</div>
<!-- /div.sidebar -->
  </div>
<!-- /div.sidebar-col -->
<!-- END:   inst/pkgdown/templates/navbar.html-->

        <!-- START: inst/pkgdown/templates/content-extra.html -->
  <div class="col-xl-8 col-lg-12 primary-content">
    <main id="main-content" class="main-content"><div class="container lesson-content">
        
        
<section id="aio-01-background"><p>Content from <a href="01-background.html">Background and Metadata</a></p>
<hr>
<p>Last updated on 2023-05-04 |

        <a href="https://github.com/datacarpentry/wrangling-genomics/edit/main/episodes/01-background.md" class="external-link">Edit this page <i aria-hidden="true" data-feather="edit"></i></a></p>
<p>Estimated time: <i aria-hidden="true" data-feather="clock"></i> 15 minutes</p>
<div class="text-end">
          <button role="button" aria-pressed="false" tabindex="0" id="expand-code" class="pull-right" data-expand="Expand All Solutions " data-collapse="Collapse All Solutions "> Expand All Solutions <i aria-hidden="true" data-feather="plus"></i></button>
        </div>
<div class="overview card">
<h2 class="card-header">Overview</h2>
<div class="row g-0">
<div class="col-md-4">
<div class="card-body">
<div class="inner">
<h3 class="card-title">Questions</h3>
<ul>
<li>What data are we using?</li>
<li>Why is this experiment important?</li>
</ul>
</div>
</div>
</div>
<div class="col-md-8">
<div class="card-body">
<div class="inner bordered">
<h3 class="card-title">Objectives</h3>
<ul>
<li>Why study <em>E. coli</em>?</li>
<li>Understand the data set.</li>
<li>What is hypermutability?</li>
</ul>
</div>
</div>
</div>
</div>
</div>
<section><h2 class="section-heading" id="background">Background<a class="anchor" aria-label="anchor" href="#background"></a>
</h2>
<hr class="half-width">
<p>We are going to use a long-term sequencing dataset from a population
of <em>Escherichia coli</em>.</p>
<ul>
<li>
<strong>What is <em>E. coli</em>?</strong>
<ul>
<li>
<em>E. coli</em> are rod-shaped bacteria that can survive under a
wide variety of conditions including variable temperatures, nutrient
availability, and oxygen levels. Most strains are harmless, but some are
associated with food-poisoning.</li>
</ul>
</li>
</ul>
<figure><img src="../fig/172px-EscherichiaColi_NIAID.jpg" alt="Wikimedia" class="figure mx-auto d-block"></figure><!-- https://species.wikimedia.org/wiki/Escherichia_coli#/media/File:EscherichiaColi_NIAID.jpg --><ul>
<li>
<strong>Why is <em>E. coli</em> important?</strong>
<ul>
<li>
<em>E. coli</em> are one of the most well-studied model organisms in
science. As a single-celled organism, <em>E. coli</em> reproduces
rapidly, typically doubling its population every 20 minutes, which means
it can be manipulated easily in experiments. In addition, most naturally
occurring strains of <em>E. coli</em> are harmless. Most importantly,
the genetics of <em>E. coli</em> are fairly well understood and can be
manipulated to study adaptation and evolution.</li>
</ul>
</li>
</ul></section><section><h2 class="section-heading" id="the-data">The data<a class="anchor" aria-label="anchor" href="#the-data"></a>
</h2>
<hr class="half-width">
<ul>
<li><p>The data we are going to use is part of a long-term evolution
experiment led by <a href="https://en.wikipedia.org/wiki/E._coli_long-term_evolution_experiment" class="external-link">Richard
Lenski</a>.</p></li>
<li><p>The experiment was designed to assess adaptation in <em>E.
coli</em>. A population was propagated for more than 40,000 generations
in a glucose-limited minimal medium (in most conditions glucose is the
best carbon source for <em>E. coli</em>, providing faster growth than
other sugars). This medium was supplemented with citrate, which <em>E.
coli</em> cannot metabolize in the aerobic conditions of the experiment.
Sequencing of the populations at regular time points revealed that
spontaneous citrate-using variant (<strong>Cit+</strong>) appeared
between 31,000 and 31,500 generations, causing an increase in population
size and diversity. In addition, this experiment showed hypermutability
in certain regions. Hypermutability is important and can help accelerate
adaptation to novel environments, but also can be selected against in
well-adapted populations.</p></li>
<li><p>To see a timeline of the experiment to date, check out this <a href="https://en.wikipedia.org/wiki/E._coli_long-term_evolution_experiment#/media/File:LTEE_Timeline_as_of_May_28,_2016.png" class="external-link">figure</a>,
and this paper <a href="https://www.pnas.org/content/105/23/7899" class="external-link">Blount
et al. 2008: Historical contingency and the evolution of a key
innovation in an experimental population of <em>Escherichia
coli</em></a>.</p></li>
</ul>
<div class="section level3">
<h3 id="view-the-metadata">View the metadata<a class="anchor" aria-label="anchor" href="#view-the-metadata"></a>
</h3>
<p>We will be working with three sample events from the
<strong>Ara-3</strong> strain of this experiment, one from 5,000
generations, one from 15,000 generations, and one from 50,000
generations. The population changed substantially during the course of
the experiment, and we will be exploring how (the evolution of a
<strong>Cit+</strong> mutant and <strong>hypermutability</strong>) with
our variant calling workflow. The metadata file associated with this
lesson can be <a href="../files/Ecoli_metadata_composite.csv">downloaded
directly here</a> or <a href="https://github.com/datacarpentry/wrangling-genomics/blob/main/episodes/files/Ecoli_metadata_composite.csv" class="external-link">viewed
in Github</a>. If you would like to know details of how the file was
created, you can look at <a href="https://github.com/datacarpentry/wrangling-genomics/blob/main/episodes/files/Ecoli_metadata_composite_README.md" class="external-link">some
notes and sources here</a>.</p>
<p>This metadata describes information on the <em>Ara-3</em> clones and
the columns represent:</p>
<table class="table">
<thead><tr class="header">
<th>Column</th>
<th>Description</th>
</tr></thead>
<tbody>
<tr class="odd">
<td>strain</td>
<td>strain name</td>
</tr>
<tr class="even">
<td>generation</td>
<td>generation when sample frozen</td>
</tr>
<tr class="odd">
<td>clade</td>
<td>based on parsimony-based tree</td>
</tr>
<tr class="even">
<td>reference</td>
<td>study the samples were originally sequenced for</td>
</tr>
<tr class="odd">
<td>population</td>
<td>ancestral population group</td>
</tr>
<tr class="even">
<td>mutator</td>
<td>hypermutability mutant status</td>
</tr>
<tr class="odd">
<td>facility</td>
<td>facility samples were sequenced at</td>
</tr>
<tr class="even">
<td>run</td>
<td>Sequence read archive sample ID</td>
</tr>
<tr class="odd">
<td>read_type</td>
<td>library type of reads</td>
</tr>
<tr class="even">
<td>read_length</td>
<td>length of reads in sample</td>
</tr>
<tr class="odd">
<td>sequencing_depth</td>
<td>depth of sequencing</td>
</tr>
<tr class="even">
<td>cit</td>
<td>citrate-using mutant status</td>
</tr>
</tbody>
</table>
<div id="challenge" class="callout challenge">
<div class="callout-square">
<i class="callout-icon" data-feather="zap"></i>
</div>
<div id="challenge" class="callout-inner">
<h3 class="callout-title">Challenge</h3>
<div class="callout-content">
<p>Based on the metadata, can you answer the following questions?</p>
<ol style="list-style-type: decimal">
<li>How many different generations exist in the data?</li>
<li>How many rows and how many columns are in this data?</li>
<li>How many citrate+ mutants have been recorded in
<strong>Ara-3</strong>?</li>
<li>How many hypermutable mutants have been recorded in
<strong>Ara-3</strong>?</li>
</ol>
</div>
</div>
</div>
<div id="accordionSolution1" class="accordion challenge-accordion accordion-flush">
<div class="accordion-item">
<button class="accordion-button solution-button collapsed" type="button" data-bs-toggle="collapse" data-bs-target="#collapseSolution1" aria-expanded="false" aria-controls="collapseSolution1">
  <h4 class="accordion-header" id="headingSolution1"> Show me the solution </h4>
</button>
<div id="collapseSolution1" class="accordion-collapse collapse" data-bs-parent="#accordionSolution1" aria-labelledby="headingSolution1">
<div class="accordion-body">
<ol style="list-style-type: decimal">
<li>25 different generations</li>
<li>62 rows, 12 columns</li>
<li>10 citrate+ mutants</li>
<li>6 hypermutable mutants</li>
</ol>
</div>
</div>
</div>
</div>
<!-- can add some additional info relevant to interplay of hypermutability and Cit+ adaptations, but keep it simple for now -->
<div id="keypoints1" class="callout keypoints">
<div class="callout-square">
<i class="callout-icon" data-feather="key"></i>
</div>
<div class="callout-inner">
<h3 class="callout-title">Key Points</h3>
<div class="callout-content">
<ul>
<li>It is important to record and understand your experiment’s
metadata.</li>
</ul>
</div>
</div>
</div>
</div>
</section></section><section id="aio-02-quality-control"><p>Content from <a href="02-quality-control.html">Assessing Read Quality</a></p>
<hr>
<p>Last updated on 2023-05-04 |

        <a href="https://github.com/datacarpentry/wrangling-genomics/edit/main/episodes/02-quality-control.md" class="external-link">Edit this page <i aria-hidden="true" data-feather="edit"></i></a></p>
<p>Estimated time: <i aria-hidden="true" data-feather="clock"></i> 50 minutes</p>
<div class="text-end">
          <button role="button" aria-pressed="false" tabindex="0" id="expand-code" class="pull-right" data-expand="Expand All Solutions " data-collapse="Collapse All Solutions "> Expand All Solutions <i aria-hidden="true" data-feather="plus"></i></button>
        </div>
<div class="overview card">
<h2 class="card-header">Overview</h2>
<div class="row g-0">
<div class="col-md-4">
<div class="card-body">
<div class="inner">
<h3 class="card-title">Questions</h3>
<ul>
<li>How can I describe the quality of my data?</li>
</ul>
</div>
</div>
</div>
<div class="col-md-8">
<div class="card-body">
<div class="inner bordered">
<h3 class="card-title">Objectives</h3>
<ul>
<li>Explain how a FASTQ file encodes per-base quality scores.</li>
<li>Interpret a FastQC plot summarizing per-base quality across all
reads.</li>
<li>Use <code>for</code> loops to automate operations on multiple
files.</li>
</ul>
</div>
</div>
</div>
</div>
</div>
<section><h2 class="section-heading" id="bioinformatic-workflows">Bioinformatic workflows<a class="anchor" aria-label="anchor" href="#bioinformatic-workflows"></a>
</h2>
<hr class="half-width">
<p>When working with high-throughput sequencing data, the raw reads you
get off of the sequencer will need to pass through a number of different
tools in order to generate your final desired output. The execution of
this set of tools in a specified order is commonly referred to as a
<em>workflow</em> or a <em>pipeline</em>.</p>
<p>An example of the workflow we will be using for our variant calling
analysis is provided below with a brief description of each step.</p>
<figure><img src="../fig/variant_calling_workflow.png" alt="workflow" class="figure mx-auto d-block"></figure><ol style="list-style-type: decimal">
<li>Quality control - Assessing quality using FastQC</li>
<li>Quality control - Trimming and/or filtering reads (if
necessary)</li>
<li>Align reads to reference genome</li>
<li>Perform post-alignment clean-up</li>
<li>Variant calling</li>
</ol>
<p>These workflows in bioinformatics adopt a plug-and-play approach in
that the output of one tool can be easily used as input to another tool
without any extensive configuration. Having standards for data formats
is what makes this feasible. Standards ensure that data is stored in a
way that is generally accepted and agreed upon within the community. The
tools that are used to analyze data at different stages of the workflow
are therefore built under the assumption that the data will be provided
in a specific format.</p>
</section><section><h2 class="section-heading" id="starting-with-data">Starting with data<a class="anchor" aria-label="anchor" href="#starting-with-data"></a>
</h2>
<hr class="half-width">
<p>Often times, the first step in a bioinformatic workflow is getting
the data you want to work with onto a computer where you can work with
it. If you have outsourced sequencing of your data, the sequencing
center will usually provide you with a link that you can use to download
your data. Today we will be working with publicly available sequencing
data.</p>
<p>We are studying a population of <em>Escherichia coli</em> (designated
Ara-3), which were propagated for more than 50,000 generations in a
glucose-limited minimal medium. We will be working with three samples
from this experiment, one from 5,000 generations, one from 15,000
generations, and one from 50,000 generations. The population changed
substantially during the course of the experiment, and we will be
exploring how with our variant calling workflow.</p>
<p>The data are paired-end, so we will download two files for each
sample. We will use the <a href="https://www.ebi.ac.uk/ena" class="external-link">European
Nucleotide Archive</a> to get our data. The ENA “provides a
comprehensive record of the world’s nucleotide sequencing information,
covering raw sequencing data, sequence assembly information and
functional annotation.” The ENA also provides sequencing data in the
fastq format, an important format for sequencing reads that we will be
learning about today.</p>
<p>To download the data, run the commands below.</p>
<p>Here we are using the <code>-p</code> option for <code>mkdir</code>.
This option allows <code>mkdir</code> to create the new directory, even
if one of the parent directories does not already exist. It also
supresses errors if the directory already exists, without overwriting
that directory.</p>
<p>It will take about 15 minutes to download the files.</p>
<div class="codewrapper sourceCode" id="cb1">
<h3 class="code-label">BASH<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="sourceCode bash" tabindex="0"><code class="sourceCode bash"><span id="cb1-1"><a href="#cb1-1" tabindex="-1"></a><span class="fu">mkdir</span> <span class="at">-p</span> ~/dc_workshop/data/untrimmed_fastq/</span>
<span id="cb1-2"><a href="#cb1-2" tabindex="-1"></a><span class="bu">cd</span> ~/dc_workshop/data/untrimmed_fastq</span>
<span id="cb1-3"><a href="#cb1-3" tabindex="-1"></a></span>
<span id="cb1-4"><a href="#cb1-4" tabindex="-1"></a><span class="ex">curl</span> <span class="at">-O</span> ftp://ftp.sra.ebi.ac.uk/vol1/fastq/SRR258/004/SRR2589044/SRR2589044_1.fastq.gz</span>
<span id="cb1-5"><a href="#cb1-5" tabindex="-1"></a><span class="ex">curl</span> <span class="at">-O</span> ftp://ftp.sra.ebi.ac.uk/vol1/fastq/SRR258/004/SRR2589044/SRR2589044_2.fastq.gz</span>
<span id="cb1-6"><a href="#cb1-6" tabindex="-1"></a><span class="ex">curl</span> <span class="at">-O</span> ftp://ftp.sra.ebi.ac.uk/vol1/fastq/SRR258/003/SRR2584863/SRR2584863_1.fastq.gz</span>
<span id="cb1-7"><a href="#cb1-7" tabindex="-1"></a><span class="ex">curl</span> <span class="at">-O</span> ftp://ftp.sra.ebi.ac.uk/vol1/fastq/SRR258/003/SRR2584863/SRR2584863_2.fastq.gz</span>
<span id="cb1-8"><a href="#cb1-8" tabindex="-1"></a><span class="ex">curl</span> <span class="at">-O</span> ftp://ftp.sra.ebi.ac.uk/vol1/fastq/SRR258/006/SRR2584866/SRR2584866_1.fastq.gz</span>
<span id="cb1-9"><a href="#cb1-9" tabindex="-1"></a><span class="ex">curl</span> <span class="at">-O</span> ftp://ftp.sra.ebi.ac.uk/vol1/fastq/SRR258/006/SRR2584866/SRR2584866_2.fastq.gz</span></code></pre>
</div>
<div id="faster-option" class="callout">
<div class="callout-square">
<i class="callout-icon" data-feather="bell"></i>
</div>
<div id="faster-option" class="callout-inner">
<h3 class="callout-title">Faster option</h3>
<div class="callout-content">
<p>If your workshop is short on time or the venue’s internet connection
is weak or unstable, learners can avoid needing to download the data and
instead use the data files provided in the <code>.backup/</code>
directory.</p>
<div class="codewrapper sourceCode" id="cb2">
<h3 class="code-label">BASH<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="sourceCode bash" tabindex="0"><code class="sourceCode bash"><span id="cb2-1"><a href="#cb2-1" tabindex="-1"></a><span class="ex">$</span> cp ~/.backup/untrimmed_fastq/<span class="pp">*</span>fastq.gz .</span></code></pre>
</div>
<p>This command creates a copy of each of the files in the
<code>.backup/untrimmed_fastq/</code> directory that end in
<code>fastq.gz</code> and places the copies in the current working
directory (signified by <code>.</code>).</p>
</div>
</div>
</div>
<p>The data comes in a compressed format, which is why there is a
<code>.gz</code> at the end of the file names. This makes it faster to
transfer, and allows it to take up less space on our computer. Let’s
unzip one of the files so that we can look at the fastq format.</p>
<div class="codewrapper sourceCode" id="cb3">
<h3 class="code-label">BASH<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="sourceCode bash" tabindex="0"><code class="sourceCode bash"><span id="cb3-1"><a href="#cb3-1" tabindex="-1"></a><span class="ex">$</span> gunzip SRR2584863_1.fastq.gz</span></code></pre>
</div>
</section><section><h2 class="section-heading" id="quality-control">Quality control<a class="anchor" aria-label="anchor" href="#quality-control"></a>
</h2>
<hr class="half-width">
<p>We will now assess the quality of the sequence reads contained in our
fastq files.</p>
<figure><img src="../fig/var_calling_workflow_qc.png" alt="workflow_qc" class="figure mx-auto d-block"></figure><div class="section level3">
<h3 id="details-on-the-fastq-format">Details on the FASTQ format<a class="anchor" aria-label="anchor" href="#details-on-the-fastq-format"></a>
</h3>
<p>Although it looks complicated (and it is), we can understand the <a href="https://en.wikipedia.org/wiki/FASTQ_format" class="external-link">fastq</a> format with
a little decoding. Some rules about the format include…</p>
<table class="table">
<colgroup>
<col width="3%">
<col width="96%">
</colgroup>
<thead><tr class="header">
<th>Line</th>
<th>Description</th>
</tr></thead>
<tbody>
<tr class="odd">
<td>1</td>
<td>Always begins with ‘@’ and then information about the read</td>
</tr>
<tr class="even">
<td>2</td>
<td>The actual DNA sequence</td>
</tr>
<tr class="odd">
<td>3</td>
<td>Always begins with a ‘+’ and sometimes the same info in line 1</td>
</tr>
<tr class="even">
<td>4</td>
<td>Has a string of characters which represent the quality scores; must
have same number of characters as line 2</td>
</tr>
</tbody>
</table>
<p>We can view the first complete read in one of the files our dataset
by using <code>head</code> to look at the first four lines.</p>
<div class="codewrapper sourceCode" id="cb4">
<h3 class="code-label">BASH<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="sourceCode bash" tabindex="0"><code class="sourceCode bash"><span id="cb4-1"><a href="#cb4-1" tabindex="-1"></a><span class="ex">$</span> head <span class="at">-n</span> 4 SRR2584863_1.fastq</span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>@SRR2584863.1 HWI-ST957:244:H73TDADXX:1:1101:4712:2181/1
TTCACATCCTGACCATTCAGTTGAGCAAAATAGTTCTTCAGTGCCTGTTTAACCGAGTCACGCAGGGGTTTTTGGGTTACCTGATCCTGAGAGTTAACGGTAGAAACGGTCAGTACGTCAGAATTTACGCGTTGTTCGAACATAGTTCTG
+
CCCFFFFFGHHHHJIJJJJIJJJIIJJJJIIIJJGFIIIJEDDFEGGJIFHHJIJJDECCGGEGIIJFHFFFACD:BBBDDACCCCAA@@CA@C&gt;C3&gt;@5(8&amp;&gt;C:9?8+89&lt;4(:83825C(:A#########################</code></pre>
</div>
<p>Line 4 shows the quality for each nucleotide in the read. Quality is
interpreted as the probability of an incorrect base call (e.g. 1 in 10)
or, equivalently, the base call accuracy (e.g. 90%). To make it possible
to line up each individual nucleotide with its quality score, the
numerical score is converted into a code where each individual character
represents the numerical quality score for an individual nucleotide. For
example, in the line above, the quality score line is:</p>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>CCCFFFFFGHHHHJIJJJJIJJJIIJJJJIIIJJGFIIIJEDDFEGGJIFHHJIJJDECCGGEGIIJFHFFFACD:BBBDDACCCCAA@@CA@C&gt;C3&gt;@5(8&amp;&gt;C:9?8+89&lt;4(:83825C(:A#########################</code></pre>
</div>
<p>The numerical value assigned to each of these characters depends on
the sequencing platform that generated the reads. The sequencing machine
used to generate our data uses the standard Sanger quality PHRED score
encoding, using Illumina version 1.8 onwards. Each character is assigned
a quality score between 0 and 41 as shown in the chart below.</p>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>Quality encoding: !"#$%&amp;'()*+,-./0123456789:;&lt;=&gt;?@ABCDEFGHIJ
                   |         |         |         |         |
Quality score:    01........11........21........31........41</code></pre>
</div>
<p>Each quality score represents the probability that the corresponding
nucleotide call is incorrect. This quality score is logarithmically
based, so a quality score of 10 reflects a base call accuracy of 90%,
but a quality score of 20 reflects a base call accuracy of 99%. These
probability values are the results from the base calling algorithm and
depend on how much signal was captured for the base incorporation.</p>
<p>Looking back at our read:</p>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>@SRR2584863.1 HWI-ST957:244:H73TDADXX:1:1101:4712:2181/1
TTCACATCCTGACCATTCAGTTGAGCAAAATAGTTCTTCAGTGCCTGTTTAACCGAGTCACGCAGGGGTTTTTGGGTTACCTGATCCTGAGAGTTAACGGTAGAAACGGTCAGTACGTCAGAATTTACGCGTTGTTCGAACATAGTTCTG
+
CCCFFFFFGHHHHJIJJJJIJJJIIJJJJIIIJJGFIIIJEDDFEGGJIFHHJIJJDECCGGEGIIJFHFFFACD:BBBDDACCCCAA@@CA@C&gt;C3&gt;@5(8&amp;&gt;C:9?8+89&lt;4(:83825C(:A#########################</code></pre>
</div>
<p>we can now see that there is a range of quality scores, but that the
end of the sequence is very poor (<code>#</code> = a quality score of
2).</p>
<div id="exercise" class="callout challenge">
<div class="callout-square">
<i class="callout-icon" data-feather="zap"></i>
</div>
<div id="exercise" class="callout-inner">
<h3 class="callout-title">Exercise</h3>
<div class="callout-content">
<p>What is the last read in the <code>SRR2584863_1.fastq</code> file?
How confident are you in this read?</p>
</div>
</div>
</div>
<div id="accordionSolution1" class="accordion challenge-accordion accordion-flush">
<div class="accordion-item">
<button class="accordion-button solution-button collapsed" type="button" data-bs-toggle="collapse" data-bs-target="#collapseSolution1" aria-expanded="false" aria-controls="collapseSolution1">
  <h4 class="accordion-header" id="headingSolution1"> Show me the solution </h4>
</button>
<div id="collapseSolution1" class="accordion-collapse collapse" aria-labelledby="headingSolution1" data-bs-parent="#accordionSolution1">
<div class="accordion-body">
<div class="codewrapper sourceCode" id="cb9">
<h3 class="code-label">BASH<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="sourceCode bash" tabindex="0"><code class="sourceCode bash"><span id="cb9-1"><a href="#cb9-1" tabindex="-1"></a><span class="ex">$</span> tail <span class="at">-n</span> 4 SRR2584863_1.fastq</span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>@SRR2584863.1553259 HWI-ST957:245:H73R4ADXX:2:2216:21048:100894/1
CTGCAATACCACGCTGATCTTTCACATGATGTAAGAAAAGTGGGATCAGCAAACCGGGTGCTGCTGTGGCTAGTTGCAGCAAACCATGCAGTGAACCCGCCTGTGCTTCGCTATAGCCGTGACTGATGAGGATCGCCGGAAGCCAGCCAA
+
CCCFFFFFHHHHGJJJJJJJJJHGIJJJIJJJJIJJJJIIIIJJJJJJJJJJJJJIIJJJHHHHHFFFFFEEEEEDDDDDDDDDDDDDDDDDCDEDDBDBDDBDDDDDDDDDBDEEDDDD7@BDDDDDD&gt;AA&gt;?B?&lt;@BDD@BDC?BDA?</code></pre>
</div>
<p>This read has more consistent quality at its end than the first read
that we looked at, but still has a range of quality scores, most of them
high. We will look at variations in position-based quality in just a
moment.</p>
</div>
</div>
</div>
</div>
<p>At this point, lets validate that all the relevant tools are
installed. If you are using the AWS AMI then these <em>should</em> be
preinstalled.</p>
<div class="codewrapper sourceCode" id="cb11">
<h3 class="code-label">BASH<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="sourceCode bash" tabindex="0"><code class="sourceCode bash"><span id="cb11-1"><a href="#cb11-1" tabindex="-1"></a><span class="ex">$</span> fastqc <span class="at">-h</span></span>
<span id="cb11-2"><a href="#cb11-2" tabindex="-1"></a>            <span class="ex">FastQC</span> <span class="at">-</span> A high throughput sequence QC analysis tool</span>
<span id="cb11-3"><a href="#cb11-3" tabindex="-1"></a></span>
<span id="cb11-4"><a href="#cb11-4" tabindex="-1"></a><span class="ex">SYNOPSIS</span></span>
<span id="cb11-5"><a href="#cb11-5" tabindex="-1"></a></span>
<span id="cb11-6"><a href="#cb11-6" tabindex="-1"></a>        <span class="ex">fastqc</span> seqfile1 seqfile2 .. seqfileN</span>
<span id="cb11-7"><a href="#cb11-7" tabindex="-1"></a></span>
<span id="cb11-8"><a href="#cb11-8" tabindex="-1"></a>    <span class="ex">fastqc</span> [-o output dir] [--<span class="er">(</span><span class="ex">no</span><span class="kw">)</span><span class="ex">extract]</span> [-f fastq<span class="kw">|</span><span class="ex">bam</span><span class="kw">|</span><span class="ex">sam]</span></span>
<span id="cb11-9"><a href="#cb11-9" tabindex="-1"></a>           <span class="ex">[-c</span> contaminant file] seqfile1 .. seqfileN</span>
<span id="cb11-10"><a href="#cb11-10" tabindex="-1"></a></span>
<span id="cb11-11"><a href="#cb11-11" tabindex="-1"></a><span class="ex">DESCRIPTION</span></span>
<span id="cb11-12"><a href="#cb11-12" tabindex="-1"></a></span>
<span id="cb11-13"><a href="#cb11-13" tabindex="-1"></a>    <span class="ex">FastQC</span> reads a set of sequence files and produces from each one a quality</span>
<span id="cb11-14"><a href="#cb11-14" tabindex="-1"></a>    <span class="ex">control</span> report consisting of a number of different modules, each one of</span>
<span id="cb11-15"><a href="#cb11-15" tabindex="-1"></a>    <span class="fu">which</span> will help to identify a different potential type of problem in your</span>
<span id="cb11-16"><a href="#cb11-16" tabindex="-1"></a>    <span class="ex">data.</span></span>
<span id="cb11-17"><a href="#cb11-17" tabindex="-1"></a></span>
<span id="cb11-18"><a href="#cb11-18" tabindex="-1"></a>    <span class="ex">If</span> no files to process are specified on the command line then the program</span>
<span id="cb11-19"><a href="#cb11-19" tabindex="-1"></a>    <span class="ex">will</span> start as an interactive graphical application.  If files are provided</span>
<span id="cb11-20"><a href="#cb11-20" tabindex="-1"></a>    <span class="ex">on</span> the command line then the program will run with no user interaction</span>
<span id="cb11-21"><a href="#cb11-21" tabindex="-1"></a>    <span class="ex">required.</span>  In this mode it is suitable for inclusion into a standardised</span>
<span id="cb11-22"><a href="#cb11-22" tabindex="-1"></a>    <span class="ex">analysis</span> pipeline.</span>
<span id="cb11-23"><a href="#cb11-23" tabindex="-1"></a></span>
<span id="cb11-24"><a href="#cb11-24" tabindex="-1"></a>    <span class="ex">The</span> options for the program as as follows:</span>
<span id="cb11-25"><a href="#cb11-25" tabindex="-1"></a></span>
<span id="cb11-26"><a href="#cb11-26" tabindex="-1"></a>    <span class="ex">-h</span> <span class="at">--help</span>       Print this help file and exit</span>
<span id="cb11-27"><a href="#cb11-27" tabindex="-1"></a></span>
<span id="cb11-28"><a href="#cb11-28" tabindex="-1"></a>    <span class="ex">-v</span> <span class="at">--version</span>    Print the version of the program and exit</span>
<span id="cb11-29"><a href="#cb11-29" tabindex="-1"></a></span>
<span id="cb11-30"><a href="#cb11-30" tabindex="-1"></a>    <span class="ex">-o</span> <span class="at">--outdir</span>     Create all output files in the specified output directory.</span>
<span id="cb11-31"><a href="#cb11-31" tabindex="-1"></a>                    <span class="ex">Please</span> note that this directory must exist as the program</span>
<span id="cb11-32"><a href="#cb11-32" tabindex="-1"></a>                    <span class="ex">will</span> not create it.  If this option is not set then the</span>
<span id="cb11-33"><a href="#cb11-33" tabindex="-1"></a>                    <span class="ex">output</span> file for each sequence file is created in the same</span>
<span id="cb11-34"><a href="#cb11-34" tabindex="-1"></a>                    <span class="ex">directory</span> as the sequence file which was processed.</span>
<span id="cb11-35"><a href="#cb11-35" tabindex="-1"></a></span>
<span id="cb11-36"><a href="#cb11-36" tabindex="-1"></a>    <span class="ex">--casava</span>        Files come from raw casava output. Files in the same sample</span>
<span id="cb11-37"><a href="#cb11-37" tabindex="-1"></a>                    <span class="ex">group</span> <span class="er">(</span><span class="ex">differing</span> only by the group number<span class="kw">)</span> <span class="ex">will</span> be analysed</span>
<span id="cb11-38"><a href="#cb11-38" tabindex="-1"></a>                    <span class="fu">as</span> a set rather than individually. Sequences with the filter</span>
<span id="cb11-39"><a href="#cb11-39" tabindex="-1"></a>                    <span class="ex">flag</span> set in the header will be excluded from the analysis.</span>
<span id="cb11-40"><a href="#cb11-40" tabindex="-1"></a>                    <span class="ex">Files</span> must have the same names given to them by casava</span>
<span id="cb11-41"><a href="#cb11-41" tabindex="-1"></a>                    <span class="kw">(</span><span class="ex">including</span> being gzipped and ending with .gz<span class="kw">)</span> <span class="ex">otherwise</span> they</span>
<span id="cb11-42"><a href="#cb11-42" tabindex="-1"></a>                    <span class="ex">will</span> not be grouped together correctly.</span>
<span id="cb11-43"><a href="#cb11-43" tabindex="-1"></a></span>
<span id="cb11-44"><a href="#cb11-44" tabindex="-1"></a>    <span class="ex">--nano</span>          Files come from naopore sequences and are in fast5 format. In</span>
<span id="cb11-45"><a href="#cb11-45" tabindex="-1"></a>                    <span class="ex">this</span> mode you can pass in directories to process and the program</span>
<span id="cb11-46"><a href="#cb11-46" tabindex="-1"></a>                    <span class="ex">will</span> take in all fast5 files within those directories and produce</span>
<span id="cb11-47"><a href="#cb11-47" tabindex="-1"></a>                    <span class="ex">a</span> single output file from the sequences found in all files.</span>
<span id="cb11-48"><a href="#cb11-48" tabindex="-1"></a></span>
<span id="cb11-49"><a href="#cb11-49" tabindex="-1"></a>    <span class="ex">--nofilter</span>      If running with <span class="at">--casava</span> then don<span class="st">'t remove read flagged by</span></span>
<span id="cb11-50"><a href="#cb11-50" tabindex="-1"></a><span class="st">                    casava as poor quality when performing the QC analysis.</span></span>
<span id="cb11-51"><a href="#cb11-51" tabindex="-1"></a></span>
<span id="cb11-52"><a href="#cb11-52" tabindex="-1"></a><span class="st">    --extract       If set then the zipped output file will be uncompressed in</span></span>
<span id="cb11-53"><a href="#cb11-53" tabindex="-1"></a><span class="st">                    the same directory after it has been created.  By default</span></span>
<span id="cb11-54"><a href="#cb11-54" tabindex="-1"></a><span class="st">                    this option will be set if fastqc is run in non-interactive</span></span>
<span id="cb11-55"><a href="#cb11-55" tabindex="-1"></a><span class="st">                    mode.</span></span>
<span id="cb11-56"><a href="#cb11-56" tabindex="-1"></a></span>
<span id="cb11-57"><a href="#cb11-57" tabindex="-1"></a><span class="st">    -j --java       Provides the full path to the java binary you want to use to</span></span>
<span id="cb11-58"><a href="#cb11-58" tabindex="-1"></a><span class="st">                    launch fastqc. If not supplied then java is assumed to be in</span></span>
<span id="cb11-59"><a href="#cb11-59" tabindex="-1"></a><span class="st">                    your path.</span></span>
<span id="cb11-60"><a href="#cb11-60" tabindex="-1"></a></span>
<span id="cb11-61"><a href="#cb11-61" tabindex="-1"></a><span class="st">    --noextract     Do not uncompress the output file after creating it.  You</span></span>
<span id="cb11-62"><a href="#cb11-62" tabindex="-1"></a><span class="st">                    should set this option if you do not wish to uncompress</span></span>
<span id="cb11-63"><a href="#cb11-63" tabindex="-1"></a><span class="st">                    the output when running in non-interactive mode.</span></span>
<span id="cb11-64"><a href="#cb11-64" tabindex="-1"></a></span>
<span id="cb11-65"><a href="#cb11-65" tabindex="-1"></a><span class="st">    --nogroup       Disable grouping of bases for reads &gt;50bp. All reports will</span></span>
<span id="cb11-66"><a href="#cb11-66" tabindex="-1"></a><span class="st">                    show data for every base in the read.  WARNING: Using this</span></span>
<span id="cb11-67"><a href="#cb11-67" tabindex="-1"></a><span class="st">                    option will cause fastqc to crash and burn if you use it on</span></span>
<span id="cb11-68"><a href="#cb11-68" tabindex="-1"></a><span class="st">                    really long reads, and your plots may end up a ridiculous size.</span></span>
<span id="cb11-69"><a href="#cb11-69" tabindex="-1"></a><span class="st">                    You have been warned!</span></span>
<span id="cb11-70"><a href="#cb11-70" tabindex="-1"></a></span>
<span id="cb11-71"><a href="#cb11-71" tabindex="-1"></a><span class="st">    -f --format     Bypasses the normal sequence file format detection and</span></span>
<span id="cb11-72"><a href="#cb11-72" tabindex="-1"></a><span class="st">                    forces the program to use the specified format.  Valid</span></span>
<span id="cb11-73"><a href="#cb11-73" tabindex="-1"></a><span class="st">                    formats are bam,sam,bam_mapped,sam_mapped and fastq</span></span>
<span id="cb11-74"><a href="#cb11-74" tabindex="-1"></a></span>
<span id="cb11-75"><a href="#cb11-75" tabindex="-1"></a><span class="st">    -t --threads    Specifies the number of files which can be processed</span></span>
<span id="cb11-76"><a href="#cb11-76" tabindex="-1"></a><span class="st">                    simultaneously.  Each thread will be allocated 250MB of</span></span>
<span id="cb11-77"><a href="#cb11-77" tabindex="-1"></a><span class="st">                    memory so you shouldn'</span>t run more threads than your</span>
<span id="cb11-78"><a href="#cb11-78" tabindex="-1"></a>                    <span class="ex">available</span> memory will cope with, and not more than</span>
<span id="cb11-79"><a href="#cb11-79" tabindex="-1"></a>                    <span class="ex">6</span> threads on a 32 bit machine</span>
<span id="cb11-80"><a href="#cb11-80" tabindex="-1"></a></span>
<span id="cb11-81"><a href="#cb11-81" tabindex="-1"></a>    <span class="ex">-c</span>              Specifies a non-default file which contains the list of</span>
<span id="cb11-82"><a href="#cb11-82" tabindex="-1"></a>    <span class="ex">--contaminants</span>  contaminants to screen overrepresented sequences against.</span>
<span id="cb11-83"><a href="#cb11-83" tabindex="-1"></a>                    <span class="ex">The</span> file must contain sets of named contaminants in the</span>
<span id="cb11-84"><a href="#cb11-84" tabindex="-1"></a>                    <span class="ex">form</span> name<span class="pp">[</span><span class="ss">tab</span><span class="pp">]</span>sequence.  Lines prefixed with a hash will</span>
<span id="cb11-85"><a href="#cb11-85" tabindex="-1"></a>                    <span class="ex">be</span> ignored.</span>
<span id="cb11-86"><a href="#cb11-86" tabindex="-1"></a></span>
<span id="cb11-87"><a href="#cb11-87" tabindex="-1"></a>    <span class="ex">-a</span>              Specifies a non-default file which contains the list of</span>
<span id="cb11-88"><a href="#cb11-88" tabindex="-1"></a>    <span class="ex">--adapters</span>      adapter sequences which will be explicity searched against</span>
<span id="cb11-89"><a href="#cb11-89" tabindex="-1"></a>                    <span class="ex">the</span> library. The file must contain sets of named adapters</span>
<span id="cb11-90"><a href="#cb11-90" tabindex="-1"></a>                    <span class="er">in</span> <span class="ex">the</span> form name<span class="pp">[</span><span class="ss">tab</span><span class="pp">]</span>sequence.  Lines prefixed with a hash</span>
<span id="cb11-91"><a href="#cb11-91" tabindex="-1"></a>                    <span class="ex">will</span> be ignored.</span>
<span id="cb11-92"><a href="#cb11-92" tabindex="-1"></a></span>
<span id="cb11-93"><a href="#cb11-93" tabindex="-1"></a>    <span class="ex">-l</span>              Specifies a non-default file which contains a set of criteria</span>
<span id="cb11-94"><a href="#cb11-94" tabindex="-1"></a>    <span class="ex">--limits</span>        which will be used to determine the warn/error limits for the</span>
<span id="cb11-95"><a href="#cb11-95" tabindex="-1"></a>                    <span class="ex">various</span> modules.  This file can also be used to selectively</span>
<span id="cb11-96"><a href="#cb11-96" tabindex="-1"></a>                    <span class="ex">remove</span> some modules from the output all together.  The format</span>
<span id="cb11-97"><a href="#cb11-97" tabindex="-1"></a>                    <span class="ex">needs</span> to mirror the default limits.txt file found in the</span>
<span id="cb11-98"><a href="#cb11-98" tabindex="-1"></a>                    <span class="ex">Configuration</span> folder.</span>
<span id="cb11-99"><a href="#cb11-99" tabindex="-1"></a></span>
<span id="cb11-100"><a href="#cb11-100" tabindex="-1"></a>   <span class="ex">-k</span> <span class="at">--kmers</span>       Specifies the length of Kmer to look for in the Kmer content</span>
<span id="cb11-101"><a href="#cb11-101" tabindex="-1"></a>                    <span class="ex">module.</span> Specified Kmer length must be between 2 and 10. Default</span>
<span id="cb11-102"><a href="#cb11-102" tabindex="-1"></a>                    <span class="ex">length</span> is 7 if not specified.</span>
<span id="cb11-103"><a href="#cb11-103" tabindex="-1"></a></span>
<span id="cb11-104"><a href="#cb11-104" tabindex="-1"></a>   <span class="ex">-q</span> <span class="at">--quiet</span>       Supress all progress messages on stdout and only report errors.</span>
<span id="cb11-105"><a href="#cb11-105" tabindex="-1"></a></span>
<span id="cb11-106"><a href="#cb11-106" tabindex="-1"></a>   <span class="ex">-d</span> <span class="at">--dir</span>         Selects a directory to be used for temporary files written when</span>
<span id="cb11-107"><a href="#cb11-107" tabindex="-1"></a>                    <span class="ex">generating</span> report images. Defaults to system temp directory if</span>
<span id="cb11-108"><a href="#cb11-108" tabindex="-1"></a>                    <span class="ex">not</span> specified.</span>
<span id="cb11-109"><a href="#cb11-109" tabindex="-1"></a></span>
<span id="cb11-110"><a href="#cb11-110" tabindex="-1"></a><span class="ex">BUGS</span></span>
<span id="cb11-111"><a href="#cb11-111" tabindex="-1"></a></span>
<span id="cb11-112"><a href="#cb11-112" tabindex="-1"></a>    <span class="ex">Any</span> bugs in fastqc should be reported either to simon.andrews@babraham.ac.uk</span>
<span id="cb11-113"><a href="#cb11-113" tabindex="-1"></a>    <span class="ex">or</span> in www.bioinformatics.babraham.ac.uk/bugzilla/</span></code></pre>
</div>
<p>if fastqc is not installed then you would expect to see an error
like</p>
<pre><code>$ fastqc -h
The program 'fastqc' is currently not installed. You can install it by typing:
sudo apt-get install fastqc</code></pre>
<p>If this happens check with your instructor before trying to install
it.</p>
</div>
<div class="section level3">
<h3 id="assessing-quality-using-fastqc">Assessing quality using FastQC<a class="anchor" aria-label="anchor" href="#assessing-quality-using-fastqc"></a>
</h3>
<p>In real life, you will not be assessing the quality of your reads by
visually inspecting your FASTQ files. Rather, you will be using a
software program to assess read quality and filter out poor quality
reads. We will first use a program called <a href="https://www.bioinformatics.babraham.ac.uk/projects/fastqc/" class="external-link">FastQC</a>
to visualize the quality of our reads. Later in our workflow, we will
use another program to filter out poor quality reads.</p>
<p>FastQC has a number of features which can give you a quick impression
of any problems your data may have, so you can take these issues into
consideration before moving forward with your analyses. Rather than
looking at quality scores for each individual read, FastQC looks at
quality collectively across all reads within a sample. The image below
shows one FastQC-generated plot that indicates a very high quality
sample:</p>
<figure><img src="../fig/good_quality1.8.png" alt="good_quality" class="figure mx-auto d-block"></figure><p>The x-axis displays the base position in the read, and the y-axis
shows quality scores. In this example, the sample contains reads that
are 40 bp long. This is much shorter than the reads we are working with
in our workflow. For each position, there is a box-and-whisker plot
showing the distribution of quality scores for all reads at that
position. The horizontal red line indicates the median quality score and
the yellow box shows the 1st to 3rd quartile range. This means that 50%
of reads have a quality score that falls within the range of the yellow
box at that position. The whiskers show the absolute range, which covers
the lowest (0th quartile) to highest (4th quartile) values.</p>
<p>For each position in this sample, the quality values do not drop much
lower than 32. This is a high quality score. The plot background is also
color-coded to identify good (green), acceptable (yellow), and bad (red)
quality scores.</p>
<p>Now let’s take a look at a quality plot on the other end of the
spectrum.</p>
<figure><img src="../fig/bad_quality1.8.png" alt="bad_quality" class="figure mx-auto d-block"></figure><p>Here, we see positions within the read in which the boxes span a much
wider range. Also, quality scores drop quite low into the “bad” range,
particularly on the tail end of the reads. The FastQC tool produces
several other diagnostic plots to assess sample quality, in addition to
the one plotted above.</p>
</div>
<div class="section level3">
<h3 id="running-fastqc">Running FastQC<a class="anchor" aria-label="anchor" href="#running-fastqc"></a>
</h3>
<p>We will now assess the quality of the reads that we downloaded.
First, make sure you are still in the <code>untrimmed_fastq</code>
directory</p>
<div class="codewrapper sourceCode" id="cb13">
<h3 class="code-label">BASH<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="sourceCode bash" tabindex="0"><code class="sourceCode bash"><span id="cb13-1"><a href="#cb13-1" tabindex="-1"></a><span class="ex">$</span> cd ~/dc_workshop/data/untrimmed_fastq/</span></code></pre>
</div>
<div id="exercise-1" class="callout challenge">
<div class="callout-square">
<i class="callout-icon" data-feather="zap"></i>
</div>
<div id="exercise-1" class="callout-inner">
<h3 class="callout-title">Exercise</h3>
<div class="callout-content">
<p>How big are the files? (Hint: Look at the options for the
<code>ls</code> command to see how to show file sizes.)</p>
</div>
</div>
</div>
<div id="accordionSolution2" class="accordion challenge-accordion accordion-flush">
<div class="accordion-item">
<button class="accordion-button solution-button collapsed" type="button" data-bs-toggle="collapse" data-bs-target="#collapseSolution2" aria-expanded="false" aria-controls="collapseSolution2">
  <h4 class="accordion-header" id="headingSolution2"> Show me the solution </h4>
</button>
<div id="collapseSolution2" class="accordion-collapse collapse" aria-labelledby="headingSolution2" data-bs-parent="#accordionSolution2">
<div class="accordion-body">
<div class="codewrapper sourceCode" id="cb14">
<h3 class="code-label">BASH<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="sourceCode bash" tabindex="0"><code class="sourceCode bash"><span id="cb14-1"><a href="#cb14-1" tabindex="-1"></a><span class="ex">$</span> ls <span class="at">-l</span> <span class="at">-h</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>-rw-rw-r-- 1 dcuser dcuser 545M Jul  6 20:27 SRR2584863_1.fastq
-rw-rw-r-- 1 dcuser dcuser 183M Jul  6 20:29 SRR2584863_2.fastq.gz
-rw-rw-r-- 1 dcuser dcuser 309M Jul  6 20:34 SRR2584866_1.fastq.gz
-rw-rw-r-- 1 dcuser dcuser 296M Jul  6 20:37 SRR2584866_2.fastq.gz
-rw-rw-r-- 1 dcuser dcuser 124M Jul  6 20:22 SRR2589044_1.fastq.gz
-rw-rw-r-- 1 dcuser dcuser 128M Jul  6 20:24 SRR2589044_2.fastq.gz</code></pre>
</div>
<p>There are six FASTQ files ranging from 124M (124MB) to 545M.</p>
</div>
</div>
</div>
</div>
<p>FastQC can accept multiple file names as input, and on both zipped
and unzipped files, so we can use the *.fastq* wildcard to run FastQC on
all of the FASTQ files in this directory.</p>
<div class="codewrapper sourceCode" id="cb16">
<h3 class="code-label">BASH<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="sourceCode bash" tabindex="0"><code class="sourceCode bash"><span id="cb16-1"><a href="#cb16-1" tabindex="-1"></a><span class="ex">$</span> fastqc <span class="pp">*</span>.fastq<span class="pp">*</span></span></code></pre>
</div>
<p>You will see an automatically updating output message telling you the
progress of the analysis. It will start like this:</p>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>Started analysis of SRR2584863_1.fastq
Approx 5% complete for SRR2584863_1.fastq
Approx 10% complete for SRR2584863_1.fastq
Approx 15% complete for SRR2584863_1.fastq
Approx 20% complete for SRR2584863_1.fastq
Approx 25% complete for SRR2584863_1.fastq
Approx 30% complete for SRR2584863_1.fastq
Approx 35% complete for SRR2584863_1.fastq
Approx 40% complete for SRR2584863_1.fastq
Approx 45% complete for SRR2584863_1.fastq</code></pre>
</div>
<p>In total, it should take about five minutes for FastQC to run on all
six of our FASTQ files. When the analysis completes, your prompt will
return. So your screen will look something like this:</p>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>Approx 80% complete for SRR2589044_2.fastq.gz
Approx 85% complete for SRR2589044_2.fastq.gz
Approx 90% complete for SRR2589044_2.fastq.gz
Approx 95% complete for SRR2589044_2.fastq.gz
Analysis complete for SRR2589044_2.fastq.gz
$</code></pre>
</div>
<p>The FastQC program has created several new files within our
<code>data/untrimmed_fastq/</code> directory.</p>
<div class="codewrapper sourceCode" id="cb19">
<h3 class="code-label">BASH<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="sourceCode bash" tabindex="0"><code class="sourceCode bash"><span id="cb19-1"><a href="#cb19-1" tabindex="-1"></a><span class="ex">$</span> ls</span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>SRR2584863_1.fastq        SRR2584866_1_fastqc.html  SRR2589044_1_fastqc.html
SRR2584863_1_fastqc.html  SRR2584866_1_fastqc.zip   SRR2589044_1_fastqc.zip
SRR2584863_1_fastqc.zip   SRR2584866_1.fastq.gz     SRR2589044_1.fastq.gz
SRR2584863_2_fastqc.html  SRR2584866_2_fastqc.html  SRR2589044_2_fastqc.html
SRR2584863_2_fastqc.zip   SRR2584866_2_fastqc.zip   SRR2589044_2_fastqc.zip
SRR2584863_2.fastq.gz     SRR2584866_2.fastq.gz     SRR2589044_2.fastq.gz</code></pre>
</div>
<p>For each input FASTQ file, FastQC has created a <code>.zip</code>
file and a</p>
<p><code>.html</code> file. The <code>.zip</code> file extension
indicates that this is actually a compressed set of multiple output
files. We will be working with these output files soon. The
<code>.html</code> file is a stable webpage displaying the summary
report for each of our samples.</p>
<p>We want to keep our data files and our results files separate, so we
will move these output files into a new directory within our
<code>results/</code> directory.</p>
<div class="codewrapper sourceCode" id="cb21">
<h3 class="code-label">BASH<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="sourceCode bash" tabindex="0"><code class="sourceCode bash"><span id="cb21-1"><a href="#cb21-1" tabindex="-1"></a><span class="ex">$</span> mkdir <span class="at">-p</span> ~/dc_workshop/results/fastqc_untrimmed_reads</span>
<span id="cb21-2"><a href="#cb21-2" tabindex="-1"></a><span class="ex">$</span> mv <span class="pp">*</span>.zip ~/dc_workshop/results/fastqc_untrimmed_reads/</span>
<span id="cb21-3"><a href="#cb21-3" tabindex="-1"></a><span class="ex">$</span> mv <span class="pp">*</span>.html ~/dc_workshop/results/fastqc_untrimmed_reads/</span></code></pre>
</div>
<p>Now we can navigate into this results directory and do some closer
inspection of our output files.</p>
<div class="codewrapper sourceCode" id="cb22">
<h3 class="code-label">BASH<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="sourceCode bash" tabindex="0"><code class="sourceCode bash"><span id="cb22-1"><a href="#cb22-1" tabindex="-1"></a><span class="ex">$</span> cd ~/dc_workshop/results/fastqc_untrimmed_reads/</span></code></pre>
</div>
</div>
<div class="section level3">
<h3 id="viewing-the-fastqc-results">Viewing the FastQC results<a class="anchor" aria-label="anchor" href="#viewing-the-fastqc-results"></a>
</h3>
<p>If we were working on our local computers, we would be able to look
at each of these HTML files by opening them in a web browser.</p>
<p>However, these files are currently sitting on our remote AWS
instance, where our local computer can not see them. And, since we are
only logging into the AWS instance via the command line - it does not
have any web browser setup to display these files either.</p>
<p>So the easiest way to look at these webpage summary reports will be
to transfer them to our local computers (i.e. your laptop).</p>
<p>To transfer a file from a remote server to our own machines, we will
use <code>scp</code>, which we learned yesterday in the Shell Genomics
lesson.</p>
<p>First we will make a new directory on our computer to store the HTML
files we are transferring. Let’s put it on our desktop for now. Open a
new tab in your terminal program (you can use the pull down menu at the
top of your screen or the Cmd+t keyboard shortcut) and type:</p>
<div class="codewrapper sourceCode" id="cb23">
<h3 class="code-label">BASH<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="sourceCode bash" tabindex="0"><code class="sourceCode bash"><span id="cb23-1"><a href="#cb23-1" tabindex="-1"></a><span class="ex">$</span> mkdir <span class="at">-p</span> ~/Desktop/fastqc_html</span></code></pre>
</div>
<p>Now we can transfer our HTML files to our local computer using
<code>scp</code>.</p>
<div class="codewrapper sourceCode" id="cb24">
<h3 class="code-label">BASH<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="sourceCode bash" tabindex="0"><code class="sourceCode bash"><span id="cb24-1"><a href="#cb24-1" tabindex="-1"></a><span class="ex">$</span> scp dcuser@ec2-34-238-162-94.compute-1.amazonaws.com:~/dc_workshop/results/fastqc_untrimmed_reads/<span class="pp">*</span>.html ~/Desktop/fastqc_html</span></code></pre>
</div>
<div id="note-on-using-zsh" class="callout">
<div class="callout-square">
<i class="callout-icon" data-feather="bell"></i>
</div>
<div id="note-on-using-zsh" class="callout-inner">
<h3 class="callout-title">Note on using zsh</h3>
<div class="callout-content">
<p>If you are using zsh instead of bash (macOS for example changed the
default recently to zsh), it is likely that a
<code>no matches found</code> error will be displayed. The reason for
this is that the wildcard (“*”) is not correctly interpreted. To fix
this problem the wildcard needs to be escaped with a “\”:</p>
<div class="codewrapper sourceCode" id="cb25">
<h3 class="code-label">BASH<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="sourceCode bash" tabindex="0"><code class="sourceCode bash"><span id="cb25-1"><a href="#cb25-1" tabindex="-1"></a><span class="ex">$</span> scp dcuser@ec2-34-238-162-94.compute-1.amazonaws.com:~/dc_workshop/results/fastqc_untrimmed_reads/\*.html ~/Desktop/fastqc_html</span></code></pre>
</div>
<p>Alternatively, you can put the whole path into quotation marks:</p>
<div class="codewrapper sourceCode" id="cb26">
<h3 class="code-label">BASH<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="sourceCode bash" tabindex="0"><code class="sourceCode bash"><span id="cb26-1"><a href="#cb26-1" tabindex="-1"></a><span class="ex">$</span> scp <span class="st">"dcuser@ec2-34-238-162-94.compute-1.amazonaws.com:~/dc_workshop/results/fastqc_untrimmed_reads/*.html"</span> ~/Desktop/fastqc_html</span></code></pre>
</div>
</div>
</div>
</div>
<p>As a reminder, the first part of the command
<code>dcuser@ec2-34-238-162-94.compute-1.amazonaws.com</code> is the
address for your remote computer. Make sure you replace everything after
<code>dcuser@</code> with your instance number (the one you used to log
in).</p>
<p>The second part starts with a <code>:</code> and then gives the
absolute path of the files you want to transfer from your remote
computer. Do not forget the <code>:</code>. We used a wildcard
(<code>*.html</code>) to indicate that we want all of the HTML
files.</p>
<p>The third part of the command gives the absolute path of the location
you want to put the files. This is on your local computer and is the
directory we just created <code>~/Desktop/fastqc_html</code>.</p>
<p>You should see a status output like this:</p>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>SRR2584863_1_fastqc.html                      100%  249KB 152.3KB/s   00:01
SRR2584863_2_fastqc.html                      100%  254KB 219.8KB/s   00:01
SRR2584866_1_fastqc.html                      100%  254KB 271.8KB/s   00:00
SRR2584866_2_fastqc.html                      100%  251KB 252.8KB/s   00:00
SRR2589044_1_fastqc.html                      100%  249KB 370.1KB/s   00:00
SRR2589044_2_fastqc.html                      100%  251KB 592.2KB/s   00:00</code></pre>
</div>
<p>Now we can go to our new directory and open the 6 HTML files.</p>
<p>Depending on your system, you should be able to select and open them
all at once via a right click menu in your file browser.</p>
<div id="exercise-2" class="callout challenge">
<div class="callout-square">
<i class="callout-icon" data-feather="zap"></i>
</div>
<div id="exercise-2" class="callout-inner">
<h3 class="callout-title">Exercise</h3>
<div class="callout-content">
<p>Discuss your results with a neighbor. Which sample(s) looks the best
in terms of per base sequence quality? Which sample(s) look the
worst?</p>
</div>
</div>
</div>
<div id="accordionSolution3" class="accordion challenge-accordion accordion-flush">
<div class="accordion-item">
<button class="accordion-button solution-button collapsed" type="button" data-bs-toggle="collapse" data-bs-target="#collapseSolution3" aria-expanded="false" aria-controls="collapseSolution3">
  <h4 class="accordion-header" id="headingSolution3"> Show me the solution </h4>
</button>
<div id="collapseSolution3" class="accordion-collapse collapse" aria-labelledby="headingSolution3" data-bs-parent="#accordionSolution3">
<div class="accordion-body">
<p>All of the reads contain usable data, but the quality decreases
toward the end of the reads.</p>
</div>
</div>
</div>
</div>
</div>
<div class="section level3">
<h3 id="decoding-the-other-fastqc-outputs">Decoding the other FastQC outputs<a class="anchor" aria-label="anchor" href="#decoding-the-other-fastqc-outputs"></a>
</h3>
<p>We have now looked at quite a few “Per base sequence quality” FastQC
graphs, but there are nine other graphs that we have not talked about!
Below we have provided a brief overview of interpretations for each of
these plots. For more information, please see the FastQC documentation
<a href="https://www.bioinformatics.babraham.ac.uk/projects/fastqc/Help/" class="external-link">here</a></p>
<ul>
<li>
<a href="https://www.bioinformatics.babraham.ac.uk/projects/fastqc/Help/3%20Analysis%20Modules/12%20Per%20Tile%20Sequence%20Quality.html" class="external-link"><strong>Per
tile sequence quality</strong></a>: the machines that perform sequencing
are divided into tiles. This plot displays patterns in base quality
along these tiles. Consistently low scores are often found around the
edges, but hot spots can also occur in the middle if an air bubble was
introduced at some point during the run.</li>
<li>
<a href="https://www.bioinformatics.babraham.ac.uk/projects/fastqc/Help/3%20Analysis%20Modules/3%20Per%20Sequence%20Quality%20Scores.html" class="external-link"><strong>Per
sequence quality scores</strong></a>: a density plot of quality for all
reads at all positions. This plot shows what quality scores are most
common.</li>
<li>
<a href="https://www.bioinformatics.babraham.ac.uk/projects/fastqc/Help/3%20Analysis%20Modules/4%20Per%20Base%20Sequence%20Content.html" class="external-link"><strong>Per
base sequence content</strong></a>: plots the proportion of each base
position over all of the reads. Typically, we expect to see each base
roughly 25% of the time at each position, but this often fails at the
beginning or end of the read due to quality or adapter content.</li>
<li>
<a href="https://www.bioinformatics.babraham.ac.uk/projects/fastqc/Help/3%20Analysis%20Modules/5%20Per%20Sequence%20GC%20Content.html" class="external-link"><strong>Per
sequence GC content</strong></a>: a density plot of average GC content
in each of the reads.</li>
<li>
<a href="https://www.bioinformatics.babraham.ac.uk/projects/fastqc/Help/3%20Analysis%20Modules/6%20Per%20Base%20N%20Content.html" class="external-link"><strong>Per
base N content</strong></a>: the percent of times that ‘N’ occurs at a
position in all reads. If there is an increase at a particular position,
this might indicate that something went wrong during sequencing.</li>
<li>
<a href="https://www.bioinformatics.babraham.ac.uk/projects/fastqc/Help/3%20Analysis%20Modules/7%20Sequence%20Length%20Distribution.html" class="external-link"><strong>Sequence
Length Distribution</strong></a>: the distribution of sequence lengths
of all reads in the file. If the data is raw, there is often on sharp
peak, however if the reads have been trimmed, there may be a
distribution of shorter lengths.</li>
<li>
<a href="https://www.bioinformatics.babraham.ac.uk/projects/fastqc/Help/3%20Analysis%20Modules/8%20Duplicate%20Sequences.html" class="external-link"><strong>Sequence
Duplication Levels</strong></a>: A distribution of duplicated sequences.
In sequencing, we expect most reads to only occur once. If some
sequences are occurring more than once, it might indicate enrichment
bias (e.g. from PCR). If the samples are high coverage (or RNA-seq or
amplicon), this might not be true.</li>
<li>
<a href="https://www.bioinformatics.babraham.ac.uk/projects/fastqc/Help/3%20Analysis%20Modules/9%20Overrepresented%20Sequences.html" class="external-link"><strong>Overrepresented
sequences</strong></a>: A list of sequences that occur more frequently
than would be expected by chance.</li>
<li>
<a href="https://www.bioinformatics.babraham.ac.uk/projects/fastqc/Help/3%20Analysis%20Modules/10%20Adapter%20Content.html" class="external-link"><strong>Adapter
Content</strong></a>: a graph indicating where adapater sequences occur
in the reads.</li>
<li>
<a href="https://www.bioinformatics.babraham.ac.uk/projects/fastqc/Help/3%20Analysis%20Modules/11%20Kmer%20Content.html" class="external-link"><strong>K-mer
Content</strong></a>: a graph showing any sequences which may show a
positional bias within the reads.</li>
</ul>
</div>
<div class="section level3">
<h3 id="working-with-the-fastqc-text-output">Working with the FastQC text output<a class="anchor" aria-label="anchor" href="#working-with-the-fastqc-text-output"></a>
</h3>
<p>Now that we have looked at our HTML reports to get a feel for the
data, let’s look more closely at the other output files. Go back to the
tab in your terminal program that is connected to your AWS instance (the
tab label will start with <code>dcuser@ip</code>) and make sure you are
in our results subdirectory.</p>
<div class="codewrapper sourceCode" id="cb28">
<h3 class="code-label">BASH<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="sourceCode bash" tabindex="0"><code class="sourceCode bash"><span id="cb28-1"><a href="#cb28-1" tabindex="-1"></a><span class="ex">$</span> cd ~/dc_workshop/results/fastqc_untrimmed_reads/</span>
<span id="cb28-2"><a href="#cb28-2" tabindex="-1"></a><span class="ex">$</span> ls</span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>SRR2584863_1_fastqc.html  SRR2584866_1_fastqc.html  SRR2589044_1_fastqc.html
SRR2584863_1_fastqc.zip   SRR2584866_1_fastqc.zip   SRR2589044_1_fastqc.zip
SRR2584863_2_fastqc.html  SRR2584866_2_fastqc.html  SRR2589044_2_fastqc.html
SRR2584863_2_fastqc.zip   SRR2584866_2_fastqc.zip   SRR2589044_2_fastqc.zip</code></pre>
</div>
<p>Our <code>.zip</code> files are compressed files. They each contain
multiple different types of output files for a single input FASTQ file.
To view the contents of a <code>.zip</code> file, we can use the program
<code>unzip</code> to decompress these files. Let’s try doing them all
at once using a wildcard.</p>
<div class="codewrapper sourceCode" id="cb30">
<h3 class="code-label">BASH<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="sourceCode bash" tabindex="0"><code class="sourceCode bash"><span id="cb30-1"><a href="#cb30-1" tabindex="-1"></a><span class="ex">$</span> unzip <span class="pp">*</span>.zip</span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>Archive:  SRR2584863_1_fastqc.zip
caution: filename not matched:  SRR2584863_2_fastqc.zip
caution: filename not matched:  SRR2584866_1_fastqc.zip
caution: filename not matched:  SRR2584866_2_fastqc.zip
caution: filename not matched:  SRR2589044_1_fastqc.zip
caution: filename not matched:  SRR2589044_2_fastqc.zip</code></pre>
</div>
<p>This did not work. We unzipped the first file and then got a warning
message for each of the other <code>.zip</code> files. This is because
<code>unzip</code> expects to get only one zip file as input. We could
go through and unzip each file one at a time, but this is very time
consuming and error-prone. Someday you may have 500 files to unzip!</p>
<p>A more efficient way is to use a <code>for</code> loop like we
learned in the Shell Genomics lesson to iterate through all of our
<code>.zip</code> files. Let’s see what that looks like and then we will
discuss what we are doing with each line of our loop.</p>
<div class="codewrapper sourceCode" id="cb32">
<h3 class="code-label">BASH<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="sourceCode bash" tabindex="0"><code class="sourceCode bash"><span id="cb32-1"><a href="#cb32-1" tabindex="-1"></a><span class="ex">$</span> for filename in <span class="pp">*</span>.zip</span>
<span id="cb32-2"><a href="#cb32-2" tabindex="-1"></a><span class="op">&gt;</span> do</span>
<span id="cb32-3"><a href="#cb32-3" tabindex="-1"></a><span class="op">&gt;</span> unzip <span class="va">$filename</span></span>
<span id="cb32-4"><a href="#cb32-4" tabindex="-1"></a><span class="op">&gt;</span> done</span></code></pre>
</div>
<p>In this example, the input is six filenames (one filename for each of
our <code>.zip</code> files). Each time the loop iterates, it will
assign a file name to the variable <code>filename</code> and run the
<code>unzip</code> command. The first time through the loop,
<code>$filename</code> is <code>SRR2584863_1_fastqc.zip</code>. The
interpreter runs the command <code>unzip</code> on
<code>SRR2584863_1_fastqc.zip</code>. For the second iteration,
<code>$filename</code> becomes <code>SRR2584863_2_fastqc.zip</code>.
This time, the shell runs <code>unzip</code> on
<code>SRR2584863_2_fastqc.zip</code>. It then repeats this process for
the four other <code>.zip</code> files in our directory.</p>
<p>When we run our <code>for</code> loop, you will see output that
starts like this:</p>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code><span><span class="va">Archive</span><span class="op">:</span>  <span class="va">SRR2589044_2_fastqc.zip</span></span>
<span>   <span class="va">creating</span><span class="op">:</span> <span class="va">SRR2589044_2_fastqc</span><span class="op">/</span></span>
<span>   <span class="va">creating</span><span class="op">:</span> <span class="va">SRR2589044_2_fastqc</span><span class="op">/</span><span class="va">Icons</span><span class="op">/</span></span>
<span>   <span class="va">creating</span><span class="op">:</span> <span class="va">SRR2589044_2_fastqc</span><span class="op">/</span><span class="va">Images</span><span class="op">/</span></span>
<span>  <span class="va">inflating</span><span class="op">:</span> <span class="va">SRR2589044_2_fastqc</span><span class="op">/</span><span class="va">Icons</span><span class="op">/</span><span class="va">fastqc_icon.png</span></span>
<span>  <span class="va">inflating</span><span class="op">:</span> <span class="va">SRR2589044_2_fastqc</span><span class="op">/</span><span class="va">Icons</span><span class="op">/</span><span class="va">warning.png</span></span>
<span>  <span class="va">inflating</span><span class="op">:</span> <span class="va">SRR2589044_2_fastqc</span><span class="op">/</span><span class="va">Icons</span><span class="op">/</span><span class="va">error.png</span></span>
<span>  <span class="va">inflating</span><span class="op">:</span> <span class="va">SRR2589044_2_fastqc</span><span class="op">/</span><span class="va">Icons</span><span class="op">/</span><span class="va">tick.png</span></span>
<span>  <span class="va">inflating</span><span class="op">:</span> <span class="va">SRR2589044_2_fastqc</span><span class="op">/</span><span class="va">summary.txt</span></span>
<span>  <span class="va">inflating</span><span class="op">:</span> <span class="va">SRR2589044_2_fastqc</span><span class="op">/</span><span class="va">Images</span><span class="op">/</span><span class="va">per_base_quality.png</span></span>
<span>  <span class="va">inflating</span><span class="op">:</span> <span class="va">SRR2589044_2_fastqc</span><span class="op">/</span><span class="va">Images</span><span class="op">/</span><span class="va">per_tile_quality.png</span></span>
<span>  <span class="va">inflating</span><span class="op">:</span> <span class="va">SRR2589044_2_fastqc</span><span class="op">/</span><span class="va">Images</span><span class="op">/</span><span class="va">per_sequence_quality.png</span></span>
<span>  <span class="va">inflating</span><span class="op">:</span> <span class="va">SRR2589044_2_fastqc</span><span class="op">/</span><span class="va">Images</span><span class="op">/</span><span class="va">per_base_sequence_content.png</span></span>
<span>  <span class="va">inflating</span><span class="op">:</span> <span class="va">SRR2589044_2_fastqc</span><span class="op">/</span><span class="va">Images</span><span class="op">/</span><span class="va">per_sequence_gc_content.png</span></span>
<span>  <span class="va">inflating</span><span class="op">:</span> <span class="va">SRR2589044_2_fastqc</span><span class="op">/</span><span class="va">Images</span><span class="op">/</span><span class="va">per_base_n_content.png</span></span>
<span>  <span class="va">inflating</span><span class="op">:</span> <span class="va">SRR2589044_2_fastqc</span><span class="op">/</span><span class="va">Images</span><span class="op">/</span><span class="va">sequence_length_distribution.png</span></span>
<span>  <span class="va">inflating</span><span class="op">:</span> <span class="va">SRR2589044_2_fastqc</span><span class="op">/</span><span class="va">Images</span><span class="op">/</span><span class="va">duplication_levels.png</span></span>
<span>  <span class="va">inflating</span><span class="op">:</span> <span class="va">SRR2589044_2_fastqc</span><span class="op">/</span><span class="va">Images</span><span class="op">/</span><span class="va">adapter_content.png</span></span>
<span>  <span class="va">inflating</span><span class="op">:</span> <span class="va">SRR2589044_2_fastqc</span><span class="op">/</span><span class="va">fastqc_report.html</span></span>
<span>  <span class="va">inflating</span><span class="op">:</span> <span class="va">SRR2589044_2_fastqc</span><span class="op">/</span><span class="va">fastqc_data.txt</span></span>
<span>  <span class="va">inflating</span><span class="op">:</span> <span class="va">SRR2589044_2_fastqc</span><span class="op">/</span><span class="va">fastqc.fo</span></span></code></pre>
</div>
<p>The <code>unzip</code> program is decompressing the <code>.zip</code>
files and creating a new directory (with subdirectories) for each of our
samples, to store all of the different output that is produced by
FastQC. There</p>
<p>are a lot of files here. The one we are going to focus on is the
<code>summary.txt</code> file.</p>
<p>If you list the files in our directory now you will see:</p>
<pre><code>SRR2584863_1_fastqc       SRR2584866_1_fastqc       SRR2589044_1_fastqc
SRR2584863_1_fastqc.html  SRR2584866_1_fastqc.html  SRR2589044_1_fastqc.html
SRR2584863_1_fastqc.zip   SRR2584866_1_fastqc.zip   SRR2589044_1_fastqc.zip
SRR2584863_2_fastqc       SRR2584866_2_fastqc       SRR2589044_2_fastqc
SRR2584863_2_fastqc.html  SRR2584866_2_fastqc.html  SRR2589044_2_fastqc.html
SRR2584863_2_fastqc.zip   SRR2584866_2_fastqc.zip   SRR2589044_2_fastqc.zip</code></pre>
<p>{:. output}</p>
<p>The <code>.html</code> files and the uncompressed <code>.zip</code>
files are still present, but now we also have a new directory for each
of our samples. We can see for sure that it is a directory if we use the
<code>-F</code> flag for <code>ls</code>.</p>
<div class="codewrapper sourceCode" id="cb35">
<h3 class="code-label">BASH<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="sourceCode bash" tabindex="0"><code class="sourceCode bash"><span id="cb35-1"><a href="#cb35-1" tabindex="-1"></a><span class="ex">$</span> ls <span class="at">-F</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>SRR2584863_1_fastqc/      SRR2584866_1_fastqc/      SRR2589044_1_fastqc/
SRR2584863_1_fastqc.html  SRR2584866_1_fastqc.html  SRR2589044_1_fastqc.html
SRR2584863_1_fastqc.zip   SRR2584866_1_fastqc.zip   SRR2589044_1_fastqc.zip
SRR2584863_2_fastqc/      SRR2584866_2_fastqc/      SRR2589044_2_fastqc/
SRR2584863_2_fastqc.html  SRR2584866_2_fastqc.html  SRR2589044_2_fastqc.html
SRR2584863_2_fastqc.zip   SRR2584866_2_fastqc.zip   SRR2589044_2_fastqc.zip</code></pre>
</div>
<p>Let’s see what files are present within one of these output
directories.</p>
<div class="codewrapper sourceCode" id="cb37">
<h3 class="code-label">BASH<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="sourceCode bash" tabindex="0"><code class="sourceCode bash"><span id="cb37-1"><a href="#cb37-1" tabindex="-1"></a><span class="ex">$</span> ls <span class="at">-F</span> SRR2584863_1_fastqc/</span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>fastqc_data.txt  fastqc.fo  fastqc_report.html	Icons/	Images/  summary.txt</code></pre>
</div>
<p>Use <code>less</code> to preview the <code>summary.txt</code> file
for this sample.</p>
<div class="codewrapper sourceCode" id="cb39">
<h3 class="code-label">BASH<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="sourceCode bash" tabindex="0"><code class="sourceCode bash"><span id="cb39-1"><a href="#cb39-1" tabindex="-1"></a><span class="ex">$</span> less SRR2584863_1_fastqc/summary.txt</span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>PASS    Basic Statistics        SRR2584863_1.fastq
PASS    Per base sequence quality       SRR2584863_1.fastq
PASS    Per tile sequence quality       SRR2584863_1.fastq
PASS    Per sequence quality scores     SRR2584863_1.fastq
WARN    Per base sequence content       SRR2584863_1.fastq
WARN    Per sequence GC content SRR2584863_1.fastq
PASS    Per base N content      SRR2584863_1.fastq
PASS    Sequence Length Distribution    SRR2584863_1.fastq
PASS    Sequence Duplication Levels     SRR2584863_1.fastq
PASS    Overrepresented sequences       SRR2584863_1.fastq
WARN    Adapter Content SRR2584863_1.fastq</code></pre>
</div>
<p>The summary file gives us a list of tests that FastQC ran, and tells
us whether this sample passed, failed, or is borderline
(<code>WARN</code>). Remember, to quit from <code>less</code> you must
type <code>q</code>.</p>
</div>
<div class="section level3">
<h3 id="documenting-our-work">Documenting our work<a class="anchor" aria-label="anchor" href="#documenting-our-work"></a>
</h3>
<p>We can make a record of the results we obtained for all our
samples</p>
<p>by concatenating all of our <code>summary.txt</code> files into a
single file using the <code>cat</code> command. We will call this
<code>fastqc_summaries.txt</code> and move it to
<code>~/dc_workshop/docs</code>.</p>
<div class="codewrapper sourceCode" id="cb41">
<h3 class="code-label">BASH<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="sourceCode bash" tabindex="0"><code class="sourceCode bash"><span id="cb41-1"><a href="#cb41-1" tabindex="-1"></a><span class="ex">$</span> cat <span class="pp">*</span>/summary.txt <span class="op">&gt;</span> ~/dc_workshop/docs/fastqc_summaries.txt</span></code></pre>
</div>
<div id="exercise-3" class="callout challenge">
<div class="callout-square">
<i class="callout-icon" data-feather="zap"></i>
</div>
<div id="exercise-3" class="callout-inner">
<h3 class="callout-title">Exercise</h3>
<div class="callout-content">
<p>Which samples failed at least one of FastQC’s quality tests? What
test(s) did those samples fail?</p>
</div>
</div>
</div>
<div id="accordionSolution4" class="accordion challenge-accordion accordion-flush">
<div class="accordion-item">
<button class="accordion-button solution-button collapsed" type="button" data-bs-toggle="collapse" data-bs-target="#collapseSolution4" aria-expanded="false" aria-controls="collapseSolution4">
  <h4 class="accordion-header" id="headingSolution4"> Show me the solution </h4>
</button>
<div id="collapseSolution4" class="accordion-collapse collapse" aria-labelledby="headingSolution4" data-bs-parent="#accordionSolution4">
<div class="accordion-body">
<p>We can get the list of all failed tests using <code>grep</code>.</p>
<div class="codewrapper sourceCode" id="cb42">
<h3 class="code-label">BASH<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="sourceCode bash" tabindex="0"><code class="sourceCode bash"><span id="cb42-1"><a href="#cb42-1" tabindex="-1"></a><span class="ex">$</span> cd ~/dc_workshop/docs</span>
<span id="cb42-2"><a href="#cb42-2" tabindex="-1"></a><span class="ex">$</span> grep FAIL fastqc_summaries.txt</span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>FAIL    Per base sequence quality       SRR2584863_2.fastq.gz
FAIL    Per tile sequence quality       SRR2584863_2.fastq.gz
FAIL    Per base sequence content       SRR2584863_2.fastq.gz
FAIL    Per base sequence quality       SRR2584866_1.fastq.gz
FAIL    Per base sequence content       SRR2584866_1.fastq.gz
FAIL    Adapter Content SRR2584866_1.fastq.gz
FAIL    Adapter Content SRR2584866_2.fastq.gz
FAIL    Adapter Content SRR2589044_1.fastq.gz
FAIL    Per base sequence quality       SRR2589044_2.fastq.gz
FAIL    Per tile sequence quality       SRR2589044_2.fastq.gz
FAIL    Per base sequence content       SRR2589044_2.fastq.gz
FAIL    Adapter Content SRR2589044_2.fastq.gz</code></pre>
</div>
</div>
</div>
</div>
</div>
</div>
</section><section><h2 class="section-heading" id="other-notes-optional">Other notes – optional<a class="anchor" aria-label="anchor" href="#other-notes-optional"></a>
</h2>
<hr class="half-width">
<div id="quality-encodings-vary" class="callout">
<div class="callout-square">
<i class="callout-icon" data-feather="bell"></i>
</div>
<div id="quality-encodings-vary" class="callout-inner">
<h3 class="callout-title">Quality encodings vary</h3>
<div class="callout-content">
<p>Although we have used a particular quality encoding system to
demonstrate interpretation of read quality, different sequencing
machines use different encoding systems. This means that, depending on
which sequencer you use to generate your data, a <code>#</code> may not
be an indicator of a poor quality base call.</p>
<p>This mainly relates to older Solexa/Illumina data, but it is
essential that you know which sequencing platform was used to generate
your data, so that you can tell your quality control program which
encoding to use. If you choose the wrong encoding, you run the risk of
throwing away good reads or (even worse) not throwing away bad
reads!</p>
</div>
</div>
</div>
<div id="same-symbols-different-meanings" class="callout">
<div class="callout-square">
<i class="callout-icon" data-feather="bell"></i>
</div>
<div id="same-symbols-different-meanings" class="callout-inner">
<h3 class="callout-title">Same symbols, different meanings</h3>
<div class="callout-content">
<p>Here we see <code>&gt;</code> being used as a shell prompt, whereas
<code>&gt;</code> is also used to redirect output. Similarly,
<code>$</code> is used as a shell prompt, but, as we saw earlier, it is
also used to ask the shell to get the value of a variable.</p>
<p>If the <em>shell</em> prints <code>&gt;</code> or <code>$</code> then
it expects you to type something, and the symbol is a prompt.</p>
<p>If <em>you</em> type <code>&gt;</code> or <code>$</code> yourself, it
is an instruction from you that the shell should redirect output or get
the value of a variable.</p>
</div>
</div>
</div>
<div id="keypoints1" class="callout keypoints">
<div class="callout-square">
<i class="callout-icon" data-feather="key"></i>
</div>
<div class="callout-inner">
<h3 class="callout-title">Key Points</h3>
<div class="callout-content">
<ul>
<li>Quality encodings vary across sequencing platforms.</li>
<li>
<code>for</code> loops let you perform the same set of operations on
multiple files with a single command.</li>
</ul>
</div>
</div>
</div>
</section></section><section id="aio-03-trimming"><p>Content from <a href="03-trimming.html">Trimming and Filtering</a></p>
<hr>
<p>Last updated on 2025-02-04 |

        <a href="https://github.com/datacarpentry/wrangling-genomics/edit/main/episodes/03-trimming.md" class="external-link">Edit this page <i aria-hidden="true" data-feather="edit"></i></a></p>
<p>Estimated time: <i aria-hidden="true" data-feather="clock"></i> 55 minutes</p>
<div class="text-end">
          <button role="button" aria-pressed="false" tabindex="0" id="expand-code" class="pull-right" data-expand="Expand All Solutions " data-collapse="Collapse All Solutions "> Expand All Solutions <i aria-hidden="true" data-feather="plus"></i></button>
        </div>
<div class="overview card">
<h2 class="card-header">Overview</h2>
<div class="row g-0">
<div class="col-md-4">
<div class="card-body">
<div class="inner">
<h3 class="card-title">Questions</h3>
<ul>
<li>How can I get rid of sequence data that does not meet my quality
standards?</li>
</ul>
</div>
</div>
</div>
<div class="col-md-8">
<div class="card-body">
<div class="inner bordered">
<h3 class="card-title">Objectives</h3>
<ul>
<li>Clean FASTQ reads using Trimmomatic.</li>
<li>Select and set multiple options for command-line bioinformatic
tools.</li>
<li>Write <code>for</code> loops with two variables.</li>
</ul>
</div>
</div>
</div>
</div>
</div>
<section><h2 class="section-heading" id="cleaning-reads">Cleaning reads<a class="anchor" aria-label="anchor" href="#cleaning-reads"></a>
</h2>
<hr class="half-width">
<p>In the previous episode, we took a high-level look at the quality of
each of our samples using FastQC. We visualized per-base quality graphs
showing the distribution of read quality at each base across all reads
in a sample and extracted information about which samples fail which
quality checks. Some of our samples failed quite a few quality metrics
used by FastQC. This does not mean, though, that our samples should be
thrown out! It is very common to have some quality metrics fail, and
this may or may not be a problem for your downstream application. For
our variant calling workflow, we will be removing some of the low
quality sequences to reduce our false positive rate due to sequencing
error.</p>
<p>We will use a program called <a href="https://www.usadellab.org/cms/?page=trimmomatic" class="external-link">Trimmomatic</a>
to filter poor quality reads and trim poor quality bases from our
samples.</p>
</section><section><h2 class="section-heading" id="trimmomatic-options">Trimmomatic options<a class="anchor" aria-label="anchor" href="#trimmomatic-options"></a>
</h2>
<hr class="half-width">
<p>Trimmomatic has a variety of options to trim your reads. If we run
the following command, we can see some of our options.</p>
<div class="codewrapper sourceCode" id="cb1">
<h3 class="code-label">BASH<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="sourceCode bash" tabindex="0"><code class="sourceCode bash"><span id="cb1-1"><a href="#cb1-1" tabindex="-1"></a><span class="ex">$</span> trimmomatic</span></code></pre>
</div>
<p>Which will give you the following output:</p>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>Usage:
       PE [-version] [-threads &lt;threads&gt;] [-phred33|-phred64] [-trimlog &lt;trimLogFile&gt;] [-summary &lt;statsSummaryFile&gt;] [-quiet] [-validatePairs] [-basein &lt;inputBase&gt; | &lt;inputFile1&gt; &lt;inputFile2&gt;] [-baseout &lt;outputBase&gt; | &lt;outputFile1P&gt; &lt;outputFile1U&gt; &lt;outputFile2P&gt; &lt;outputFile2U&gt;] &lt;trimmer1&gt;...
   or:
       SE [-version] [-threads &lt;threads&gt;] [-phred33|-phred64] [-trimlog &lt;trimLogFile&gt;] [-summary &lt;statsSummaryFile&gt;] [-quiet] &lt;inputFile&gt; &lt;outputFile&gt; &lt;trimmer1&gt;...
   or:
       -version</code></pre>
</div>
<p>This output shows us that we must first specify whether we have
paired end (<code>PE</code>) or single end (<code>SE</code>) reads.
Next, we specify what flag we would like to run. For example, you can
specify <code>threads</code> to indicate the number of processors on
your computer that you want Trimmomatic to use. In most cases using
multiple threads (processors) can help to run the trimming faster. These
flags are not necessary, but they can give you more control over the
command. The flags are followed by positional arguments, meaning the
order in which you specify them is important. In paired end mode,
Trimmomatic expects the two input files, and then the names of the
output files. These files are described below. While, in single end
mode, Trimmomatic will expect 1 file as input, after which you can enter
the optional settings and lastly the name of the output file.</p>
<table class="table">
<colgroup>
<col width="11%">
<col width="88%">
</colgroup>
<thead><tr class="header">
<th>option</th>
<th>meaning</th>
</tr></thead>
<tbody>
<tr class="odd">
<td>&lt;inputFile1&gt;</td>
<td>Input reads to be trimmed. Typically the file name will contain an
<code>_1</code> or <code>_R1</code> in the name.</td>
</tr>
<tr class="even">
<td>&lt;inputFile2&gt;</td>
<td>Input reads to be trimmed. Typically the file name will contain an
<code>_2</code> or <code>_R2</code> in the name.</td>
</tr>
<tr class="odd">
<td>&lt;outputFile1P&gt;</td>
<td>Output file that contains surviving pairs from the <code>_1</code>
file.</td>
</tr>
<tr class="even">
<td>&lt;outputFile1U&gt;</td>
<td>Output file that contains orphaned reads from the <code>_1</code>
file.</td>
</tr>
<tr class="odd">
<td>&lt;outputFile2P&gt;</td>
<td>Output file that contains surviving pairs from the <code>_2</code>
file.</td>
</tr>
<tr class="even">
<td>&lt;outputFile2U&gt;</td>
<td>Output file that contains orphaned reads from the <code>_2</code>
file.</td>
</tr>
</tbody>
</table>
<p>The last thing trimmomatic expects to see is the trimming
parameters:</p>
<table class="table">
<colgroup>
<col width="11%">
<col width="88%">
</colgroup>
<thead><tr class="header">
<th>step</th>
<th>meaning</th>
</tr></thead>
<tbody>
<tr class="odd">
<td><code>ILLUMINACLIP</code></td>
<td>Perform adapter removal.</td>
</tr>
<tr class="even">
<td><code>SLIDINGWINDOW</code></td>
<td>Perform sliding window trimming, cutting once the average quality
within the window falls below a threshold.</td>
</tr>
<tr class="odd">
<td><code>LEADING</code></td>
<td>Cut bases off the start of a read, if below a threshold
quality.</td>
</tr>
<tr class="even">
<td><code>TRAILING</code></td>
<td>Cut bases off the end of a read, if below a threshold quality.</td>
</tr>
<tr class="odd">
<td><code>CROP</code></td>
<td>Cut the read to a specified length.</td>
</tr>
<tr class="even">
<td><code>HEADCROP</code></td>
<td>Cut the specified number of bases from the start of the read.</td>
</tr>
<tr class="odd">
<td><code>MINLEN</code></td>
<td>Drop an entire read if it is below a specified length.</td>
</tr>
<tr class="even">
<td><code>TOPHRED33</code></td>
<td>Convert quality scores to Phred-33.</td>
</tr>
<tr class="odd">
<td><code>TOPHRED64</code></td>
<td>Convert quality scores to Phred-64.</td>
</tr>
</tbody>
</table>
<p>We will use only a few of these options and trimming steps in our
analysis. It is important to understand the steps you are using to clean
your data. For more information about the Trimmomatic arguments and
options, see <a href="https://www.usadellab.org/cms/uploads/supplementary/Trimmomatic/TrimmomaticManual_V0.32.pdf" class="external-link">the
Trimmomatic manual</a>.</p>
<p>However, a complete command for Trimmomatic will look something like
the command below. This command is an example and will not work, as we
do not have the files it refers to:</p>
<div class="codewrapper sourceCode" id="cb3">
<h3 class="code-label">BASH<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="sourceCode bash" tabindex="0"><code class="sourceCode bash"><span id="cb3-1"><a href="#cb3-1" tabindex="-1"></a><span class="ex">$</span> trimmomatic PE <span class="at">-threads</span> 4 SRR_1056_1.fastq SRR_1056_2.fastq  <span class="dt">\</span></span>
<span id="cb3-2"><a href="#cb3-2" tabindex="-1"></a>              SRR_1056_1.trimmed.fastq SRR_1056_1un.trimmed.fastq <span class="dt">\</span></span>
<span id="cb3-3"><a href="#cb3-3" tabindex="-1"></a>              SRR_1056_2.trimmed.fastq SRR_1056_2un.trimmed.fastq <span class="dt">\</span></span>
<span id="cb3-4"><a href="#cb3-4" tabindex="-1"></a>              ILLUMINACLIP:SRR_adapters.fa SLIDINGWINDOW:4:20</span></code></pre>
</div>
<p>In this example, we have told Trimmomatic:</p>
<table class="table">
<colgroup>
<col width="11%">
<col width="88%">
</colgroup>
<thead><tr class="header">
<th>code</th>
<th>meaning</th>
</tr></thead>
<tbody>
<tr class="odd">
<td><code>PE</code></td>
<td>that it will be taking a paired end file as input</td>
</tr>
<tr class="even">
<td><code>-threads 4</code></td>
<td>to use four computing threads to run (this will speed up our
run)</td>
</tr>
<tr class="odd">
<td><code>SRR_1056_1.fastq</code></td>
<td>the first input file name</td>
</tr>
<tr class="even">
<td><code>SRR_1056_2.fastq</code></td>
<td>the second input file name</td>
</tr>
<tr class="odd">
<td><code>SRR_1056_1.trimmed.fastq</code></td>
<td>the output file for surviving pairs from the <code>_1</code>
file</td>
</tr>
<tr class="even">
<td><code>SRR_1056_1un.trimmed.fastq</code></td>
<td>the output file for orphaned reads from the <code>_1</code>
file</td>
</tr>
<tr class="odd">
<td><code>SRR_1056_2.trimmed.fastq</code></td>
<td>the output file for surviving pairs from the <code>_2</code>
file</td>
</tr>
<tr class="even">
<td><code>SRR_1056_2un.trimmed.fastq</code></td>
<td>the output file for orphaned reads from the <code>_2</code>
file</td>
</tr>
<tr class="odd">
<td><code>ILLUMINACLIP:SRR_adapters.fa</code></td>
<td>to clip the Illumina adapters from the input file using the adapter
sequences listed in <code>SRR_adapters.fa</code>
</td>
</tr>
<tr class="even">
<td><code>SLIDINGWINDOW:4:20</code></td>
<td>to use a sliding window of size 4 that will remove bases if their
phred score is below 20</td>
</tr>
</tbody>
</table>
<div id="multi-line-commands" class="callout">
<div class="callout-square">
<i class="callout-icon" data-feather="bell"></i>
</div>
<div id="multi-line-commands" class="callout-inner">
<h3 class="callout-title">Multi-line commands</h3>
<div class="callout-content">
<p>Some of the commands we ran in this lesson are long! When typing a
long command into your terminal, you can use the <code>\</code>
character to separate code chunks onto separate lines. This can make
your code more readable.</p>
</div>
</div>
</div>
</section><section><h2 class="section-heading" id="running-trimmomatic">Running Trimmomatic<a class="anchor" aria-label="anchor" href="#running-trimmomatic"></a>
</h2>
<hr class="half-width">
<p>Now we will run Trimmomatic on our data. To begin, navigate to your
<code>untrimmed_fastq</code> data directory:</p>
<div class="codewrapper sourceCode" id="cb4">
<h3 class="code-label">BASH<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="sourceCode bash" tabindex="0"><code class="sourceCode bash"><span id="cb4-1"><a href="#cb4-1" tabindex="-1"></a><span class="ex">$</span> cd ~/dc_workshop/data/untrimmed_fastq</span></code></pre>
</div>
<p>We are going to run Trimmomatic on one of our paired-end samples.
While using FastQC we saw that Nextera adapters were present in our
samples. The adapter sequences came with the installation of
trimmomatic, so we will first copy these sequences into our current
directory.</p>
<div class="codewrapper sourceCode" id="cb5">
<h3 class="code-label">BASH<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="sourceCode bash" tabindex="0"><code class="sourceCode bash"><span id="cb5-1"><a href="#cb5-1" tabindex="-1"></a><span class="ex">$</span> cp ~/.miniconda3/pkgs/trimmomatic-0.38-0/share/trimmomatic-0.38-0/adapters/NexteraPE-PE.fa .</span></code></pre>
</div>
<p>We will also use a sliding window of size 4 that will remove bases if
their phred score is below 20 (like in our example above). We will also
discard any reads that do not have at least 25 bases remaining after
this trimming step. Three additional pieces of code are also added to
the end of the ILLUMINACLIP step. These three additional numbers
(2:40:15) tell Trimmimatic how to handle sequence matches to the Nextera
adapters. A detailed explanation of how they work is advanced for this
particular lesson. For now we will use these numbers as a default and
recognize they are needed to for Trimmomatic to run properly. This
command will take a few minutes to run.</p>
<div class="codewrapper sourceCode" id="cb6">
<h3 class="code-label">BASH<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="sourceCode bash" tabindex="0"><code class="sourceCode bash"><span id="cb6-1"><a href="#cb6-1" tabindex="-1"></a><span class="ex">$</span> trimmomatic PE SRR2589044_1.fastq.gz SRR2589044_2.fastq.gz <span class="dt">\</span></span>
<span id="cb6-2"><a href="#cb6-2" tabindex="-1"></a>                SRR2589044_1.trim.fastq.gz SRR2589044_1un.trim.fastq.gz <span class="dt">\</span></span>
<span id="cb6-3"><a href="#cb6-3" tabindex="-1"></a>                SRR2589044_2.trim.fastq.gz SRR2589044_2un.trim.fastq.gz <span class="dt">\</span></span>
<span id="cb6-4"><a href="#cb6-4" tabindex="-1"></a>                SLIDINGWINDOW:4:20 MINLEN:25 ILLUMINACLIP:NexteraPE-PE.fa:2:40:15</span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>TrimmomaticPE: Started with arguments:
 SRR2589044_1.fastq.gz SRR2589044_2.fastq.gz SRR2589044_1.trim.fastq.gz SRR2589044_1un.trim.fastq.gz SRR2589044_2.trim.fastq.gz SRR2589044_2un.trim.fastq.gz SLIDINGWINDOW:4:20 MINLEN:25 ILLUMINACLIP:NexteraPE-PE.fa:2:40:15
Multiple cores found: Using 2 threads
Using PrefixPair: 'AGATGTGTATAAGAGACAG' and 'AGATGTGTATAAGAGACAG'
Using Long Clipping Sequence: 'GTCTCGTGGGCTCGGAGATGTGTATAAGAGACAG'
Using Long Clipping Sequence: 'TCGTCGGCAGCGTCAGATGTGTATAAGAGACAG'
Using Long Clipping Sequence: 'CTGTCTCTTATACACATCTCCGAGCCCACGAGAC'
Using Long Clipping Sequence: 'CTGTCTCTTATACACATCTGACGCTGCCGACGA'
ILLUMINACLIP: Using 1 prefix pairs, 4 forward/reverse sequences, 0 forward only sequences, 0 reverse only sequences
Quality encoding detected as phred33
Input Read Pairs: 1107090 Both Surviving: 885220 (79.96%) Forward Only Surviving: 216472 (19.55%) Reverse Only Surviving: 2850 (0.26%) Dropped: 2548 (0.23%)
TrimmomaticPE: Completed successfully</code></pre>
</div>
<div id="exercise" class="callout challenge">
<div class="callout-square">
<i class="callout-icon" data-feather="zap"></i>
</div>
<div id="exercise" class="callout-inner">
<h3 class="callout-title">Exercise</h3>
<div class="callout-content">
<p>Use the output from your Trimmomatic command to answer the following
questions.</p>
<ol style="list-style-type: decimal">
<li>What percent of reads did we discard from our sample?</li>
<li>What percent of reads did we keep both pairs?</li>
</ol>
</div>
</div>
</div>
<div id="accordionSolution1" class="accordion challenge-accordion accordion-flush">
<div class="accordion-item">
<button class="accordion-button solution-button collapsed" type="button" data-bs-toggle="collapse" data-bs-target="#collapseSolution1" aria-expanded="false" aria-controls="collapseSolution1">
  <h4 class="accordion-header" id="headingSolution1"> Show me the solution </h4>
</button>
<div id="collapseSolution1" class="accordion-collapse collapse" data-bs-parent="#accordionSolution1" aria-labelledby="headingSolution1">
<div class="accordion-body">
<ol style="list-style-type: decimal">
<li>0.23%</li>
<li>79.96%</li>
</ol>
</div>
</div>
</div>
</div>
<p>You may have noticed that Trimmomatic automatically detected the
quality encoding of our sample. It is always a good idea to double-check
this or to enter the quality encoding manually.</p>
<p>We can confirm that we have our output files:</p>
<div class="codewrapper sourceCode" id="cb8">
<h3 class="code-label">BASH<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="sourceCode bash" tabindex="0"><code class="sourceCode bash"><span id="cb8-1"><a href="#cb8-1" tabindex="-1"></a><span class="ex">$</span> ls SRR2589044<span class="pp">*</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>SRR2589044_1.fastq.gz       SRR2589044_1un.trim.fastq.gz  SRR2589044_2.trim.fastq.gz
SRR2589044_1.trim.fastq.gz  SRR2589044_2.fastq.gz         SRR2589044_2un.trim.fastq.gz</code></pre>
</div>
<p>The output files are also FASTQ files. It should be smaller than our
input file, because we have removed reads. We can confirm this:</p>
<div class="codewrapper sourceCode" id="cb10">
<h3 class="code-label">BASH<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="sourceCode bash" tabindex="0"><code class="sourceCode bash"><span id="cb10-1"><a href="#cb10-1" tabindex="-1"></a><span class="ex">$</span> ls SRR2589044<span class="pp">*</span> <span class="at">-l</span> <span class="at">-h</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>-rw-rw-r-- 1 dcuser dcuser 124M Jul  6 20:22 SRR2589044_1.fastq.gz
-rw-rw-r-- 1 dcuser dcuser  94M Jul  6 22:33 SRR2589044_1.trim.fastq.gz
-rw-rw-r-- 1 dcuser dcuser  18M Jul  6 22:33 SRR2589044_1un.trim.fastq.gz
-rw-rw-r-- 1 dcuser dcuser 128M Jul  6 20:24 SRR2589044_2.fastq.gz
-rw-rw-r-- 1 dcuser dcuser  91M Jul  6 22:33 SRR2589044_2.trim.fastq.gz
-rw-rw-r-- 1 dcuser dcuser 271K Jul  6 22:33 SRR2589044_2un.trim.fastq.gz</code></pre>
</div>
<p>We have just successfully run Trimmomatic on one of our FASTQ files!
However, there is some bad news. Trimmomatic can only operate on one
sample at a time and we have more than one sample. The good news is that
we can use a <code>for</code> loop to iterate through our sample files
quickly!</p>
<p>We unzipped one of our files before to work with it, let’s compress
it again before we run our for loop.</p>
<div class="codewrapper sourceCode" id="cb12">
<h3 class="code-label">BASH<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="sourceCode bash" tabindex="0"><code class="sourceCode bash"><span id="cb12-1"><a href="#cb12-1" tabindex="-1"></a><span class="fu">gzip</span> SRR2584863_1.fastq </span></code></pre>
</div>
<div class="codewrapper sourceCode" id="cb13">
<h3 class="code-label">BASH<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="sourceCode bash" tabindex="0"><code class="sourceCode bash"><span id="cb13-1"><a href="#cb13-1" tabindex="-1"></a><span class="ex">$</span> for infile in <span class="pp">*</span>_1.fastq.gz</span>
<span id="cb13-2"><a href="#cb13-2" tabindex="-1"></a><span class="op">&gt;</span> do</span>
<span id="cb13-3"><a href="#cb13-3" tabindex="-1"></a><span class="op">&gt;</span>   base=<span class="va">$(</span><span class="fu">basename</span> <span class="va">${infile}</span> _1.fastq.gz<span class="va">)</span></span>
<span id="cb13-4"><a href="#cb13-4" tabindex="-1"></a><span class="op">&gt;</span>   trimmomatic <span class="ex">PE</span> <span class="va">${infile}</span> <span class="va">${base}</span>_2.fastq.gz <span class="dt">\</span></span>
<span id="cb13-5"><a href="#cb13-5" tabindex="-1"></a><span class="op">&gt;</span>                <span class="va">${base}</span>_1.trim.fastq.gz <span class="va">${base}</span>_1un.trim.fastq.gz <span class="dt">\</span></span>
<span id="cb13-6"><a href="#cb13-6" tabindex="-1"></a><span class="op">&gt;</span>                <span class="va">${base}</span>_2.trim.fastq.gz <span class="va">${base}</span>_2un.trim.fastq.gz <span class="dt">\</span></span>
<span id="cb13-7"><a href="#cb13-7" tabindex="-1"></a><span class="op">&gt;</span>                SLIDINGWINDOW:4:20 MINLEN:25 ILLUMINACLIP:NexteraPE-PE.fa:2:40:15 </span>
<span id="cb13-8"><a href="#cb13-8" tabindex="-1"></a><span class="op">&gt;</span> done</span></code></pre>
</div>
<p>Go ahead and run the for loop. It should take a few minutes for
Trimmomatic to run for each of our six input files. Once it is done
running, take a look at your directory contents. You will notice that
even though we ran Trimmomatic on file <code>SRR2589044</code> before
running the for loop, there is only one set of files for it. Because we
matched the ending <code>_1.fastq.gz</code>, we re-ran Trimmomatic on
this file, overwriting our first results. That is ok, but it is good to
be aware that it happened.</p>
<div class="codewrapper sourceCode" id="cb14">
<h3 class="code-label">BASH<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="sourceCode bash" tabindex="0"><code class="sourceCode bash"><span id="cb14-1"><a href="#cb14-1" tabindex="-1"></a><span class="ex">$</span> ls</span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>NexteraPE-PE.fa               SRR2584866_1.fastq.gz         SRR2589044_1.trim.fastq.gz
SRR2584863_1.fastq.gz         SRR2584866_1.trim.fastq.gz    SRR2589044_1un.trim.fastq.gz
SRR2584863_1.trim.fastq.gz    SRR2584866_1un.trim.fastq.gz  SRR2589044_2.fastq.gz
SRR2584863_1un.trim.fastq.gz  SRR2584866_2.fastq.gz         SRR2589044_2.trim.fastq.gz
SRR2584863_2.fastq.gz         SRR2584866_2.trim.fastq.gz    SRR2589044_2un.trim.fastq.gz
SRR2584863_2.trim.fastq.gz    SRR2584866_2un.trim.fastq.gz
SRR2584863_2un.trim.fastq.gz  SRR2589044_1.fastq.gz</code></pre>
</div>
<div id="exercise-1" class="callout challenge">
<div class="callout-square">
<i class="callout-icon" data-feather="zap"></i>
</div>
<div id="exercise-1" class="callout-inner">
<h3 class="callout-title">Exercise</h3>
<div class="callout-content">
<p>We trimmed our fastq files with Nextera adapters, but there are other
adapters that are commonly used. What other adapter files came with
Trimmomatic?</p>
</div>
</div>
</div>
<div id="accordionSolution2" class="accordion challenge-accordion accordion-flush">
<div class="accordion-item">
<button class="accordion-button solution-button collapsed" type="button" data-bs-toggle="collapse" data-bs-target="#collapseSolution2" aria-expanded="false" aria-controls="collapseSolution2">
  <h4 class="accordion-header" id="headingSolution2"> Show me the solution </h4>
</button>
<div id="collapseSolution2" class="accordion-collapse collapse" data-bs-parent="#accordionSolution2" aria-labelledby="headingSolution2">
<div class="accordion-body">
<div class="codewrapper sourceCode" id="cb16">
<h3 class="code-label">BASH<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="sourceCode bash" tabindex="0"><code class="sourceCode bash"><span id="cb16-1"><a href="#cb16-1" tabindex="-1"></a><span class="ex">$</span> ls ~/.miniconda3/pkgs/trimmomatic-0.38-0/share/trimmomatic-0.38-0/adapters/</span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>NexteraPE-PE.fa  TruSeq2-SE.fa    TruSeq3-PE.fa
TruSeq2-PE.fa    TruSeq3-PE-2.fa  TruSeq3-SE.fa</code></pre>
</div>
</div>
</div>
</div>
</div>
<p>We have now completed the trimming and filtering steps of our quality
control process! Before we move on, let’s move our trimmed FASTQ files
to a new subdirectory within our <code>data/</code> directory.</p>
<div class="codewrapper sourceCode" id="cb18">
<h3 class="code-label">BASH<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="sourceCode bash" tabindex="0"><code class="sourceCode bash"><span id="cb18-1"><a href="#cb18-1" tabindex="-1"></a><span class="ex">$</span> cd ~/dc_workshop/data/untrimmed_fastq</span>
<span id="cb18-2"><a href="#cb18-2" tabindex="-1"></a><span class="ex">$</span> mkdir ../trimmed_fastq</span>
<span id="cb18-3"><a href="#cb18-3" tabindex="-1"></a><span class="ex">$</span> mv <span class="pp">*</span>.trim<span class="pp">*</span> ../trimmed_fastq</span>
<span id="cb18-4"><a href="#cb18-4" tabindex="-1"></a><span class="ex">$</span> cd ../trimmed_fastq</span>
<span id="cb18-5"><a href="#cb18-5" tabindex="-1"></a><span class="ex">$</span> ls</span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>SRR2584863_1.trim.fastq.gz    SRR2584866_1.trim.fastq.gz    SRR2589044_1.trim.fastq.gz
SRR2584863_1un.trim.fastq.gz  SRR2584866_1un.trim.fastq.gz  SRR2589044_1un.trim.fastq.gz
SRR2584863_2.trim.fastq.gz    SRR2584866_2.trim.fastq.gz    SRR2589044_2.trim.fastq.gz
SRR2584863_2un.trim.fastq.gz  SRR2584866_2un.trim.fastq.gz  SRR2589044_2un.trim.fastq.gz</code></pre>
</div>
<div id="bonus-exercise-advanced" class="callout challenge">
<div class="callout-square">
<i class="callout-icon" data-feather="zap"></i>
</div>
<div id="bonus-exercise-advanced" class="callout-inner">
<h3 class="callout-title">Bonus exercise (advanced)</h3>
<div class="callout-content">
<p>Now that our samples have gone through quality control, they should
perform better on the quality tests run by FastQC. Go ahead and re-run
FastQC on your trimmed FASTQ files and visualize the HTML files to see
whether your per base sequence quality is higher after trimming.</p>
</div>
</div>
</div>
<div id="accordionSolution3" class="accordion challenge-accordion accordion-flush">
<div class="accordion-item">
<button class="accordion-button solution-button collapsed" type="button" data-bs-toggle="collapse" data-bs-target="#collapseSolution3" aria-expanded="false" aria-controls="collapseSolution3">
  <h4 class="accordion-header" id="headingSolution3"> Show me the solution </h4>
</button>
<div id="collapseSolution3" class="accordion-collapse collapse" data-bs-parent="#accordionSolution3" aria-labelledby="headingSolution3">
<div class="accordion-body">
<p>In your AWS terminal window do:</p>
<div class="codewrapper sourceCode" id="cb20">
<h3 class="code-label">BASH<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="sourceCode bash" tabindex="0"><code class="sourceCode bash"><span id="cb20-1"><a href="#cb20-1" tabindex="-1"></a><span class="ex">$</span> fastqc ~/dc_workshop/data/trimmed_fastq/<span class="pp">*</span>.fastq<span class="pp">*</span></span></code></pre>
</div>
<p>In a new tab in your terminal do:</p>
<div class="codewrapper sourceCode" id="cb21">
<h3 class="code-label">BASH<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="sourceCode bash" tabindex="0"><code class="sourceCode bash"><span id="cb21-1"><a href="#cb21-1" tabindex="-1"></a><span class="ex">$</span> mkdir ~/Desktop/fastqc_html/trimmed</span>
<span id="cb21-2"><a href="#cb21-2" tabindex="-1"></a><span class="ex">$</span> scp dcuser@ec2-34-203-203-131.compute-1.amazonaws.com:~/dc_workshop/data/trimmed_fastq/<span class="pp">*</span>.html ~/Desktop/fastqc_html/trimmed</span></code></pre>
</div>
<p>Then take a look at the html files in your browser.</p>
<p>Remember to replace everything between the <code>@</code> and
<code>:</code> in your scp command with your AWS instance number.</p>
<p>After trimming and filtering, our overall quality is much higher, we
have a distribution of sequence lengths, and more samples pass adapter
content. However, quality trimming is not perfect, and some programs are
better at removing some sequences than others. Because our sequences
still contain 3’ adapters, it could be important to explore other
trimming tools like <a href="https://cutadapt.readthedocs.io/en/stable/" class="external-link">cutadapt</a> to remove
these, depending on your downstream application. Trimmomatic did pretty
well though, and its performance is good enough for our workflow.</p>
</div>
</div>
</div>
</div>
<div id="keypoints1" class="callout keypoints">
<div class="callout-square">
<i class="callout-icon" data-feather="key"></i>
</div>
<div class="callout-inner">
<h3 class="callout-title">Key Points</h3>
<div class="callout-content">
<ul>
<li>The options you set for the command-line tools you use are
important!</li>
<li>Data cleaning is an essential step in a genomics workflow.</li>
</ul>
</div>
</div>
</div>
</section></section><section id="aio-04-variant_calling"><p>Content from <a href="04-variant_calling.html">Variant Calling Workflow</a></p>
<hr>
<p>Last updated on 2024-07-08 |

        <a href="https://github.com/datacarpentry/wrangling-genomics/edit/main/episodes/04-variant_calling.md" class="external-link">Edit this page <i aria-hidden="true" data-feather="edit"></i></a></p>
<p>Estimated time: <i aria-hidden="true" data-feather="clock"></i> 60 minutes</p>
<div class="text-end">
          <button role="button" aria-pressed="false" tabindex="0" id="expand-code" class="pull-right" data-expand="Expand All Solutions " data-collapse="Collapse All Solutions "> Expand All Solutions <i aria-hidden="true" data-feather="plus"></i></button>
        </div>
<div class="overview card">
<h2 class="card-header">Overview</h2>
<div class="row g-0">
<div class="col-md-4">
<div class="card-body">
<div class="inner">
<h3 class="card-title">Questions</h3>
<ul>
<li>How do I find sequence variants between my sample and a reference
genome?</li>
</ul>
</div>
</div>
</div>
<div class="col-md-8">
<div class="card-body">
<div class="inner bordered">
<h3 class="card-title">Objectives</h3>
<ul>
<li>Understand the steps involved in variant calling.</li>
<li>Describe the types of data formats encountered during variant
calling.</li>
<li>Use command line tools to perform variant calling.</li>
</ul>
</div>
</div>
</div>
</div>
</div>
<p>We mentioned before that we are working with files from a long-term
evolution study of an <em>E. coli</em> population (designated Ara-3).
Now that we have looked at our data to make sure that it is high
quality, and removed low-quality base calls, we can perform variant
calling to see how the population changed over time. We care how this
population changed relative to the original population, <em>E. coli</em>
strain REL606. Therefore, we will align each of our samples to the
<em>E. coli</em> REL606 reference genome, and see what differences exist
in our reads versus the genome.</p>
<section><h2 class="section-heading" id="alignment-to-a-reference-genome">Alignment to a reference genome<a class="anchor" aria-label="anchor" href="#alignment-to-a-reference-genome"></a>
</h2>
<hr class="half-width">
<figure><img src="../fig/variant_calling_workflow_align.png" alt="workflow_align" class="figure mx-auto d-block"></figure><p>We perform read alignment or mapping to determine where in the genome
our reads originated from. There are a number of tools to choose from
and, while there is no gold standard, there are some tools that are
better suited for particular NGS analyses. We will be using the <a href="https://bio-bwa.sourceforge.net/" class="external-link">Burrows Wheeler Aligner
(BWA)</a>, which is a software package for mapping low-divergent
sequences against a large reference genome.</p>
<p>The alignment process consists of two steps:</p>
<ol style="list-style-type: decimal">
<li>Indexing the reference genome</li>
<li>Aligning the reads to the reference genome</li>
</ol></section><section><h2 class="section-heading" id="setting-up">Setting up<a class="anchor" aria-label="anchor" href="#setting-up"></a>
</h2>
<hr class="half-width">
<p>First we download the reference genome for <em>E. coli</em> REL606.
Although we could copy or move the file with <code>cp</code> or
<code>mv</code>, most genomics workflows begin with a download step, so
we will practice that here.</p>
<div class="codewrapper sourceCode" id="cb1">
<h3 class="code-label">BASH<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="sourceCode bash" tabindex="0"><code class="sourceCode bash"><span id="cb1-1"><a href="#cb1-1" tabindex="-1"></a><span class="ex">$</span> cd ~/dc_workshop</span>
<span id="cb1-2"><a href="#cb1-2" tabindex="-1"></a><span class="ex">$</span> mkdir <span class="at">-p</span> data/ref_genome</span>
<span id="cb1-3"><a href="#cb1-3" tabindex="-1"></a><span class="ex">$</span> curl <span class="at">-L</span> <span class="at">-o</span> data/ref_genome/ecoli_rel606.fasta.gz ftp://ftp.ncbi.nlm.nih.gov/genomes/all/GCA/000/017/985/GCA_000017985.1_ASM1798v1/GCA_000017985.1_ASM1798v1_genomic.fna.gz</span>
<span id="cb1-4"><a href="#cb1-4" tabindex="-1"></a><span class="ex">$</span> gunzip data/ref_genome/ecoli_rel606.fasta.gz</span></code></pre>
</div>
<div id="exercise" class="callout challenge">
<div class="callout-square">
<i class="callout-icon" data-feather="zap"></i>
</div>
<div id="exercise" class="callout-inner">
<h3 class="callout-title">Exercise</h3>
<div class="callout-content">
<p>We saved this file as
<code>data/ref_genome/ecoli_rel606.fasta.gz</code> and then decompressed
it. What is the real name of the genome?</p>
</div>
</div>
</div>
<div id="accordionSolution1" class="accordion challenge-accordion accordion-flush">
<div class="accordion-item">
<button class="accordion-button solution-button collapsed" type="button" data-bs-toggle="collapse" data-bs-target="#collapseSolution1" aria-expanded="false" aria-controls="collapseSolution1">
  <h4 class="accordion-header" id="headingSolution1"> Show me the solution </h4>
</button>
<div id="collapseSolution1" class="accordion-collapse collapse" aria-labelledby="headingSolution1" data-bs-parent="#accordionSolution1">
<div class="accordion-body">
<div class="codewrapper sourceCode" id="cb2">
<h3 class="code-label">BASH<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="sourceCode bash" tabindex="0"><code class="sourceCode bash"><span id="cb2-1"><a href="#cb2-1" tabindex="-1"></a><span class="ex">$</span> head data/ref_genome/ecoli_rel606.fasta</span></code></pre>
</div>
<p>The name of the sequence follows the <code>&gt;</code> character. The
name is
<code>CP000819.1 Escherichia coli B str. REL606, complete genome</code>.
Keep this chromosome name (<code>CP000819.1</code>) in mind, as we will
use it later in the lesson.</p>
</div>
</div>
</div>
</div>
<p>We will also download a set of trimmed FASTQ files to work with.
These are small subsets of our real trimmed data, and will enable us to
run our variant calling workflow quite quickly.</p>
<div class="codewrapper sourceCode" id="cb3">
<h3 class="code-label">BASH<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="sourceCode bash" tabindex="0"><code class="sourceCode bash"><span id="cb3-1"><a href="#cb3-1" tabindex="-1"></a><span class="ex">$</span> curl <span class="at">-L</span> <span class="at">-o</span> sub.tar.gz https://ndownloader.figshare.com/files/14418248</span>
<span id="cb3-2"><a href="#cb3-2" tabindex="-1"></a><span class="ex">$</span> tar xvf sub.tar.gz</span>
<span id="cb3-3"><a href="#cb3-3" tabindex="-1"></a><span class="ex">$</span> mv sub/ ~/dc_workshop/data/trimmed_fastq_small</span></code></pre>
</div>
<p>You will also need to create directories for the results that will be
generated as part of this workflow. We can do this in a single line of
code, because <code>mkdir</code> can accept multiple new directory names
as input.</p>
<div class="codewrapper sourceCode" id="cb4">
<h3 class="code-label">BASH<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="sourceCode bash" tabindex="0"><code class="sourceCode bash"><span id="cb4-1"><a href="#cb4-1" tabindex="-1"></a><span class="ex">$</span> mkdir <span class="at">-p</span> results/sam results/bam results/bcf results/vcf</span></code></pre>
</div>
<div class="section level4">
<h4 id="index-the-reference-genome">Index the reference genome<a class="anchor" aria-label="anchor" href="#index-the-reference-genome"></a>
</h4>
<p>Our first step is to index the reference genome for use by BWA.
Indexing allows the aligner to quickly find potential alignment sites
for query sequences in a genome, which saves time during alignment.
Indexing the reference only has to be run once. The only reason you
would want to create a new index is if you are working with a different
reference genome or you are using a different tool for alignment.</p>
<div class="codewrapper sourceCode" id="cb5">
<h3 class="code-label">BASH<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="sourceCode bash" tabindex="0"><code class="sourceCode bash"><span id="cb5-1"><a href="#cb5-1" tabindex="-1"></a><span class="ex">$</span> bwa index data/ref_genome/ecoli_rel606.fasta</span></code></pre>
</div>
<p>While the index is created, you will see output that looks something
like this:</p>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>[bwa_index] Pack FASTA... 0.04 sec
[bwa_index] Construct BWT for the packed sequence...
[bwa_index] 1.05 seconds elapse.
[bwa_index] Update BWT... 0.03 sec
[bwa_index] Pack forward-only FASTA... 0.02 sec
[bwa_index] Construct SA from BWT and Occ... 0.57 sec
[main] Version: 0.7.17-r1188
[main] CMD: bwa index data/ref_genome/ecoli_rel606.fasta
[main] Real time: 1.765 sec; CPU: 1.715 sec</code></pre>
</div>
</div>
<div class="section level4">
<h4 id="align-reads-to-reference-genome">Align reads to reference genome<a class="anchor" aria-label="anchor" href="#align-reads-to-reference-genome"></a>
</h4>
<p>The alignment process consists of choosing an appropriate reference
genome to map our reads against and then deciding on an aligner. We will
use the BWA-MEM algorithm, which is the latest and is generally
recommended for high-quality queries as it is faster and more
accurate.</p>
<p>An example of what a <code>bwa</code> command looks like is below.
This command will not run, as we do not have the files
<code>ref_genome.fa</code>, <code>input_file_R1.fastq</code>, or
<code>input_file_R2.fastq</code>.</p>
<div class="codewrapper sourceCode" id="cb7">
<h3 class="code-label">BASH<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="sourceCode bash" tabindex="0"><code class="sourceCode bash"><span id="cb7-1"><a href="#cb7-1" tabindex="-1"></a><span class="ex">$</span> bwa mem ref_genome.fasta input_file_R1.fastq input_file_R2.fastq <span class="op">&gt;</span> output.sam</span></code></pre>
</div>
<p>Have a look at the <a href="https://bio-bwa.sourceforge.net/bwa.shtml" class="external-link">bwa options page</a>.
While we are running bwa with the default parameters here, your use case
might require a change of parameters. <em>NOTE: Always read the manual
page for any tool before using and make sure the options you use are
appropriate for your data.</em></p>
<p>We are going to start by aligning the reads from just one of the
samples in our dataset (<code>SRR2584866</code>). Later, we will be
iterating this whole process on all of our sample files.</p>
<div class="codewrapper sourceCode" id="cb8">
<h3 class="code-label">BASH<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="sourceCode bash" tabindex="0"><code class="sourceCode bash"><span id="cb8-1"><a href="#cb8-1" tabindex="-1"></a><span class="ex">$</span> bwa mem data/ref_genome/ecoli_rel606.fasta data/trimmed_fastq_small/SRR2584866_1.trim.sub.fastq data/trimmed_fastq_small/SRR2584866_2.trim.sub.fastq <span class="op">&gt;</span> results/sam/SRR2584866.aligned.sam</span></code></pre>
</div>
<p>You will see output that starts like this:</p>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>[M::bwa_idx_load_from_disk] read 0 ALT contigs
[M::process] read 77446 sequences (10000033 bp)...
[M::process] read 77296 sequences (10000182 bp)...
[M::mem_pestat] # candidate unique pairs for (FF, FR, RF, RR): (48, 36728, 21, 61)
[M::mem_pestat] analyzing insert size distribution for orientation FF...
[M::mem_pestat] (25, 50, 75) percentile: (420, 660, 1774)
[M::mem_pestat] low and high boundaries for computing mean and std.dev: (1, 4482)
[M::mem_pestat] mean and std.dev: (784.68, 700.87)
[M::mem_pestat] low and high boundaries for proper pairs: (1, 5836)
[M::mem_pestat] analyzing insert size distribution for orientation FR...</code></pre>
</div>
<div class="section level5">
<h5 id="sambam-format">SAM/BAM format<a class="anchor" aria-label="anchor" href="#sambam-format"></a>
</h5>
<p>The <a href="https://genome.sph.umich.edu/wiki/SAM" class="external-link">SAM file</a>, is
a tab-delimited text file that contains information for each individual
read and its alignment to the genome. While we do not have time to go
into detail about the features of the SAM format, the paper by <a href="https://bioinformatics.oxfordjournals.org/content/25/16/2078.full" class="external-link">Heng
Li et al.</a> provides a lot more detail on the specification.</p>
<p><strong>The compressed binary version of SAM is called a BAM
file.</strong> We use this version to reduce size and to allow for
<em>indexing</em>, which enables efficient random access of the data
contained within the file.</p>
<p>The file begins with a <strong>header</strong>, which is optional.
The header is used to describe the source of data, reference sequence,
method of alignment, etc., this will change depending on the aligner
being used. Following the header is the <strong>alignment
section</strong>. Each line that follows corresponds to alignment
information for a single read. Each alignment line has <strong>11
mandatory fields</strong> for essential mapping information and a
variable number of other fields for aligner specific information. An
example entry from a SAM file is displayed below with the different
fields highlighted.</p>
<figure><img src="../fig/sam_bam.png" alt="sam_bam1" class="figure mx-auto d-block"></figure><figure><img src="../fig/sam_bam3.png" alt="sam_bam2" class="figure mx-auto d-block"></figure><p>We will convert the SAM file to BAM format using the
<code>samtools</code> program with the <code>view</code> command and
tell this command that the input is in SAM format (<code>-S</code>) and
to output BAM format (<code>-b</code>):</p>
<div class="codewrapper sourceCode" id="cb10">
<h3 class="code-label">BASH<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="sourceCode bash" tabindex="0"><code class="sourceCode bash"><span id="cb10-1"><a href="#cb10-1" tabindex="-1"></a><span class="ex">$</span> samtools view <span class="at">-S</span> <span class="at">-b</span> results/sam/SRR2584866.aligned.sam <span class="op">&gt;</span> results/bam/SRR2584866.aligned.bam</span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>[samopen] SAM header is present: 1 sequences.</code></pre>
</div>
</div>
</div>
<div class="section level4">
<h4 id="sort-bam-file-by-coordinates">Sort BAM file by coordinates<a class="anchor" aria-label="anchor" href="#sort-bam-file-by-coordinates"></a>
</h4>
<p>Next we sort the BAM file using the <code>sort</code> command from
<code>samtools</code>. <code>-o</code> tells the command where to write
the output.</p>
<div class="codewrapper sourceCode" id="cb12">
<h3 class="code-label">BASH<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="sourceCode bash" tabindex="0"><code class="sourceCode bash"><span id="cb12-1"><a href="#cb12-1" tabindex="-1"></a><span class="ex">$</span> samtools sort <span class="at">-o</span> results/bam/SRR2584866.aligned.sorted.bam results/bam/SRR2584866.aligned.bam </span></code></pre>
</div>
<p>Our files are pretty small, so we will not see this output. If you
run the workflow with larger files, you will see something like
this:</p>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>[bam_sort_core] merging from 2 files...</code></pre>
</div>
<p>SAM/BAM files can be sorted in multiple ways, e.g. by location of
alignment on the chromosome, by read name, etc. It is important to be
aware that different alignment tools will output differently sorted
SAM/BAM, and different downstream tools require differently sorted
alignment files as input.</p>
<p>You can use samtools to learn more about this bam file as well.</p>
<div class="codewrapper sourceCode" id="cb14">
<h3 class="code-label">BASH<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="sourceCode bash" tabindex="0"><code class="sourceCode bash"><span id="cb14-1"><a href="#cb14-1" tabindex="-1"></a><span class="ex">samtools</span> flagstat results/bam/SRR2584866.aligned.sorted.bam</span></code></pre>
</div>
<p>This will give you the following statistics about your sorted bam
file:</p>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>351169 + 0 in total (QC-passed reads + QC-failed reads)
0 + 0 secondary
1169 + 0 supplementary
0 + 0 duplicates
351103 + 0 mapped (99.98% : N/A)
350000 + 0 paired in sequencing
175000 + 0 read1
175000 + 0 read2
346688 + 0 properly paired (99.05% : N/A)
349876 + 0 with itself and mate mapped
58 + 0 singletons (0.02% : N/A)
0 + 0 with mate mapped to a different chr
0 + 0 with mate mapped to a different chr (mapQ&gt;=5)</code></pre>
</div>
</div>
<div class="section level3">
<h3 id="variant-calling">Variant calling<a class="anchor" aria-label="anchor" href="#variant-calling"></a>
</h3>
<p>A variant call is a conclusion that there is a nucleotide difference
vs. some reference at a given position in an individual genome or
transcriptome, often referred to as a Single Nucleotide Variant (SNV).
The call is usually accompanied by an estimate of variant frequency and
some measure of confidence. Similar to other steps in this workflow,
there are a number of tools available for variant calling. In this
workshop we will be using <code>bcftools</code>, but there are a few
things we need to do before actually calling the variants.</p>
<figure><img src="../fig/variant_calling_workflow.png" alt="workflow" class="figure mx-auto d-block"></figure><div class="section level4">
<h4 id="step-1-calculate-the-read-coverage-of-positions-in-the-genome">Step 1: Calculate the read coverage of positions in the genome<a class="anchor" aria-label="anchor" href="#step-1-calculate-the-read-coverage-of-positions-in-the-genome"></a>
</h4>
<p>Do the first pass on variant calling by counting read coverage with
<a href="https://samtools.github.io/bcftools/bcftools.html" class="external-link">bcftools</a>.
We will use the command <code>mpileup</code>. The flag <code>-O b</code>
tells bcftools to generate a bcf format output file, <code>-o</code>
specifies where to write the output file, and <code>-f</code> flags the
path to the reference genome:</p>
<div class="codewrapper sourceCode" id="cb16">
<h3 class="code-label">BASH<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="sourceCode bash" tabindex="0"><code class="sourceCode bash"><span id="cb16-1"><a href="#cb16-1" tabindex="-1"></a><span class="ex">$</span> bcftools mpileup <span class="at">-O</span> b <span class="at">-o</span> results/bcf/SRR2584866_raw.bcf <span class="dt">\</span></span>
<span id="cb16-2"><a href="#cb16-2" tabindex="-1"></a>-f data/ref_genome/ecoli_rel606.fasta results/bam/SRR2584866.aligned.sorted.bam </span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>[mpileup] 1 samples in 1 input files</code></pre>
</div>
<p>We have now generated a file with coverage information for every
base.</p>
</div>
<div class="section level4">
<h4 id="step-2-detect-the-single-nucleotide-variants-snvs">Step 2: Detect the single nucleotide variants (SNVs)<a class="anchor" aria-label="anchor" href="#step-2-detect-the-single-nucleotide-variants-snvs"></a>
</h4>
<p>Identify SNVs using bcftools <code>call</code>. We have to specify
ploidy with the flag <code>--ploidy</code>, which is one for the haploid
<em>E. coli</em>. <code>-m</code> allows for multiallelic and
rare-variant calling, <code>-v</code> tells the program to output
variant sites only (not every site in the genome), and <code>-o</code>
specifies where to write the output file:</p>
<div class="codewrapper sourceCode" id="cb18">
<h3 class="code-label">BASH<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="sourceCode bash" tabindex="0"><code class="sourceCode bash"><span id="cb18-1"><a href="#cb18-1" tabindex="-1"></a><span class="ex">$</span> bcftools call <span class="at">--ploidy</span> 1 <span class="at">-m</span> <span class="at">-v</span> <span class="at">-o</span> results/vcf/SRR2584866_variants.vcf results/bcf/SRR2584866_raw.bcf </span></code></pre>
</div>
</div>
<div class="section level4">
<h4 id="step-3-filter-and-report-the-snv-variants-in-variant-calling-format-vcf">Step 3: Filter and report the SNV variants in variant calling format
(VCF)<a class="anchor" aria-label="anchor" href="#step-3-filter-and-report-the-snv-variants-in-variant-calling-format-vcf"></a>
</h4>
<p>Filter the SNVs for the final output in VCF format, using
<code>vcfutils.pl</code>:</p>
<div class="codewrapper sourceCode" id="cb19">
<h3 class="code-label">BASH<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="sourceCode bash" tabindex="0"><code class="sourceCode bash"><span id="cb19-1"><a href="#cb19-1" tabindex="-1"></a><span class="ex">$</span> vcfutils.pl varFilter results/vcf/SRR2584866_variants.vcf  <span class="op">&gt;</span> results/vcf/SRR2584866_final_variants.vcf</span></code></pre>
</div>
<div id="filtering" class="callout">
<div class="callout-square">
<i class="callout-icon" data-feather="bell"></i>
</div>
<div id="filtering" class="callout-inner">
<h3 class="callout-title">Filtering</h3>
<div class="callout-content">
<p>The <code>vcfutils.pl varFilter</code> call filters out variants that
do not meet minimum quality default criteria, which can be changed
through its options. Using <code>bcftools</code> we can verify that the
quality of the variant call set has improved after this filtering step
by calculating the ratio of <a href="https://en.wikipedia.org/wiki/Transition_%28genetics%29" class="external-link">transitions(TS)</a>
to <a href="https://en.wikipedia.org/wiki/Transversion" class="external-link">transversions
(TV)</a> ratio (TS/TV), where transitions should be more likely to occur
than transversions:</p>
<div class="codewrapper sourceCode" id="cb20">
<h3 class="code-label">BASH<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="sourceCode bash" tabindex="0"><code class="sourceCode bash"><span id="cb20-1"><a href="#cb20-1" tabindex="-1"></a><span class="ex">$</span> bcftools stats results/vcf/SRR2584866_variants.vcf <span class="kw">|</span> <span class="fu">grep</span> TSTV</span>
<span id="cb20-2"><a href="#cb20-2" tabindex="-1"></a><span class="co"># TSTV, transitions/transversions:</span></span>
<span id="cb20-3"><a href="#cb20-3" tabindex="-1"></a><span class="co"># TSTV	[2]id	[3]ts	[4]tv	[5]ts/tv	[6]ts (1st ALT)	[7]tv (1st ALT)	[8]ts/tv (1st ALT)</span></span>
<span id="cb20-4"><a href="#cb20-4" tabindex="-1"></a><span class="ex">TSTV</span>	0	628	58	10.83	628	58	10.83</span>
<span id="cb20-5"><a href="#cb20-5" tabindex="-1"></a><span class="ex">$</span> bcftools stats results/vcf/SRR2584866_final_variants.vcf <span class="kw">|</span> <span class="fu">grep</span> TSTV</span>
<span id="cb20-6"><a href="#cb20-6" tabindex="-1"></a><span class="co"># TSTV, transitions/transversions:</span></span>
<span id="cb20-7"><a href="#cb20-7" tabindex="-1"></a><span class="co"># TSTV	[2]id	[3]ts	[4]tv	[5]ts/tv	[6]ts (1st ALT)	[7]tv (1st ALT)	[8]ts/tv (1st ALT)</span></span>
<span id="cb20-8"><a href="#cb20-8" tabindex="-1"></a><span class="ex">TSTV</span>	0	621	54	11.50	621	54	11.50</span></code></pre>
</div>
</div>
</div>
</div>
</div>
</div>
<div class="section level3">
<h3 id="explore-the-vcf-format">Explore the VCF format:<a class="anchor" aria-label="anchor" href="#explore-the-vcf-format"></a>
</h3>
<div class="codewrapper sourceCode" id="cb21">
<h3 class="code-label">BASH<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="sourceCode bash" tabindex="0"><code class="sourceCode bash"><span id="cb21-1"><a href="#cb21-1" tabindex="-1"></a><span class="ex">$</span> less <span class="at">-S</span> results/vcf/SRR2584866_final_variants.vcf</span></code></pre>
</div>
<p>You will see the header (which describes the format), the time and
date the file was created, the version of bcftools that was used, the
command line parameters used, and some additional information:</p>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code><span><span class="co">##fileformat=VCFv4.2</span></span>
<span><span class="co">##FILTER=&lt;ID=PASS,Description="All filters passed"&gt;</span></span>
<span><span class="co">##bcftoolsVersion=1.8+htslib-1.8</span></span>
<span><span class="co">##bcftoolsCommand=mpileup -O b -o results/bcf/SRR2584866_raw.bcf -f data/ref_genome/ecoli_rel606.fasta results/bam/SRR2584866.aligned.sorted.bam</span></span>
<span><span class="co">##reference=file://data/ref_genome/ecoli_rel606.fasta</span></span>
<span><span class="co">##contig=&lt;ID=CP000819.1,length=4629812&gt;</span></span>
<span><span class="co">##ALT=&lt;ID=*,Description="Represents allele(s) other than observed."&gt;</span></span>
<span><span class="co">##INFO=&lt;ID=INDEL,Number=0,Type=Flag,Description="Indicates that the variant is an INDEL."&gt;</span></span>
<span><span class="co">##INFO=&lt;ID=IDV,Number=1,Type=Integer,Description="Maximum number of reads supporting an indel"&gt;</span></span>
<span><span class="co">##INFO=&lt;ID=IMF,Number=1,Type=Float,Description="Maximum fraction of reads supporting an indel"&gt;</span></span>
<span><span class="co">##INFO=&lt;ID=DP,Number=1,Type=Integer,Description="Raw read depth"&gt;</span></span>
<span><span class="co">##INFO=&lt;ID=VDB,Number=1,Type=Float,Description="Variant Distance Bias for filtering splice-site artefacts in RNA-seq data (bigger is better)",Version=</span></span>
<span><span class="co">##INFO=&lt;ID=RPB,Number=1,Type=Float,Description="Mann-Whitney U test of Read Position Bias (bigger is better)"&gt;</span></span>
<span><span class="co">##INFO=&lt;ID=MQB,Number=1,Type=Float,Description="Mann-Whitney U test of Mapping Quality Bias (bigger is better)"&gt;</span></span>
<span><span class="co">##INFO=&lt;ID=BQB,Number=1,Type=Float,Description="Mann-Whitney U test of Base Quality Bias (bigger is better)"&gt;</span></span>
<span><span class="co">##INFO=&lt;ID=MQSB,Number=1,Type=Float,Description="Mann-Whitney U test of Mapping Quality vs Strand Bias (bigger is better)"&gt;</span></span>
<span><span class="co">##INFO=&lt;ID=SGB,Number=1,Type=Float,Description="Segregation based metric."&gt;</span></span>
<span><span class="co">##INFO=&lt;ID=MQ0F,Number=1,Type=Float,Description="Fraction of MQ0 reads (smaller is better)"&gt;</span></span>
<span><span class="co">##FORMAT=&lt;ID=PL,Number=G,Type=Integer,Description="List of Phred-scaled genotype likelihoods"&gt;</span></span>
<span><span class="co">##FORMAT=&lt;ID=GT,Number=1,Type=String,Description="Genotype"&gt;</span></span>
<span><span class="co">##INFO=&lt;ID=ICB,Number=1,Type=Float,Description="Inbreeding Coefficient Binomial test (bigger is better)"&gt;</span></span>
<span><span class="co">##INFO=&lt;ID=HOB,Number=1,Type=Float,Description="Bias in the number of HOMs number (smaller is better)"&gt;</span></span>
<span><span class="co">##INFO=&lt;ID=AC,Number=A,Type=Integer,Description="Allele count in genotypes for each ALT allele, in the same order as listed"&gt;</span></span>
<span><span class="co">##INFO=&lt;ID=AN,Number=1,Type=Integer,Description="Total number of alleles in called genotypes"&gt;</span></span>
<span><span class="co">##INFO=&lt;ID=DP4,Number=4,Type=Integer,Description="Number of high-quality ref-forward , ref-reverse, alt-forward and alt-reverse bases"&gt;</span></span>
<span><span class="co">##INFO=&lt;ID=MQ,Number=1,Type=Integer,Description="Average mapping quality"&gt;</span></span>
<span><span class="co">##bcftools_callVersion=1.8+htslib-1.8</span></span>
<span><span class="co">##bcftools_callCommand=call --ploidy 1 -m -v -o results/bcf/SRR2584866_variants.vcf results/bcf/SRR2584866_raw.bcf; Date=Tue Oct  9 18:48:10 2018</span></span></code></pre>
</div>
<p>Followed by information on each of the variations observed:</p>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>#CHROM  POS     ID      REF     ALT     QUAL    FILTER  INFO    FORMAT  results/bam/SRR2584866.aligned.sorted.bam
CP000819.1      1521    .       C       T       207     .       DP=9;VDB=0.993024;SGB=-0.662043;MQSB=0.974597;MQ0F=0;AC=1;AN=1;DP4=0,0,4,5;MQ=60
CP000819.1      1612    .       A       G       225     .       DP=13;VDB=0.52194;SGB=-0.676189;MQSB=0.950952;MQ0F=0;AC=1;AN=1;DP4=0,0,6,5;MQ=60
CP000819.1      9092    .       A       G       225     .       DP=14;VDB=0.717543;SGB=-0.670168;MQSB=0.916482;MQ0F=0;AC=1;AN=1;DP4=0,0,7,3;MQ=60
CP000819.1      9972    .       T       G       214     .       DP=10;VDB=0.022095;SGB=-0.670168;MQSB=1;MQ0F=0;AC=1;AN=1;DP4=0,0,2,8;MQ=60      GT:PL
CP000819.1      10563   .       G       A       225     .       DP=11;VDB=0.958658;SGB=-0.670168;MQSB=0.952347;MQ0F=0;AC=1;AN=1;DP4=0,0,5,5;MQ=60
CP000819.1      22257   .       C       T       127     .       DP=5;VDB=0.0765947;SGB=-0.590765;MQSB=1;MQ0F=0;AC=1;AN=1;DP4=0,0,2,3;MQ=60      GT:PL
CP000819.1      38971   .       A       G       225     .       DP=14;VDB=0.872139;SGB=-0.680642;MQSB=1;MQ0F=0;AC=1;AN=1;DP4=0,0,4,8;MQ=60      GT:PL
CP000819.1      42306   .       A       G       225     .       DP=15;VDB=0.969686;SGB=-0.686358;MQSB=1;MQ0F=0;AC=1;AN=1;DP4=0,0,5,9;MQ=60      GT:PL
CP000819.1      45277   .       A       G       225     .       DP=15;VDB=0.470998;SGB=-0.680642;MQSB=0.95494;MQ0F=0;AC=1;AN=1;DP4=0,0,7,5;MQ=60
CP000819.1      56613   .       C       G       183     .       DP=12;VDB=0.879703;SGB=-0.676189;MQSB=1;MQ0F=0;AC=1;AN=1;DP4=0,0,8,3;MQ=60      GT:PL
CP000819.1      62118   .       A       G       225     .       DP=19;VDB=0.414981;SGB=-0.691153;MQSB=0.906029;MQ0F=0;AC=1;AN=1;DP4=0,0,8,10;MQ=59
CP000819.1      64042   .       G       A       225     .       DP=18;VDB=0.451328;SGB=-0.689466;MQSB=1;MQ0F=0;AC=1;AN=1;DP4=0,0,7,9;MQ=60      GT:PL</code></pre>
</div>
<p>This is a lot of information, so let’s take some time to make sure we
understand our output.</p>
<p>The first few columns represent the information we have about a
predicted variation.</p>
<table class="table">
<colgroup>
<col width="2%">
<col width="97%">
</colgroup>
<thead><tr class="header">
<th>column</th>
<th>info</th>
</tr></thead>
<tbody>
<tr class="odd">
<td>CHROM</td>
<td>contig location where the variation occurs</td>
</tr>
<tr class="even">
<td>POS</td>
<td>position within the contig where the variation occurs</td>
</tr>
<tr class="odd">
<td>ID</td>
<td>a <code>.</code> until we add annotation information</td>
</tr>
<tr class="even">
<td>REF</td>
<td>reference genotype (forward strand)</td>
</tr>
<tr class="odd">
<td>ALT</td>
<td>sample genotype (forward strand)</td>
</tr>
<tr class="even">
<td>QUAL</td>
<td>Phred-scaled probability that the observed variant exists at this
site (higher is better)</td>
</tr>
<tr class="odd">
<td>FILTER</td>
<td>a <code>.</code> if no quality filters have been applied, PASS if a
filter is passed, or the name of the filters this variant failed</td>
</tr>
</tbody>
</table>
<p>In an ideal world, the information in the <code>QUAL</code> column
would be all we needed to filter out bad variant calls. However, in
reality we need to filter on multiple other metrics.</p>
<p>The last two columns contain the genotypes and can be tricky to
decode.</p>
<table class="table">
<colgroup>
<col width="2%">
<col width="97%">
</colgroup>
<thead><tr class="header">
<th>column</th>
<th>info</th>
</tr></thead>
<tbody>
<tr class="odd">
<td>FORMAT</td>
<td>lists in order the metrics presented in the final column</td>
</tr>
<tr class="even">
<td>results</td>
<td>lists the values associated with those metrics in order</td>
</tr>
</tbody>
</table>
<p>For our file, the metrics presented are GT:PL:GQ.</p>
<table class="table">
<colgroup>
<col width="2%">
<col width="97%">
</colgroup>
<thead><tr class="header">
<th>metric</th>
<th>definition</th>
</tr></thead>
<tbody>
<tr class="odd">
<td>AD, DP</td>
<td>the depth per allele by sample and coverage</td>
</tr>
<tr class="even">
<td>GT</td>
<td>the genotype for the sample at this loci. For a diploid organism,
the GT field indicates the two alleles carried by the sample, encoded by
a 0 for the REF allele, 1 for the first ALT allele, 2 for the second ALT
allele, etc. A 0/0 means homozygous reference, 0/1 is heterozygous, and
1/1 is homozygous for the alternate allele.</td>
</tr>
<tr class="odd">
<td>PL</td>
<td>the likelihoods of the given genotypes</td>
</tr>
<tr class="even">
<td>GQ</td>
<td>the Phred-scaled confidence for the genotype</td>
</tr>
</tbody>
</table>
<p>The Broad Institute’s <a href="https://www.broadinstitute.org/gatk/guide/article?id=1268" class="external-link">VCF
guide</a> is an excellent place to learn more about the VCF file
format.</p>
<div id="exercise-1" class="callout challenge">
<div class="callout-square">
<i class="callout-icon" data-feather="zap"></i>
</div>
<div id="exercise-1" class="callout-inner">
<h3 class="callout-title">Exercise</h3>
<div class="callout-content">
<p>Use the <code>grep</code> and <code>wc</code> commands you have
learned to assess how many variants are in the vcf file.</p>
</div>
</div>
</div>
<div id="accordionSolution2" class="accordion challenge-accordion accordion-flush">
<div class="accordion-item">
<button class="accordion-button solution-button collapsed" type="button" data-bs-toggle="collapse" data-bs-target="#collapseSolution2" aria-expanded="false" aria-controls="collapseSolution2">
  <h4 class="accordion-header" id="headingSolution2"> Show me the solution </h4>
</button>
<div id="collapseSolution2" class="accordion-collapse collapse" aria-labelledby="headingSolution2" data-bs-parent="#accordionSolution2">
<div class="accordion-body">
<div class="codewrapper sourceCode" id="cb24">
<h3 class="code-label">BASH<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="sourceCode bash" tabindex="0"><code class="sourceCode bash"><span id="cb24-1"><a href="#cb24-1" tabindex="-1"></a><span class="ex">$</span> grep <span class="at">-v</span> <span class="st">"#"</span> results/vcf/SRR2584866_final_variants.vcf <span class="kw">|</span> <span class="fu">wc</span> <span class="at">-l</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code><span><span class="fl">766</span></span></code></pre>
</div>
<p>There are 766 variants in this file.</p>
</div>
</div>
</div>
</div>
</div>
<div class="section level3">
<h3 id="assess-the-alignment-visualization---optional-step">Assess the alignment (visualization) - optional step<a class="anchor" aria-label="anchor" href="#assess-the-alignment-visualization---optional-step"></a>
</h3>
<p>It is often instructive to look at your data in a genome browser.
Visualization will allow you to get a “feel” for the data, as well as
detecting abnormalities and problems. Also, exploring the data in such a
way may give you ideas for further analyses. As such, visualization
tools are useful for exploratory analysis. In this lesson we will
describe two different tools for visualization: a light-weight
command-line based one and the Broad Institute’s Integrative Genomics
Viewer (IGV) which requires software installation and transfer of
files.</p>
<p>In order for us to visualize the alignment files, we will need to
index the BAM file using <code>samtools</code>:</p>
<div class="codewrapper sourceCode" id="cb26">
<h3 class="code-label">BASH<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="sourceCode bash" tabindex="0"><code class="sourceCode bash"><span id="cb26-1"><a href="#cb26-1" tabindex="-1"></a><span class="ex">$</span> samtools index results/bam/SRR2584866.aligned.sorted.bam</span></code></pre>
</div>
<div class="section level4">
<h4 id="viewing-with-tview">Viewing with <code>tview</code>
<a class="anchor" aria-label="anchor" href="#viewing-with-tview"></a>
</h4>
<p><a href="https://www.htslib.org/" class="external-link">Samtools</a> implements a very
simple text alignment viewer based on the GNU <code>ncurses</code>
library, called <code>tview</code>. This alignment viewer works with
short indels and shows <a href="https://maq.sourceforge.net/" class="external-link">MAQ</a>
consensus. It uses different colors to display mapping quality or base
quality, subjected to users’ choice. Samtools viewer is known to work
with a 130 GB alignment swiftly. Due to its text interface, displaying
alignments over network is also very fast.</p>
<p>In order to visualize our mapped reads, we use <code>tview</code>,
giving it the sorted bam file and the reference file:</p>
<div class="codewrapper sourceCode" id="cb27">
<h3 class="code-label">BASH<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="sourceCode bash" tabindex="0"><code class="sourceCode bash"><span id="cb27-1"><a href="#cb27-1" tabindex="-1"></a><span class="ex">$</span> samtools tview results/bam/SRR2584866.aligned.sorted.bam data/ref_genome/ecoli_rel606.fasta</span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>1         11        21        31        41        51        61        71        81        91        101       111       121
AGCTTTTCATTCTGACTGCAACGGGCAATATGTCTCTGTGTGGATTAAAAAAAGAGTGTCTGATAGCAGCTTCTGAACTGGTTACCTGCCGTGAGTAAATTAAAATTTTATTGACTTAGGTCACTAAATAC
..................................................................................................................................
,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,, ..................N................. ,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,........................
,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,, ..................N................. ,,,,,,,,,,,,,,,,,,,,,,,,,,,.............................
...................................,g,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,  ....................................   ................
,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,....................................   ....................................      ,,,,,,,,,,
,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,  ....................................  ,,a,,,,,,,,,,,,,,,,,,,,,,,,,,,,,     .......
,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,, .............................  ,,,,,,,,,,,,,,,,,g,,,,,    ,,,,,,,,,,,,,,,,,,,,,,,,,,,,
,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,  ...........................T.......   ,,,,,,,,,,,,,,,,,,,,,,,c,          ......
......................... ................................   ,g,,,,,,,,,,,,,,,,,,,      ...........................
,,,,,,,,,,,,,,,,,,,,, ,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,, ,,,,,,,,,,,,,,,,,,,,,,,,,,,       ..........................
,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,   ................................T..  ..............................   ,,,,,,
...........................       ,,,,,,g,,,,,,,,,,,,,,,,,   ....................................         ,,,,,,
,,,,,,,,,,,,,,,,,,,,,,,,,, ....................................  ...................................        ....
....................................  ........................  ,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,      ....
,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,   ,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,  ,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,
........................            .................................. .............................     ....
,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,   ....................................        ..........................
...............................       ,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,, ....................................
...................................  ,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,  ,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,
,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,, ,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,  ..................................
.................................... ,,,,,,,,,,,,,,,,,,a,,,,,,,,,,,,,,,,,        ,,,,,,,,,,,,,,,,,,,,,,,,,
,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,  ............................ ,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,</code></pre>
</div>
<p>The first line of output shows the genome coordinates in our
reference genome. The second line shows the reference genome sequence.
The third line shows the consensus sequence determined from the sequence
reads. A <code>.</code> indicates a match to the reference sequence, so
we can see that the consensus from our sample matches the reference in
most locations. That is good! If that was not the case, we should
probably reconsider our choice of reference.</p>
<p>Below the horizontal line, we can see all of the reads in our sample
aligned with the reference genome. Only positions where the called base
differs from the reference are shown. You can use the arrow keys on your
keyboard to scroll or type <code>?</code> for a help menu. To navigate
to a specific position, type <code>g</code>. A dialogue box will appear.
In this box, type the name of the “chromosome” followed by a colon and
the position of the variant you would like to view (e.g. for this
sample, type <code>CP000819.1:50</code> to view the 50th base. Type
<code>Ctrl^C</code> or <code>q</code> to exit <code>tview</code>.</p>
<div id="exercise-2" class="callout challenge">
<div class="callout-square">
<i class="callout-icon" data-feather="zap"></i>
</div>
<div id="exercise-2" class="callout-inner">
<h3 class="callout-title">Exercise</h3>
<div class="callout-content">
<p>Visualize the alignment of the reads for our <code>SRR2584866</code>
sample. What variant is present at position 4377265? What is the
canonical nucleotide in that position?</p>
</div>
</div>
</div>
<div id="accordionSolution3" class="accordion challenge-accordion accordion-flush">
<div class="accordion-item">
<button class="accordion-button solution-button collapsed" type="button" data-bs-toggle="collapse" data-bs-target="#collapseSolution3" aria-expanded="false" aria-controls="collapseSolution3">
  <h4 class="accordion-header" id="headingSolution3"> Show me the solution </h4>
</button>
<div id="collapseSolution3" class="accordion-collapse collapse" aria-labelledby="headingSolution3" data-bs-parent="#accordionSolution3">
<div class="accordion-body">
<div class="codewrapper sourceCode" id="cb29">
<h3 class="code-label">BASH<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="sourceCode bash" tabindex="0"><code class="sourceCode bash"><span id="cb29-1"><a href="#cb29-1" tabindex="-1"></a><span class="ex">$</span> samtools tview ~/dc_workshop/results/bam/SRR2584866.aligned.sorted.bam ~/dc_workshop/data/ref_genome/ecoli_rel606.fasta</span></code></pre>
</div>
<p>Then type <code>g</code>. In the dialogue box, type
<code>CP000819.1:4377265</code>. <code>G</code> is the variant.
<code>A</code> is canonical. This variant possibly changes the phenotype
of this sample to hypermutable. It occurs in the gene <em>mutL</em>,
which controls DNA mismatch repair.</p>
</div>
</div>
</div>
</div>
</div>
<div class="section level4">
<h4 id="viewing-with-igv">Viewing with IGV<a class="anchor" aria-label="anchor" href="#viewing-with-igv"></a>
</h4>
<p><a href="https://www.broadinstitute.org/igv/" class="external-link">IGV</a> is a
stand-alone browser, which has the advantage of being installed locally
and providing fast access. Web-based genome browsers, like <a href="https://www.ensembl.org/index.html" class="external-link">Ensembl</a> or the <a href="https://genome.ucsc.edu/" class="external-link">UCSC browser</a>, are slower, but
provide more functionality. They not only allow for more polished and
flexible visualization, but also provide easy access to a wealth of
annotations and external data sources. This makes it straightforward to
relate your data with information about repeat regions, known genes,
epigenetic features or areas of cross-species conservation, to name just
a few.</p>
<p>In order to use IGV, we will need to transfer some files to our local
machine. We know how to do this with <code>scp</code>. Open a new tab in
your terminal window and create a new folder. We will put this folder on
our Desktop for demonstration purposes, but in general you should avoid
proliferating folders and files on your Desktop and instead organize
files within a directory structure like we have been using in our
<code>dc_workshop</code> directory.</p>
<div class="codewrapper sourceCode" id="cb30">
<h3 class="code-label">BASH<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="sourceCode bash" tabindex="0"><code class="sourceCode bash"><span id="cb30-1"><a href="#cb30-1" tabindex="-1"></a><span class="ex">$</span> mkdir ~/Desktop/files_for_igv</span>
<span id="cb30-2"><a href="#cb30-2" tabindex="-1"></a><span class="ex">$</span> cd ~/Desktop/files_for_igv</span></code></pre>
</div>
<p>Now we will transfer our files to that new directory. Remember to
replace the text between the <code>@</code> and the <code>:</code> with
your AWS instance number. The commands to <code>scp</code> always go in
the terminal window that is connected to your local computer (not your
AWS instance).</p>
<div class="codewrapper sourceCode" id="cb31">
<h3 class="code-label">BASH<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="sourceCode bash" tabindex="0"><code class="sourceCode bash"><span id="cb31-1"><a href="#cb31-1" tabindex="-1"></a><span class="ex">$</span> scp dcuser@ec2-34-203-203-131.compute-1.amazonaws.com:~/dc_workshop/results/bam/SRR2584866.aligned.sorted.bam ~/Desktop/files_for_igv</span>
<span id="cb31-2"><a href="#cb31-2" tabindex="-1"></a><span class="ex">$</span> scp dcuser@ec2-34-203-203-131.compute-1.amazonaws.com:~/dc_workshop/results/bam/SRR2584866.aligned.sorted.bam.bai ~/Desktop/files_for_igv</span>
<span id="cb31-3"><a href="#cb31-3" tabindex="-1"></a><span class="ex">$</span> scp dcuser@ec2-34-203-203-131.compute-1.amazonaws.com:~/dc_workshop/data/ref_genome/ecoli_rel606.fasta ~/Desktop/files_for_igv</span>
<span id="cb31-4"><a href="#cb31-4" tabindex="-1"></a><span class="ex">$</span> scp dcuser@ec2-34-203-203-131.compute-1.amazonaws.com:~/dc_workshop/results/vcf/SRR2584866_final_variants.vcf ~/Desktop/files_for_igv</span></code></pre>
</div>
<p>You will need to type the password for your AWS instance each time
you call <code>scp</code>.</p>
<p>Next, we need to open the IGV software. If you have not done so
already, you can download IGV from the <a href="https://www.broadinstitute.org/software/igv/download" class="external-link">Broad
Institute’s software page</a>, double-click the <code>.zip</code> file
to unzip it, and then drag the program into your Applications
folder.</p>
<ol style="list-style-type: decimal">
<li>Open IGV.</li>
<li>Load our reference genome file (<code>ecoli_rel606.fasta</code>)
into IGV using the <strong>“Load Genomes from File…”</strong> option
under the <strong>“Genomes”</strong> pull-down menu.</li>
<li>Load our BAM file (<code>SRR2584866.aligned.sorted.bam</code>) using
the <strong>“Load from File…”</strong> option under the
<strong>“File”</strong> pull-down menu.</li>
<li>Do the same with our VCF file
(<code>SRR2584866_final_variants.vcf</code>).</li>
</ol>
<p>Your IGV browser should look like the screenshot below:</p>
<figure><img src="../fig/igv-screenshot.png" alt="IGV" class="figure mx-auto d-block"></figure><p>There should be two tracks: one coresponding to our BAM file and the
other for our VCF file.</p>
<p>In the <strong>VCF track</strong>, each bar across the top of the
plot shows the allele fraction for a single locus. The second bar shows
the genotypes for each locus in each <em>sample</em>. We only have one
sample called here, so we only see a single line. Dark blue =
heterozygous, Cyan = homozygous variant, Grey = reference. Filtered
entries are transparent.</p>
<p>Zoom in to inspect variants you see in your filtered VCF file to
become more familiar with IGV. See how quality information corresponds
to alignment information at those loci. Use <a href="https://software.broadinstitute.org/software/igv/AlignmentData" class="external-link">this
website</a> and the links therein to understand how IGV colors the
alignments.</p>
<p>Now that we have run through our workflow for a single sample, we
want to repeat this workflow for our other five samples. However, we do
not want to type each of these individual steps again five more times.
That would be very time consuming and error-prone, and would become
impossible as we gathered more and more samples. Luckily, we already
know the tools we need to use to automate this workflow and run it on as
many files as we want using a single line of code. Those tools are:
wildcards, for loops, and bash scripts. We will use all three in the
next lesson.</p>
<div id="installing-software" class="callout">
<div class="callout-square">
<i class="callout-icon" data-feather="bell"></i>
</div>
<div id="installing-software" class="callout-inner">
<h3 class="callout-title">Installing software</h3>
<div class="callout-content">
<p>It is worth noting that all of the software we are using for this
workshop has been pre-installed on our remote computer. This saves us a
lot of time - installing software can be a time-consuming and
frustrating task - however, this does mean that you will not be able to
walk out the door and start doing these analyses on your own computer.
You will need to install the software first. Look at the <a href="https://datacarpentry.org/genomics-workshop/index.html#setup" class="external-link">setup
instructions</a> for more information on installing these software
packages.</p>
</div>
</div>
</div>
<div id="bwa-alignment-options" class="callout">
<div class="callout-square">
<i class="callout-icon" data-feather="bell"></i>
</div>
<div id="bwa-alignment-options" class="callout-inner">
<h3 class="callout-title">BWA alignment options</h3>
<div class="callout-content">
<p>BWA consists of three algorithms: BWA-backtrack, BWA-SW and BWA-MEM.
The first algorithm is designed for Illumina sequence reads up to 100bp,
while the other two are for sequences ranging from 70bp to 1Mbp. BWA-MEM
and BWA-SW share similar features such as long-read support and split
alignment, but BWA-MEM, which is the latest, is generally recommended
for high-quality queries as it is faster and more accurate.</p>
</div>
</div>
</div>
<div id="keypoints1" class="callout keypoints">
<div class="callout-square">
<i class="callout-icon" data-feather="key"></i>
</div>
<div class="callout-inner">
<h3 class="callout-title">Key Points</h3>
<div class="callout-content">
<ul>
<li>Bioinformatic command line tools are collections of commands that
can be used to carry out bioinformatic analyses.</li>
<li>To use most powerful bioinformatic tools, you will need to use the
command line.</li>
<li>There are many different file formats for storing genomics data. It
is important to understand what type of information is contained in each
file, and how it was derived.</li>
</ul>
</div>
</div>
</div>
</div>
</div>
</section></section><section id="aio-05-automation"><p>Content from <a href="05-automation.html">Automating a Variant Calling Workflow</a></p>
<hr>
<p>Last updated on 2024-07-08 |

        <a href="https://github.com/datacarpentry/wrangling-genomics/edit/main/episodes/05-automation.md" class="external-link">Edit this page <i aria-hidden="true" data-feather="edit"></i></a></p>
<p>Estimated time: <i aria-hidden="true" data-feather="clock"></i> 45 minutes</p>
<div class="text-end">
          <button role="button" aria-pressed="false" tabindex="0" id="expand-code" class="pull-right" data-expand="Expand All Solutions " data-collapse="Collapse All Solutions "> Expand All Solutions <i aria-hidden="true" data-feather="plus"></i></button>
        </div>
<div class="overview card">
<h2 class="card-header">Overview</h2>
<div class="row g-0">
<div class="col-md-4">
<div class="card-body">
<div class="inner">
<h3 class="card-title">Questions</h3>
<ul>
<li>How can I make my workflow more efficient and less error-prone?</li>
</ul>
</div>
</div>
</div>
<div class="col-md-8">
<div class="card-body">
<div class="inner bordered">
<h3 class="card-title">Objectives</h3>
<ul>
<li>Write a shell script with multiple variables.</li>
<li>Incorporate a <code>for</code> loop into a shell script.</li>
</ul>
</div>
</div>
</div>
</div>
</div>
<section><h2 class="section-heading" id="what-is-a-shell-script">What is a shell script?<a class="anchor" aria-label="anchor" href="#what-is-a-shell-script"></a>
</h2>
<hr class="half-width">
<p>You wrote a simple shell script in a <a href="https://www.datacarpentry.org/shell-genomics/05-writing-scripts" class="external-link">previous
lesson</a> that we used to extract bad reads from our FASTQ files and
put them into a new file.</p>
<p>Here is the script you wrote:</p>
<div class="codewrapper sourceCode" id="cb1">
<h3 class="code-label">BASH<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="sourceCode bash" tabindex="0"><code class="sourceCode bash"><span id="cb1-1"><a href="#cb1-1" tabindex="-1"></a><span class="fu">grep</span> <span class="at">-B1</span> <span class="at">-A2</span> NNNNNNNNNN <span class="pp">*</span>.fastq <span class="op">&gt;</span> scripted_bad_reads.txt</span>
<span id="cb1-2"><a href="#cb1-2" tabindex="-1"></a></span>
<span id="cb1-3"><a href="#cb1-3" tabindex="-1"></a><span class="bu">echo</span> <span class="st">"Script finished!"</span></span></code></pre>
</div>
<p>That script was only two lines long, but shell scripts can be much
more complicated than that and can be used to perform a large number of
operations on one or many files. This saves you the effort of having to
type each of those commands over for each of your data files and makes
your work less error-prone and more reproducible. For example, the
variant calling workflow we just carried out had about eight steps where
we had to type a command into our terminal. Most of these commands were
pretty long. If we wanted to do this for all six of our data files, that
would be forty-eight steps. If we had 50 samples (a more realistic
number), it would be 400 steps! You can see why we want to automate
this.</p>
<p>We have also used <code>for</code> loops in previous lessons to
iterate one or two commands over multiple input files. In these
<code>for</code> loops, the filename was defined as a variable in the
<code>for</code> statement, which enabled you to run the loop on
multiple files. We will be using variable assignments like this in our
new shell scripts.</p>
<p>Here is the <code>for</code> loop you wrote for unzipping
<code>.zip</code> files:</p>
<div class="codewrapper sourceCode" id="cb2">
<h3 class="code-label">BASH<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="sourceCode bash" tabindex="0"><code class="sourceCode bash"><span id="cb2-1"><a href="#cb2-1" tabindex="-1"></a><span class="ex">$</span> for filename in <span class="pp">*</span>.zip</span>
<span id="cb2-2"><a href="#cb2-2" tabindex="-1"></a><span class="op">&gt;</span> do</span>
<span id="cb2-3"><a href="#cb2-3" tabindex="-1"></a><span class="op">&gt;</span> unzip <span class="va">$filename</span></span>
<span id="cb2-4"><a href="#cb2-4" tabindex="-1"></a><span class="op">&gt;</span> done</span></code></pre>
</div>
<p>And here is the one you wrote for running Trimmomatic on all of our
<code>.fastq</code> sample files:</p>
<div class="codewrapper sourceCode" id="cb3">
<h3 class="code-label">BASH<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="sourceCode bash" tabindex="0"><code class="sourceCode bash"><span id="cb3-1"><a href="#cb3-1" tabindex="-1"></a><span class="ex">$</span> for infile in <span class="pp">*</span>_1.fastq.gz</span>
<span id="cb3-2"><a href="#cb3-2" tabindex="-1"></a><span class="op">&gt;</span> do</span>
<span id="cb3-3"><a href="#cb3-3" tabindex="-1"></a><span class="op">&gt;</span>   base=<span class="va">$(</span><span class="fu">basename</span> <span class="va">${infile}</span> _1.fastq.gz<span class="va">)</span></span>
<span id="cb3-4"><a href="#cb3-4" tabindex="-1"></a><span class="op">&gt;</span>   trimmomatic <span class="ex">PE</span> <span class="va">${infile}</span> <span class="va">${base}</span>_2.fastq.gz <span class="dt">\</span></span>
<span id="cb3-5"><a href="#cb3-5" tabindex="-1"></a><span class="op">&gt;</span>                <span class="va">${base}</span>_1.trim.fastq.gz <span class="va">${base}</span>_1un.trim.fastq.gz <span class="dt">\</span></span>
<span id="cb3-6"><a href="#cb3-6" tabindex="-1"></a><span class="op">&gt;</span>                <span class="va">${base}</span>_2.trim.fastq.gz <span class="va">${base}</span>_2un.trim.fastq.gz <span class="dt">\</span></span>
<span id="cb3-7"><a href="#cb3-7" tabindex="-1"></a><span class="op">&gt;</span>                SLIDINGWINDOW:4:20 MINLEN:25 ILLUMINACLIP:NexteraPE-PE.fa:2:40:15 </span>
<span id="cb3-8"><a href="#cb3-8" tabindex="-1"></a><span class="op">&gt;</span> done</span></code></pre>
</div>
<p>Notice that in this <code>for</code> loop, we used two variables,
<code>infile</code>, which was defined in the <code>for</code>
statement, and <code>base</code>, which was created from the filename
during each iteration of the loop.</p>
<div id="creating-variables" class="callout">
<div class="callout-square">
<i class="callout-icon" data-feather="bell"></i>
</div>
<div id="creating-variables" class="callout-inner">
<h3 class="callout-title">Creating variables</h3>
<div class="callout-content">
<p>Within the Bash shell you can create variables at any time (as we did
above, and during the ‘for’ loop lesson). Assign any name and the value
using the assignment operator: ‘=’. You can check the current definition
of your variable by typing into your script: echo $variable_name.</p>
</div>
</div>
</div>
<p>In this lesson, we will use two shell scripts to automate the variant
calling analysis: one for FastQC analysis (including creating our
summary file), and a second for the remaining variant calling. To write
a script to run our FastQC analysis, we will take each of the commands
we entered to run FastQC and process the output files and put them into
a single file with a <code>.sh</code> extension. The <code>.sh</code> is
not essential, but serves as a reminder to ourselves and to the computer
that this is a shell script.</p>
</section><section><h2 class="section-heading" id="analyzing-quality-with-fastqc">Analyzing quality with FastQC<a class="anchor" aria-label="anchor" href="#analyzing-quality-with-fastqc"></a>
</h2>
<hr class="half-width">
<p>We will use the command <code>touch</code> to create a new file where
we will write our shell script. We will create this script in a new
directory called <code>scripts/</code>. Previously, we used
<code>nano</code> to create and open a new file. The command
<code>touch</code> allows us to create a new file without opening that
file.</p>
<div class="codewrapper sourceCode" id="cb4">
<h3 class="code-label">BASH<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="sourceCode bash" tabindex="0"><code class="sourceCode bash"><span id="cb4-1"><a href="#cb4-1" tabindex="-1"></a><span class="ex">$</span> mkdir <span class="at">-p</span> ~/dc_workshop/scripts</span>
<span id="cb4-2"><a href="#cb4-2" tabindex="-1"></a><span class="ex">$</span> cd ~/dc_workshop/scripts</span>
<span id="cb4-3"><a href="#cb4-3" tabindex="-1"></a><span class="ex">$</span> touch read_qc.sh</span>
<span id="cb4-4"><a href="#cb4-4" tabindex="-1"></a><span class="ex">$</span> ls </span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code><span><span class="va">read_qc.sh</span></span></code></pre>
</div>
<p>We now have an empty file called <code>read_qc.sh</code> in our
<code>scripts/</code> directory. We will now open this file in
<code>nano</code> and start building our script.</p>
<div class="codewrapper sourceCode" id="cb6">
<h3 class="code-label">BASH<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="sourceCode bash" tabindex="0"><code class="sourceCode bash"><span id="cb6-1"><a href="#cb6-1" tabindex="-1"></a><span class="ex">$</span> nano read_qc.sh</span></code></pre>
</div>
<p><strong>Enter the following pieces of code into your shell script
(not into your terminal prompt).</strong></p>
<p>Our first line will ensure that our script will exit if an error
occurs, and is a good idea to include at the beginning of your scripts.
The second line will move us into the <code>untrimmed_fastq/</code>
directory when we run our script.</p>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>set -e
cd ~/dc_workshop/data/untrimmed_fastq/</code></pre>
</div>
<p>These next two lines will give us a status message to tell us that we
are currently running FastQC, then will run FastQC on all of the files
in our current directory with a <code>.fastq</code> extension.</p>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>echo "Running FastQC ..."
fastqc *.fastq*</code></pre>
</div>
<p>Our next line will create a new directory to hold our FastQC output
files. Here we are using the <code>-p</code> option for
<code>mkdir</code> again. It is a good idea to use this option in your
shell scripts to avoid running into errors if you do not have the
directory structure you think you do.</p>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>mkdir -p ~/dc_workshop/results/fastqc_untrimmed_reads</code></pre>
</div>
<p>Our next three lines first give us a status message to tell us we are
saving the results from FastQC, then moves all of the files with a
<code>.zip</code> or a <code>.html</code> extension to the directory we
just created for storing our FastQC results.</p>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>echo "Saving FastQC results..."
mv *.zip ~/dc_workshop/results/fastqc_untrimmed_reads/
mv *.html ~/dc_workshop/results/fastqc_untrimmed_reads/</code></pre>
</div>
<p>The next line moves us to the results directory where we have stored
our output.</p>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>cd ~/dc_workshop/results/fastqc_untrimmed_reads/</code></pre>
</div>
<p>The next five lines should look very familiar. First we give
ourselves a status message to tell us that we are unzipping our ZIP
files. Then we run our for loop to unzip all of the <code>.zip</code>
files in this directory.</p>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>echo "Unzipping..."
for filename in *.zip
    do
    unzip $filename
    done</code></pre>
</div>
<p>Next we concatenate all of our summary files into a single output
file, with a status message to remind ourselves that this is what we are
doing.</p>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>echo "Saving summary..."
mkdir -p ~/dc_workshops/docs
cat */summary.txt &gt; ~/dc_workshop/docs/fastqc_summaries.txt</code></pre>
</div>
<div id="using-echo-statements" class="callout">
<div class="callout-square">
<i class="callout-icon" data-feather="bell"></i>
</div>
<div id="using-echo-statements" class="callout-inner">
<h3 class="callout-title">Using <code>echo</code> statements</h3>
<div class="callout-content">
<p>We have used <code>echo</code> statements to add progress statements
to our script. Our script will print these statements as it is running
and therefore we will be able to see how far our script has
progressed.</p>
</div>
</div>
</div>
<p>Your full shell script should now look like this:</p>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>set -e
cd ~/dc_workshop/data/untrimmed_fastq/

echo "Running FastQC ..."
fastqc *.fastq*

mkdir -p ~/dc_workshop/results/fastqc_untrimmed_reads

echo "Saving FastQC results..."
mv *.zip ~/dc_workshop/results/fastqc_untrimmed_reads/
mv *.html ~/dc_workshop/results/fastqc_untrimmed_reads/

cd ~/dc_workshop/results/fastqc_untrimmed_reads/

echo "Unzipping..."
for filename in *.zip
    do
    unzip $filename
    done

echo "Saving summary..."
cat */summary.txt &gt; ~/dc_workshop/docs/fastqc_summaries.txt</code></pre>
</div>
<p>Save your file and exit <code>nano</code>. We can now run our
script:</p>
<div class="codewrapper sourceCode" id="cb15">
<h3 class="code-label">BASH<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="sourceCode bash" tabindex="0"><code class="sourceCode bash"><span id="cb15-1"><a href="#cb15-1" tabindex="-1"></a><span class="ex">$</span> bash read_qc.sh</span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>Running FastQC ...
Started analysis of SRR2584866.fastq
Approx 5% complete for SRR2584866.fastq
Approx 10% complete for SRR2584866.fastq
Approx 15% complete for SRR2584866.fastq
Approx 20% complete for SRR2584866.fastq
Approx 25% complete for SRR2584866.fastq
.
.
. </code></pre>
</div>
<p>For each of your sample files, FastQC will ask if you want to replace
the existing version with a new version. This is because we have already
run FastQC on this samples files and generated all of the outputs. We
are now doing this again using our scripts. Go ahead and select
<code>A</code> each time this message appears. It will appear once per
sample file (six times total).</p>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>replace SRR2584866_fastqc/Icons/fastqc_icon.png? [y]es, [n]o, [A]ll, [N]one, [r]ename:</code></pre>
</div>
</section><section><h2 class="section-heading" id="automating-the-rest-of-our-variant-calling-workflow">Automating the rest of our variant calling workflow<a class="anchor" aria-label="anchor" href="#automating-the-rest-of-our-variant-calling-workflow"></a>
</h2>
<hr class="half-width">
<p>We can extend these principles to the entire variant calling
workflow. To do this, we will take all of the individual commands that
we wrote before, put them into a single file, add variables so that the
script knows to iterate through our input files and write to the
appropriate output files. This is very similar to what we did with our
<code>read_qc.sh</code> script, but will be a bit more complex.</p>
<p>Download the script from <a href="../files/run_variant_calling.sh">here</a>. Download to
<code>~/dc_workshop/scripts</code>.</p>
<div class="codewrapper sourceCode" id="cb18">
<h3 class="code-label">BASH<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="sourceCode bash" tabindex="0"><code class="sourceCode bash"><span id="cb18-1"><a href="#cb18-1" tabindex="-1"></a><span class="ex">curl</span> <span class="at">-O</span> https://datacarpentry.org/wrangling-genomics/files/run_variant_calling.sh</span></code></pre>
</div>
<p>Our variant calling workflow has the following steps:</p>
<ol style="list-style-type: decimal">
<li>Index the reference genome for use by bwa and samtools.</li>
<li>Align reads to reference genome.</li>
<li>Convert the format of the alignment to sorted BAM, with some
intermediate steps.</li>
<li>Calculate the read coverage of positions in the genome.</li>
<li>Detect the single nucleotide variants (SNVs).</li>
<li>Filter and report the SNVs in VCF (variant calling format).</li>
</ol>
<p>Let’s go through this script together:</p>
<div class="codewrapper sourceCode" id="cb19">
<h3 class="code-label">BASH<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="sourceCode bash" tabindex="0"><code class="sourceCode bash"><span id="cb19-1"><a href="#cb19-1" tabindex="-1"></a><span class="ex">$</span> cd ~/dc_workshop/scripts</span>
<span id="cb19-2"><a href="#cb19-2" tabindex="-1"></a><span class="ex">$</span> less run_variant_calling.sh</span></code></pre>
</div>
<p>The script should look like this:</p>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>set -e
cd ~/dc_workshop/results

genome=~/dc_workshop/data/ref_genome/ecoli_rel606.fasta

bwa index $genome

mkdir -p sam bam bcf vcf

for fq1 in ~/dc_workshop/data/trimmed_fastq_small/*_1.trim.sub.fastq
    do
    echo "working with file $fq1"

    base=$(basename $fq1 _1.trim.sub.fastq)
    echo "base name is $base"

    fq1=~/dc_workshop/data/trimmed_fastq_small/${base}_1.trim.sub.fastq
    fq2=~/dc_workshop/data/trimmed_fastq_small/${base}_2.trim.sub.fastq
    sam=~/dc_workshop/results/sam/${base}.aligned.sam
    bam=~/dc_workshop/results/bam/${base}.aligned.bam
    sorted_bam=~/dc_workshop/results/bam/${base}.aligned.sorted.bam
    raw_bcf=~/dc_workshop/results/bcf/${base}_raw.bcf
    variants=~/dc_workshop/results/vcf/${base}_variants.vcf
    final_variants=~/dc_workshop/results/vcf/${base}_final_variants.vcf

    bwa mem $genome $fq1 $fq2 &gt; $sam
    samtools view -S -b $sam &gt; $bam
    samtools sort -o $sorted_bam $bam
    samtools index $sorted_bam
    bcftools mpileup -O b -o $raw_bcf -f $genome $sorted_bam
    bcftools call --ploidy 1 -m -v -o $variants $raw_bcf
    vcfutils.pl varFilter $variants &gt; $final_variants

    done</code></pre>
</div>
<p>Now, we will go through each line in the script before running
it.</p>
<p>First, notice that we change our working directory so that we can
create new results subdirectories in the right location.</p>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>cd ~/dc_workshop/results</code></pre>
</div>
<p>Next we tell our script where to find the reference genome by
assigning the <code>genome</code> variable to the path to our reference
genome:</p>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>genome=~/dc_workshop/data/ref_genome/ecoli_rel606.fasta</code></pre>
</div>
<p>Next we index our reference genome for BWA:</p>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>bwa index $genome</code></pre>
</div>
<p>And create the directory structure to store our results in:</p>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>mkdir -p sam bam bcf vcf</code></pre>
</div>
<p>Then, we use a loop to run the variant calling workflow on each of
our FASTQ files. The full list of commands within the loop will be
executed once for each of the FASTQ files in the
<code>data/trimmed_fastq_small/</code> directory. We will include a few
<code>echo</code> statements to give us status updates on our
progress.</p>
<p>The first thing we do is assign the name of the FASTQ file we are
currently working with to a variable called <code>fq1</code> and tell
the script to <code>echo</code> the filename back to us so we can check
which file we are on.</p>
<div class="codewrapper sourceCode" id="cb25">
<h3 class="code-label">BASH<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="sourceCode bash" tabindex="0"><code class="sourceCode bash"><span id="cb25-1"><a href="#cb25-1" tabindex="-1"></a><span class="cf">for</span> fq1 <span class="kw">in</span> ~/dc_workshop/data/trimmed_fastq_small/<span class="pp">*</span>_1.trim.sub.fastq</span>
<span id="cb25-2"><a href="#cb25-2" tabindex="-1"></a>    <span class="cf">do</span></span>
<span id="cb25-3"><a href="#cb25-3" tabindex="-1"></a>    <span class="bu">echo</span> <span class="st">"working with file </span><span class="va">$fq1</span><span class="st">"</span></span></code></pre>
</div>
<p>We then extract the base name of the file (excluding the path and
<code>.fastq</code> extension) and assign it to a new variable called
<code>base</code>.</p>
<div class="codewrapper sourceCode" id="cb26">
<h3 class="code-label">BASH<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="sourceCode bash" tabindex="0"><code class="sourceCode bash"><span id="cb26-1"><a href="#cb26-1" tabindex="-1"></a>    <span class="va">base</span><span class="op">=</span><span class="va">$(</span><span class="fu">basename</span> <span class="va">$fq1</span> _1.trim.sub.fastq<span class="va">)</span></span>
<span id="cb26-2"><a href="#cb26-2" tabindex="-1"></a>    <span class="bu">echo</span> <span class="st">"base name is </span><span class="va">$base</span><span class="st">"</span></span></code></pre>
</div>
<p>We can use the <code>base</code> variable to access both the
<code>base_1.fastq</code> and <code>base_2.fastq</code> input files, and
create variables to store the names of our output files. This makes the
script easier to read because we do not need to type out the full name
of each of the files: instead, we use the <code>base</code> variable,
but add a different extension (e.g. <code>.sam</code>,
<code>.bam</code>) for each file produced by our workflow.</p>
<div class="codewrapper sourceCode" id="cb27">
<h3 class="code-label">BASH<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="sourceCode bash" tabindex="0"><code class="sourceCode bash"><span id="cb27-1"><a href="#cb27-1" tabindex="-1"></a>    <span class="co">#input fastq files</span></span>
<span id="cb27-2"><a href="#cb27-2" tabindex="-1"></a>    <span class="va">fq1</span><span class="op">=</span>~/dc_workshop/data/trimmed_fastq_small/<span class="va">${base}</span>_1.trim.sub.fastq</span>
<span id="cb27-3"><a href="#cb27-3" tabindex="-1"></a>    <span class="va">fq2</span><span class="op">=</span>~/dc_workshop/data/trimmed_fastq_small/<span class="va">${base}</span>_2.trim.sub.fastq</span>
<span id="cb27-4"><a href="#cb27-4" tabindex="-1"></a>    </span>
<span id="cb27-5"><a href="#cb27-5" tabindex="-1"></a>    <span class="co"># output files</span></span>
<span id="cb27-6"><a href="#cb27-6" tabindex="-1"></a>    <span class="va">sam</span><span class="op">=</span>~/dc_workshop/results/sam/<span class="va">${base}</span>.aligned.sam</span>
<span id="cb27-7"><a href="#cb27-7" tabindex="-1"></a>    <span class="va">bam</span><span class="op">=</span>~/dc_workshop/results/bam/<span class="va">${base}</span>.aligned.bam</span>
<span id="cb27-8"><a href="#cb27-8" tabindex="-1"></a>    <span class="va">sorted_bam</span><span class="op">=</span>~/dc_workshop/results/bam/<span class="va">${base}</span>.aligned.sorted.bam</span>
<span id="cb27-9"><a href="#cb27-9" tabindex="-1"></a>    <span class="va">raw_bcf</span><span class="op">=</span>~/dc_workshop/results/bcf/<span class="va">${base}</span>_raw.bcf</span>
<span id="cb27-10"><a href="#cb27-10" tabindex="-1"></a>    <span class="va">variants</span><span class="op">=</span>~/dc_workshop/results/bcf/<span class="va">${base}</span>_variants.vcf</span>
<span id="cb27-11"><a href="#cb27-11" tabindex="-1"></a>    <span class="va">final_variants</span><span class="op">=</span>~/dc_workshop/results/vcf/<span class="va">${base}</span>_final_variants.vcf     </span></code></pre>
</div>
<p>And finally, the actual workflow steps:</p>
<ol style="list-style-type: decimal">
<li>align the reads to the reference genome and output a
<code>.sam</code> file:</li>
</ol>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>    bwa mem $genome $fq1 $fq2 &gt; $sam</code></pre>
</div>
<ol start="2" style="list-style-type: decimal">
<li>convert the SAM file to BAM format:</li>
</ol>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>    samtools view -S -b $sam &gt; $bam</code></pre>
</div>
<ol start="3" style="list-style-type: decimal">
<li>sort the BAM file:</li>
</ol>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>    samtools sort -o $sorted_bam $bam </code></pre>
</div>
<ol start="4" style="list-style-type: decimal">
<li>index the BAM file for display purposes:</li>
</ol>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>    samtools index $sorted_bam</code></pre>
</div>
<ol start="5" style="list-style-type: decimal">
<li>calculate the read coverage of positions in the genome:</li>
</ol>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>    bcftools mpileup -O b -o $raw_bcf -f $genome $sorted_bam </code></pre>
</div>
<ol start="6" style="list-style-type: decimal">
<li>call SNVs with bcftools:</li>
</ol>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>    bcftools call --ploidy 1 -m -v -o $variants $raw_bcf </code></pre>
</div>
<ol start="7" style="list-style-type: decimal">
<li>filter and report the SNVs in variant calling format (VCF):</li>
</ol>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>    vcfutils.pl varFilter $variants  &gt; $final_variants</code></pre>
</div>
<div id="exercise" class="callout discussion">
<div class="callout-square">
<i class="callout-icon" data-feather="message-circle"></i>
</div>
<div id="exercise" class="callout-inner">
<h3 class="callout-title">Exercise</h3>
<div class="callout-content">
<p>It is a good idea to add comments to your code so that you (or a
collaborator) can make sense of what you did later. Look through your
existing script. Discuss with a neighbor where you should add comments.
Add comments (anything following a <code>#</code> character will be
interpreted as a comment, bash will not try to run these comments as
code).</p>
</div>
</div>
</div>
<p>Now we can run our script:</p>
<div class="codewrapper sourceCode" id="cb35">
<h3 class="code-label">BASH<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="sourceCode bash" tabindex="0"><code class="sourceCode bash"><span id="cb35-1"><a href="#cb35-1" tabindex="-1"></a><span class="ex">$</span> bash run_variant_calling.sh</span></code></pre>
</div>
<div id="exercise-1" class="callout challenge">
<div class="callout-square">
<i class="callout-icon" data-feather="zap"></i>
</div>
<div id="exercise-1" class="callout-inner">
<h3 class="callout-title">Exercise</h3>
<div class="callout-content">
<p>The samples we just performed variant calling on are part of the
long-term evolution experiment introduced at the beginning of our
variant calling workflow. From the metadata table, we know that
SRR2589044 was from generation 5000, SRR2584863 was from generation
15000, and SRR2584866 was from generation 50000. How did the number of
mutations per sample change over time? Examine the metadata table. What
is one reason the number of mutations may have changed the way they
did?</p>
<p>Hint: You can find a copy of the output files for the subsampled
trimmed FASTQ file variant calling in the
<code>~/.solutions/wrangling-solutions/variant_calling_auto/</code>
directory.</p>
</div>
</div>
</div>
<div id="accordionSolution1" class="accordion challenge-accordion accordion-flush">
<div class="accordion-item">
<button class="accordion-button solution-button collapsed" type="button" data-bs-toggle="collapse" data-bs-target="#collapseSolution1" aria-expanded="false" aria-controls="collapseSolution1">
  <h4 class="accordion-header" id="headingSolution1"> Show me the solution </h4>
</button>
<div id="collapseSolution1" class="accordion-collapse collapse" aria-labelledby="headingSolution1" data-bs-parent="#accordionSolution1">
<div class="accordion-body">
<div class="codewrapper sourceCode" id="cb36">
<h3 class="code-label">BASH<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="sourceCode bash" tabindex="0"><code class="sourceCode bash"><span id="cb36-1"><a href="#cb36-1" tabindex="-1"></a><span class="ex">$</span> for infile in ~/dc_workshop/results/vcf/<span class="pp">*</span>_final_variants.vcf</span>
<span id="cb36-2"><a href="#cb36-2" tabindex="-1"></a><span class="op">&gt;</span> do</span>
<span id="cb36-3"><a href="#cb36-3" tabindex="-1"></a><span class="op">&gt;</span>     echo <span class="va">${infile}</span></span>
<span id="cb36-4"><a href="#cb36-4" tabindex="-1"></a><span class="op">&gt;</span>     grep <span class="ex">-v</span> <span class="st">"#"</span> <span class="va">${infile}</span> <span class="kw">|</span> <span class="fu">wc</span> <span class="at">-l</span></span>
<span id="cb36-5"><a href="#cb36-5" tabindex="-1"></a><span class="op">&gt;</span> done</span></code></pre>
</div>
<p>For SRR2589044 from generation 5000 there were 10 mutations, for
SRR2584863 from generation 15000 there were 25 mutations, and SRR2584866
from generation 50000 there were 766 mutations. In the last generation,
a hypermutable phenotype had evolved, causing this strain to have more
mutations.</p>
</div>
</div>
</div>
</div>
<div id="bonus-exercise" class="callout discussion">
<div class="callout-square">
<i class="callout-icon" data-feather="message-circle"></i>
</div>
<div id="bonus-exercise" class="callout-inner">
<h3 class="callout-title">Bonus exercise</h3>
<div class="callout-content">
<p>If you have time after completing the previous exercise, use
<code>run_variant_calling.sh</code> to run the variant calling pipeline
on the full-sized trimmed FASTQ files. You should have a copy of these
already in <code>~/dc_workshop/data/trimmed_fastq</code>, but if you do
not, there is a copy in
<code>~/.solutions/wrangling-solutions/trimmed_fastq</code>. Does the
number of variants change per sample?</p>
</div>
</div>
</div>
<div id="keypoints1" class="callout keypoints">
<div class="callout-square">
<i class="callout-icon" data-feather="key"></i>
</div>
<div class="callout-inner">
<h3 class="callout-title">Key Points</h3>
<div class="callout-content">
<ul>
<li>We can combine multiple commands into a shell script to automate a
workflow.</li>
<li>Use <code>echo</code> statements within your scripts to get an
automated progress update.</li>
</ul>
</div>
</div>
</div>
</section></section>
</div>
    </main>
</div>
<!-- END  : inst/pkgdown/templates/content-extra.html -->

      </div>
<!--/div.row-->
      		<footer class="row footer mx-md-3"><hr>
<div class="col-md-6">
        <p>This lesson is subject to the <a href="CODE_OF_CONDUCT.html">Code of Conduct</a></p>
        <p>

        <a href="https://github.com/datacarpentry/wrangling-genomics/edit/main/README.md" class="external-link">Edit on GitHub</a>

	
        | <a href="https://github.com/datacarpentry/wrangling-genomics/blob/main/CONTRIBUTING.md" class="external-link">Contributing</a>
        | <a href="https://github.com/datacarpentry/wrangling-genomics/" class="external-link">Source</a></p>
				<p><a href="https://github.com/datacarpentry/wrangling-genomics/blob/main/CITATION" class="external-link">Cite</a> | <a href="mailto:team@carpentries.org">Contact</a> | <a href="https://carpentries.org/about/" class="external-link">About</a></p>
			</div>
			<div class="col-md-6">

        <p>Materials licensed under <a href="LICENSE.html">CC-BY 4.0</a> by the authors</p>

        <p>Template licensed under <a href="https://creativecommons.org/licenses/by-sa/4.0/" class="external-link">CC-BY 4.0</a> by <a href="https://carpentries.org/" class="external-link">The Carpentries</a></p>
        <p>Built with <a href="https://github.com/carpentries/sandpaper/tree/0.16.11" class="external-link">sandpaper (0.16.11)</a>, <a href="https://github.com/carpentries/pegboard/tree/0.7.9" class="external-link">pegboard (0.7.9)</a>, and <a href="https://github.com/carpentries/varnish/tree/1.0.5" class="external-link">varnish (1.0.5)</a></p>
			</div>
		</footer>
</div> <!-- / div.container -->
	<div id="to-top">
		<a href="#top">
      <i class="search-icon" data-feather="arrow-up" role="img" aria-label="Back To Top"></i><br><!-- <span class="d-none d-sm-none d-md-none d-lg-none d-xl-block">Back</span> To Top --><span class="d-none d-sm-none d-md-none d-lg-none d-xl-block">Back</span> To Top
		</a>
	</div>
  <script type="application/ld+json">
    {
  "@context": "https://schema.org",
  "@type": "TrainingMaterial",
  "@id": "https://datacarpentry.github.io/wrangling-genomics/instructor/aio.html",
  "inLanguage": "en",
  "dct:conformsTo": "https://bioschemas.org/profiles/TrainingMaterial/1.0-RELEASE",
  "description": "A Carpentries Lesson teaching foundational data and coding skills to researchers worldwide",
  "keywords": "software, data, lesson, The Carpentries",
  "name": "All in One View",
  "creativeWorkStatus": "active",
  "url": "https://datacarpentry.github.io/wrangling-genomics/instructor/aio.html",
  "identifier": "https://datacarpentry.github.io/wrangling-genomics/instructor/aio.html",
  "dateCreated": "2015-03-24",
  "dateModified": "2025-04-15",
  "datePublished": "2025-04-15"
}

  </script><script>
		feather.replace();
	</script><!-- Matomo --><script>
          var _paq = window._paq = window._paq || [];
          /* tracker methods like "setCustomDimension" should be called before "trackPageView" */
          _paq.push(["setDocumentTitle", document.domain + "/" + document.title]);
          _paq.push(["setDomains", ["*.lessons.carpentries.org","*.datacarpentry.github.io","*.datacarpentry.org","*.librarycarpentry.github.io","*.librarycarpentry.org","*.swcarpentry.github.io", "*.carpentries.github.io"]]);
          _paq.push(["setDoNotTrack", true]);
          _paq.push(["disableCookies"]);
          _paq.push(["trackPageView"]);
          _paq.push(["enableLinkTracking"]);
          (function() {
              var u="https://matomo.carpentries.org/";
              _paq.push(["setTrackerUrl", u+"matomo.php"]);
              _paq.push(["setSiteId", "1"]);
              var d=document, g=d.createElement("script"), s=d.getElementsByTagName("script")[0];
              g.async=true; g.src="https://matomo.carpentries.org/matomo.js"; s.parentNode.insertBefore(g,s);
          })();
        </script><!-- End Matomo Code -->
</body>
</html><!-- END:   inst/pkgdown/templates/layout.html-->

